<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Linux Basics &mdash; ROPES summer school - CRG - May 2023  documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
    <link rel="canonical" href="https://biocorecrg.github.io/ropes-linux-mop2-2023/linux_basics.html" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="author" title="About these documents" href="about.html" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Containers" href="containers.html" />
    <link rel="prev" title="About the course" href="about.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            ROPES summer school - CRG - May 2023
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="about.html">About the course</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Linux Basics</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#introduction-to-linux">Introduction to Linux</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#what-is-linux">What is Linux?</a></li>
<li class="toctree-l3"><a class="reference internal" href="#what-is-a-command-line-interface-gui">What is a Command Line Interface (GUI)?</a></li>
<li class="toctree-l3"><a class="reference internal" href="#practical-part">Practical part</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#create-directory-and-practice-moving-around">Create directory and practice moving around</a></li>
<li class="toctree-l4"><a class="reference internal" href="#recap"><strong>Recap</strong></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#download-files-from-repositories">Download files from repositories</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id1"><strong>Recap</strong></a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="containers.html">Containers</a></li>
<li class="toctree-l1"><a class="reference internal" href="nextflow.html">Nextflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="mop2_practical.html">Master of Pores 2 practical</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">ROPES summer school - CRG - May 2023</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Linux Basics</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/linux_basics.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="linux-basics">
<span id="linux-basics-page"></span><h1>Linux Basics<a class="headerlink" href="#linux-basics" title="Permalink to this heading"></a></h1>
<section id="introduction-to-linux">
<h2>Introduction to Linux<a class="headerlink" href="#introduction-to-linux" title="Permalink to this heading"></a></h2>
<ul class="simple">
<li><p><a class="reference internal" href="#what-is-linux">What is Linux?</a></p></li>
<li><p><a class="reference internal" href="#what-is-a-command-line-interface-gui">What is a Command Line Interface (GUI)?</a></p></li>
<li><p><a href="#id2"><span class="problematic" id="id3">`Practical part I`_</span></a></p></li>
<li><p><a class="reference internal" href="#download-files-from-repositories">Download files from repositories</a></p></li>
<li><p><a href="#id4"><span class="problematic" id="id5">`Practical part II`_</span></a></p></li>
</ul>
<section id="what-is-linux">
<h3>What is Linux?<a class="headerlink" href="#what-is-linux" title="Permalink to this heading"></a></h3>
<a class="reference internal image-reference" href="_images/Linux-Logo.png"><img alt="_images/Linux-Logo.png" src="_images/Linux-Logo.png" style="width: 400px;" /></a>
<p>Linux is an open-source operating system that was developed in the 1990s by <a class="reference external" href="https://en.wikipedia.org/wiki/Linus_Torvalds">Linus Torvalds</a>. Unlike other popular operating systems like Windows and MacOS, Linux is free to download, use, and distribute.</p>
<p>One of the most distinctive features of Linux is that it is highly customizable. Users can choose from a wide variety of “distributions” (or “distros”) of Linux, each with its own unique set of pre-installed software and user interface. Some popular distros include <a class="reference external" href="https://ubuntu.com/">Ubuntu</a>, <a class="reference external" href="https://www.debian.org/">Debian</a>, and <a class="reference external" href="https://fedoraproject.org/">Fedora</a>.</p>
<p>Linux is known for being incredibly stable and secure. It is also highly efficient, which means it can run well even on older or less powerful hardware. This has made Linux a popular choice for servers and other industrial applications.</p>
<p>Because it is <strong>open-source</strong>, Linux has a large and active community of developers who are constantly working to improve the operating system. This means that there are always new updates and features being added to Linux, and users can be confident that their operating system is always up-to-date and secure.</p>
<p>Linux is a is widely used in <strong>bioinformatics</strong> due to its ability to handle large data sets and complex computational tasks. Bioinformaticians use Linux to analyze DNA sequences, study gene expression, and develop computational models of biological systems.</p>
<p>One of the most important features of Linux for bioinformatics is its support for powerful command-line tools. These tools allow bioinformaticians to perform complex data processing and analysis tasks with ease. In addition, Linux provides a rich array of programming languages and libraries that are essential for bioinformatics research, such as <strong>Python</strong>, <strong>Java</strong>, <strong>Perl</strong>, <strong>C</strong> and <strong>C++</strong>.</p>
<p>One of the key advantages of Linux for bioinformatics is its ability to run on a wide range of hardware, from powerful servers to small embedded devices. This makes it possible to run bioinformatics analyses on a variety of different systems, including <strong>high-performance computing clusters</strong>, <strong>cloud-based platforms</strong>, and <strong>personal computers</strong>.</p>
<p>Another advantage of Linux for bioinformatics is its strong focus on security and reliability. Because Linux is <strong>open source</strong>, the community is constantly working to improve the operating system and fix any security vulnerabilities that are discovered. This makes Linux a trusted platform for sensitive bioinformatics data.</p>
</section>
<section id="what-is-a-command-line-interface-gui">
<h3>What is a Command Line Interface (GUI)?<a class="headerlink" href="#what-is-a-command-line-interface-gui" title="Permalink to this heading"></a></h3>
<a class="reference internal image-reference" href="_images/cli.png"><img alt="_images/cli.png" src="_images/cli.png" style="width: 600px;" /></a>
<p>A command line interface (<strong>CLI</strong>) is a text-based interface used to interact with a computer’s operating system or software by entering commands through a command prompt.</p>
<p>The command prompt usually consists of a text area where the user can enter a command, and the output of the command is displayed in the same area.</p>
<p><strong>Commands</strong> can be entered using specific keywords or phrases, which are interpreted by the operating system or software.</p>
<p>For example, in the Windows command prompt, the user can type “<strong>dir</strong>” to list the files and directories in the current directory, and in the Unix/Linux command line, the user can type “<strong>ls</strong>” to achieve the same result.</p>
<p>In addition to simple commands, more complex operations can be performed by chaining commands together using special symbols.</p>
<p>Overall, command line interfaces offer a fast and powerful way to interact with a computer or software system, and are often preferred to graphical user interface (<strong>GUI</strong>) by experienced users or developers due to their flexibility and efficiency.</p>
</section>
<section id="practical-part">
<h3>Practical part<a class="headerlink" href="#practical-part" title="Permalink to this heading"></a></h3>
<section id="create-directory-and-practice-moving-around">
<h4>Create directory and practice moving around<a class="headerlink" href="#create-directory-and-practice-moving-around" title="Permalink to this heading"></a></h4>
<p>To create file and folders in linux is quite simple. You can use a number of programs for creating an empty file (<strong>touch</strong>) or an empty directory (<strong>mkdir</strong>)</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>touch<span class="w"> </span>my_beautiful_file.txt

mkdir<span class="w"> </span>my_beautiful_folder
</pre></div>
</div>
<p>To display the list of files and folder we can use the command <strong>ls</strong></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>ls
my_beautiful_file.txt<span class="w">  </span>my_beautiful_folder
</pre></div>
</div>
<p>To change the name of a file (or a directory) you can use the command <strong>mv</strong> while for copying the file you can use <strong>cp</strong>. Adding the option <strong>-r</strong> (recursive) to <strong>cp</strong> allows to copy a whole folder and its content.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>mv<span class="w"> </span>my_beautiful_file.txt<span class="w"> </span>my_ugly_file.txt
mv<span class="w"> </span>my_beautiful_folder<span class="w"> </span>my_ugly_folder

cp<span class="w"> </span>my_ugly_file.txt<span class="w"> </span>my_beautiful_file.txt
cp<span class="w"> </span>my_ugly_folder<span class="w"> </span>-r<span class="w"> </span>my_beautiful_folder
</pre></div>
</div>
<p>If you omit the <strong>-r</strong> option the system will complain</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>cp<span class="w"> </span>my_ugly_folder<span class="w"> </span>my_other_folder
</pre></div>
</div>
<p>You can use <strong>mv</strong> also for moving a file (or a directory) inside a folder. Also <strong>cp</strong> will allow you to make a copy inside a folder.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>mv<span class="w"> </span>my_beautiful_file.txt<span class="w"> </span>my_beautiful_folder
cp<span class="w"> </span>my_ugly_file.txt<span class="w"> </span>my_ugly_folder

ls

my_beautiful_folder<span class="w">  </span>my_ugly_file.txt<span class="w">  </span>my_ugly_folder
</pre></div>
</div>
<p>For entering in a folder we can use the tool <strong>cd</strong></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">cd</span><span class="w"> </span>my_ugly_folder

ls

my_ugly_file.txt
</pre></div>
</div>
<p>For going out we can move one level out</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">cd</span><span class="w"> </span>../

ls

my_beautiful_folder<span class="w">  </span>my_ugly_file.txt<span class="w">  </span>my_ugly_folder
</pre></div>
</div>
<p>Sometimes we get lost and would like to know where we are.</p>
<a class="reference internal image-reference" href="_images/lost.png"><img alt="_images/lost.png" src="_images/lost.png" style="width: 600px;" /></a>
<p>We can use the command <strong>pwd</strong>.</p>
<p>We can write to a file using the character <strong>&gt;</strong>, that means output redirection.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">echo</span><span class="w"> </span><span class="s2">&quot;ATGTACTGACTGCATGCATGCCATGCA&quot;</span><span class="w"> </span>&gt;<span class="w"> </span>my_dna.txt
</pre></div>
</div>
<p>And display the content of the file using the program <strong>cat</strong></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>cat<span class="w"> </span>my_dna.txt

ATGTACTGACTGCATGCATGCCATGCA
</pre></div>
</div>
<p>To convert this sequence to a RNA one we can just replace the <strong>T</strong> base with <strong>U</strong> by using the program <strong>sed</strong>. The sintax of this program is the following <strong>s/&lt;TO BE REPLACED&gt;/&lt;TO REPLACE&gt;/</strong>.</p>
<p>You can add a <strong>g</strong> at the end if you want to replace every character found <strong>s/&lt;TO BE REPLACED&gt;/&lt;TO REPLACE&gt;/g</strong>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sed<span class="w"> </span>s/T/U/g<span class="w"> </span>my_dna.txt<span class="w"> </span>&gt;<span class="w"> </span>my_rna.txt

cat<span class="w"> </span>my_rna.txt

AUGUACUGACUGCAUGCAUGCCAUGCA
</pre></div>
</div>
<p>Every command has a manual, you can read it by using the program <strong>man</strong> with the name of the tool.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>man<span class="w"> </span>ls

LS<span class="o">(</span><span class="m">1</span><span class="o">)</span><span class="w">                                                                   </span>User<span class="w"> </span>Commands<span class="w">                                                                   </span>LS<span class="o">(</span><span class="m">1</span><span class="o">)</span>

NAME
<span class="w">      </span>ls<span class="w"> </span>-<span class="w"> </span>list<span class="w"> </span>directory<span class="w"> </span>contents

SYNOPSIS
<span class="w">       </span>ls<span class="w"> </span><span class="o">[</span>OPTION<span class="o">]</span>...<span class="w"> </span><span class="o">[</span>FILE<span class="o">]</span>...

DESCRIPTION
<span class="w">       </span>List<span class="w"> </span>information<span class="w"> </span>about<span class="w"> </span>the<span class="w"> </span>FILEs<span class="w"> </span><span class="o">(</span>the<span class="w"> </span>current<span class="w"> </span>directory<span class="w"> </span>by<span class="w"> </span>default<span class="o">)</span>.<span class="w">  </span>Sort<span class="w"> </span>entries<span class="w"> </span>alphabetically<span class="w"> </span><span class="k">if</span><span class="w"> </span>none<span class="w"> </span>of<span class="w"> </span>-cftuvSUX<span class="w"> </span>nor<span class="w"> </span>--sort<span class="w"> </span>is<span class="w"> </span>specified.

<span class="w">       </span>Mandatory<span class="w"> </span>arguments<span class="w"> </span>to<span class="w"> </span>long<span class="w"> </span>options<span class="w"> </span>are<span class="w"> </span>mandatory<span class="w"> </span><span class="k">for</span><span class="w"> </span>short<span class="w"> </span>options<span class="w"> </span>too.

<span class="w">       </span>-a,<span class="w"> </span>--all
<span class="w">              </span><span class="k">do</span><span class="w"> </span>not<span class="w"> </span>ignore<span class="w"> </span>entries<span class="w"> </span>starting<span class="w"> </span>with<span class="w"> </span>.

<span class="w">      </span>-A,<span class="w"> </span>--almost-all
<span class="w">              </span><span class="k">do</span><span class="w"> </span>not<span class="w"> </span>list<span class="w"> </span>implied<span class="w"> </span>.<span class="w"> </span>and<span class="w"> </span>..

<span class="w">      </span>--author
<span class="w">            </span>with<span class="w"> </span>-l,<span class="w"> </span>print<span class="w"> </span>the<span class="w"> </span>author<span class="w"> </span>of<span class="w"> </span>each<span class="w"> </span>file

<span class="w">     </span>-b,<span class="w"> </span>--escape
<span class="w">            </span>print<span class="w"> </span>C-style<span class="w"> </span>escapes<span class="w"> </span><span class="k">for</span><span class="w"> </span>nongraphic<span class="w"> </span>characters
Manual<span class="w"> </span>page<span class="w"> </span>ls<span class="o">(</span><span class="m">1</span><span class="o">)</span><span class="w"> </span>line<span class="w"> </span><span class="m">1</span><span class="w"> </span><span class="o">(</span>press<span class="w"> </span>h<span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="nb">help</span><span class="w"> </span>or<span class="w"> </span>q<span class="w"> </span>to<span class="w"> </span>quit<span class="o">)</span>
</pre></div>
</div>
</section>
<section id="recap">
<h4><strong>Recap</strong><a class="headerlink" href="#recap" title="Permalink to this heading"></a></h4>
<ul class="simple">
<li><p><strong>touch</strong> writes empty files <strong>mkdir</strong> empty directories</p></li>
<li><p><strong>mv</strong> move files (or directory) or change their name</p></li>
<li><p><strong>ls</strong> list files and directories</p></li>
<li><p><strong>cp</strong> copy files and direcotries</p></li>
<li><p><strong>cd</strong> change the directory</p></li>
<li><p><strong>echo</strong> print values to standard output</p></li>
<li><p><strong>cat</strong> print the content of a file to standard output</p></li>
<li><p><strong>sed</strong> replace a string with another</p></li>
<li><p><strong>man</strong> print the manual for a function</p></li>
</ul>
</section>
</section>
<section id="download-files-from-repositories">
<h3>Download files from repositories<a class="headerlink" href="#download-files-from-repositories" title="Permalink to this heading"></a></h3>
<p>Several institutions host different kind of genomics data.</p>
<p>For example the genome browser <a class="reference external" href="https://www.ensembl.org/index.html">Ensembl</a> is also a public repository of genomes and annotation that can be freely downloaded and used for any kind of analysis</p>
<p>The resource <a class="reference external" href="https://bacteria.ensembl.org/index.html">Ensembl Bacteria</a> contains a large number of bacterial genomes and their annotation. As an example we can browse the page corresponding to <a class="reference external" href="https://bacteria.ensembl.org/Escherichia_coli_bl21_gold_de3_plyss_ag_/Info/Index/">*Escherichia coli ‘BL21-Gold(DE3)pLysS AG’*</a></p>
<p>&lt;img src=”images/ensembl_escherichia.png” width=”800”/&gt;</p>
<p>We can click on “Download genes, cDNAs, ncRNA, proteins <strong>FASTA</strong>”</p>
<p>&lt;img src=”images/list_ensembl_escherichia.png” width=”800”/&gt;</p>
<p>And then on <strong>DNA</strong></p>
<p>&lt;img src=”images/file_list_escherichia.png” width=”800”/&gt;</p>
<p>Then as an example we can use the copy the link address of the <strong>README</strong> file using the mouse right button.</p>
<p>&lt;img src=”images/right_click.png” width=”800”/&gt;</p>
<p>Then we can go back to our command line and use the program <strong>wget</strong> to download that file and using <strong>CTRL+C</strong> to paste the address:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>wget<span class="w"> </span>ftp://ftp.ensemblgenomes.org/pub/bacteria/release-42/fasta/bacteria_22_collection/escherichia_coli_bl21_gold_de3_plyss_ag_/dna/README

--2019-03-06<span class="w"> </span><span class="m">18</span>:59:13--<span class="w">  </span>ftp://ftp.ensemblgenomes.org/pub/bacteria/release-42/fasta/bacteria_22_collection/escherichia_coli_bl21_gold_de3_plyss_ag_/dna/README
<span class="w">           </span><span class="o">=</span>&gt;<span class="w"> </span>‘README’
Resolving<span class="w"> </span>ftp.ensemblgenomes.org<span class="w"> </span><span class="o">(</span>ftp.ensemblgenomes.org<span class="o">)</span>...<span class="w"> </span><span class="m">193</span>.62.197.94
Connecting<span class="w"> </span>to<span class="w"> </span>ftp.ensemblgenomes.org<span class="w"> </span><span class="o">(</span>ftp.ensemblgenomes.org<span class="o">)</span><span class="p">|</span><span class="m">193</span>.62.197.94<span class="p">|</span>:21...<span class="w"> </span>connected.
Logging<span class="w"> </span><span class="k">in</span><span class="w"> </span>as<span class="w"> </span>anonymous<span class="w"> </span>...<span class="w"> </span>Logged<span class="w"> </span><span class="k">in</span>!
<span class="o">==</span>&gt;<span class="w"> </span>SYST<span class="w"> </span>...<span class="w"> </span><span class="k">done</span>.<span class="w">    </span><span class="o">==</span>&gt;<span class="w"> </span>PWD<span class="w"> </span>...<span class="w"> </span><span class="k">done</span>.
<span class="o">==</span>&gt;<span class="w"> </span>TYPE<span class="w"> </span>I<span class="w"> </span>...<span class="w"> </span><span class="k">done</span>.<span class="w">  </span><span class="o">==</span>&gt;<span class="w"> </span>CWD<span class="w"> </span><span class="o">(</span><span class="m">1</span><span class="o">)</span><span class="w"> </span>/pub/bacteria/release-42/fasta/bacteria_22_collection/escherichia_coli_bl21_gold_de3_plyss_ag_/dna<span class="w"> </span>...<span class="w"> </span><span class="k">done</span>.
<span class="o">==</span>&gt;<span class="w"> </span>SIZE<span class="w"> </span>README<span class="w"> </span>...<span class="w"> </span><span class="nv">4923</span>
<span class="o">==</span>&gt;<span class="w"> </span>PASV<span class="w"> </span>...<span class="w"> </span><span class="k">done</span>.<span class="w">    </span><span class="o">==</span>&gt;<span class="w"> </span>RETR<span class="w"> </span>README<span class="w"> </span>...<span class="w"> </span><span class="k">done</span>.
Length:<span class="w"> </span><span class="m">4923</span><span class="w"> </span><span class="o">(</span><span class="m">4</span>.8K<span class="o">)</span><span class="w"> </span><span class="o">(</span>unauthoritative<span class="o">)</span>

<span class="m">100</span>%<span class="o">[======================================================================================================================</span>&gt;<span class="o">]</span><span class="w"> </span><span class="m">4</span>,923<span class="w">       </span>--.-K/s<span class="w">   </span><span class="k">in</span><span class="w"> </span>0s

<span class="m">2019</span>-03-06<span class="w"> </span><span class="m">18</span>:59:14<span class="w"> </span><span class="o">(</span><span class="m">295</span><span class="w"> </span>MB/s<span class="o">)</span><span class="w"> </span>-<span class="w"> </span>‘README’<span class="w"> </span>saved<span class="w"> </span><span class="o">[</span><span class="m">4923</span><span class="o">]</span>
</pre></div>
</div>
<p>we can then use the program <strong>more</strong> to display part of the content of the file:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>more<span class="w"> </span>README


<span class="c1">#### README ####</span>

IMPORTANT:<span class="w"> </span>Please<span class="w"> </span>note<span class="w"> </span>you<span class="w"> </span>can<span class="w"> </span>download<span class="w"> </span>correlation<span class="w"> </span>data<span class="w"> </span>tables,
supported<span class="w"> </span>by<span class="w"> </span>Ensembl,<span class="w"> </span>via<span class="w"> </span>the<span class="w"> </span>highly<span class="w"> </span>customisable<span class="w"> </span>BioMart<span class="w"> </span>and
EnsMart<span class="w"> </span>data<span class="w"> </span>mining<span class="w"> </span>tools.<span class="w"> </span>See<span class="w"> </span>http://www.ensembl.org/biomart/martview<span class="w"> </span>or
http://www.ebi.ac.uk/biomart/<span class="w"> </span><span class="k">for</span><span class="w"> </span>more<span class="w"> </span>information.

The<span class="w"> </span>genome<span class="w"> </span>assembly<span class="w"> </span>represented<span class="w"> </span>here<span class="w"> </span>corresponds<span class="w"> </span>to
GCA_000023665.1

<span class="c1">#######################</span>
Fasta<span class="w"> </span>DNA<span class="w"> </span>dumps
<span class="c1">#######################</span>

-----------
FILE<span class="w"> </span>NAMES
------------
The<span class="w"> </span>files<span class="w"> </span>are<span class="w"> </span>consistently<span class="w"> </span>named<span class="w"> </span>following<span class="w"> </span>this<span class="w"> </span>pattern:
<span class="w">   </span>&lt;species&gt;.&lt;assembly&gt;.&lt;sequence<span class="w"> </span>type&gt;.&lt;id<span class="w"> </span>type&gt;.&lt;id&gt;.fa.gz

&lt;species&gt;:<span class="w">   </span>The<span class="w"> </span>systematic<span class="w"> </span>name<span class="w"> </span>of<span class="w"> </span>the<span class="w"> </span>species.
&lt;assembly&gt;:<span class="w">  </span>The<span class="w"> </span>assembly<span class="w"> </span>build<span class="w"> </span>name.
&lt;sequence<span class="w"> </span>type&gt;:
<span class="w"> </span>*<span class="w"> </span><span class="s1">&#39;dna&#39;</span><span class="w"> </span>-<span class="w"> </span>unmasked<span class="w"> </span>genomic<span class="w"> </span>DNA<span class="w"> </span>sequences.
--More--<span class="o">(</span><span class="m">14</span>%<span class="o">)</span>
</pre></div>
</div>
<p>Pressing the bar allows us to scroll down the file, while for exiting you just click <strong>CTRL+C</strong>.&lt;br&gt;
After reading the README we can download the file named <strong>toplevel</strong> that contains chromosomes, regions not assembled into chromosomes and N padded haplotype/patch regions:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>wget<span class="w"> </span>ftp://ftp.ensemblgenomes.org/pub/bacteria/release-42/fasta/bacteria_22_collection/escherichia_coli_bl21_gold_de3_plyss_ag_/dna/Escherichia_coli_bl21_gold_de3_plyss_ag_.ASM2366v1.dna.toplevel.fa.gz
<span class="w">                </span><span class="sb">```</span>
</pre></div>
</div>
<p>We can use the options <strong>-lh</strong> of the program <strong>ls</strong> to list attributes of the files and show in human readable format the size fo the files</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>ls<span class="w"> </span>-lh

total<span class="w"> </span><span class="m">2</span>.0M
drwxr-xr-x<span class="w"> </span><span class="m">5</span><span class="w"> </span>lcozzuto<span class="w"> </span>Bioinformatics_Unit<span class="w">  </span><span class="m">209</span><span class="w"> </span>Mar<span class="w">  </span><span class="m">7</span><span class="w"> </span><span class="m">11</span>:48<span class="w"> </span>advanced_linux_2019
-rw-r--r--<span class="w"> </span><span class="m">1</span><span class="w"> </span>lcozzuto<span class="w"> </span>Bioinformatics_Unit<span class="w"> </span><span class="m">1</span>.4M<span class="w"> </span>Mar<span class="w">  </span><span class="m">7</span><span class="w"> </span><span class="m">13</span>:06<span class="w"> </span>Escherichia_coli_bl21_gold_de3_plyss_ag_.ASM2366v1.dna.toplevel.fa.gz
drwxr-xr-x<span class="w"> </span><span class="m">2</span><span class="w"> </span>lcozzuto<span class="w"> </span>Bioinformatics_Unit<span class="w">   </span><span class="m">39</span><span class="w"> </span>Mar<span class="w">  </span><span class="m">6</span><span class="w"> </span><span class="m">18</span>:17<span class="w"> </span>my_beautiful_folder
-rw-r--r--<span class="w"> </span><span class="m">1</span><span class="w"> </span>lcozzuto<span class="w"> </span>Bioinformatics_Unit<span class="w">    </span><span class="m">0</span><span class="w"> </span>Mar<span class="w">  </span><span class="m">6</span><span class="w"> </span><span class="m">18</span>:15<span class="w"> </span>my_ugly_file.txt
drwxr-xr-x<span class="w"> </span><span class="m">2</span><span class="w"> </span>lcozzuto<span class="w"> </span>Bioinformatics_Unit<span class="w">   </span><span class="m">34</span><span class="w"> </span>Mar<span class="w">  </span><span class="m">6</span><span class="w"> </span><span class="m">18</span>:17<span class="w"> </span>my_ugly_folder
-rw-r--r--<span class="w"> </span><span class="m">1</span><span class="w"> </span>lcozzuto<span class="w"> </span>Bioinformatics_Unit<span class="w"> </span><span class="m">4</span>.9K<span class="w"> </span>Mar<span class="w">  </span><span class="m">6</span><span class="w"> </span><span class="m">18</span>:59<span class="w"> </span>README
</pre></div>
</div>
<p>For unzipping the file we can use the program <strong>gunzip</strong>. &lt;br&gt;The uncompressed file is now <strong>4.5M</strong>.</p>
<p>Let’s see the content of the file.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>more<span class="w"> </span>Escherichia_coli_bl21_gold_de3_plyss_ag_.ASM2366v1.dna.toplevel.fa

&gt;CP001665<span class="w"> </span>dna:supercontig<span class="w"> </span>supercontig:ASM2366v1:CP001665:1:4570938:1<span class="w"> </span>REF
CGTCCTGGATCTTTATTAGATCGATTAAGCCAATTTTTGTCTATGGTCATTAAATTTTCC
AATATGCGGCGTAAATCGTGCCCGCCTCGCGGCAGGATCGTTTACACTTAGCGAGTTCTG
GAAAGTCCTGTGGATAAATCGGGAAAATCTGTGAGAAACAGAAGATCTCTTGCGCAGTTT
AGGCTATGATCCGCGGTCCCGATCGTTTTGCAGGATCTTGATCGGGCATATAACCGCAGA
CAGCGGTTCGTGCGTCACCCTCAAGCAGGGTCTTTTCGACGTACGTCAACAATCATGAAT
GTTTCAGCCTTAGTCATTATCGACTTTTGTTCGAGTGGAGTCCGCCGTGTCACTTTCGCT
TTGGCAGCAGTGTCTTGCCCGATTGCAGGATGAGTTACCAGCCACAGAATTCAGTATGTG
GATACGCCCATTGCAGGCGGAACTGAGCGATAACACGCTGGCCCTGTACGCGCCAAACCG
TTTTGTCCTCGATTGGGTACGGGACAAGTACCTTAATAATATCAATGGACTGCTAACCAG
TTTCTGCGGAGCGGATGCCCCACAGCTGCGTTTTGAAGTCGGCACCAAACCGGTGACGCA
AACGCCACAAGCGGCAGTGACGAGCAACGTCGCGGCCCCTGCACAGGTGGCGCAAACGCA
GCCGCAACGTGCTGCGCCTTCTACGCGCTCAGGTTGGGATAACGTCCCGGCCCCGGCAGA
ACCGACCTATCGTTCTAACGTAAACGTCAAACACACGTTTGATAACTTCGTTGAAGGTAA
ATCTAACCAACTGGCGCGCGCGGCGGCTCGCCAGGTGGCGGATAACCCTGGCGGTGCCTA
TAACCCGTTGTTCCTTTATGGCGGCACGGGTCTGGGTAAAACTCACCTGCTGCATGCGGT
GGGTAACGGCATTATGGCGCGCAAGCCGAATGCCAAAGTGGTTTATATGCACTCCGAGCG
CTTTGTTCAGGACATGGTTAAAGCCCTGCAAAACAACGCGATCGAAGAGTTTAAACGCTA
CTACCGTTCCGTAGATGCACTGCTGATCGACGATATTCAGTTTTTTGCTAATAAAGAACG
ATCTCAGGAAGAGTTTTTCCACACCTTCAACGCCCTGCTGGAAGGTAATCAACAGATCAT
TCTCACCTCGGATCGCTATCCGAAAGAGATCAACGGCGTTGAGGATCGTTTGAAATCCCG
CTTCGGTTGGGGACTGACTGTGGCGATCGAACCGCCAGAGCTGGAAACCCGTGTGGCGAT
CCTGATGAAAAAGGCCGACGAAAACGACATTCGTTTGCCGGGCGAAGTGGCGTTCTTTAT
CGCCAAGCGTCTACGATCTAACGTACGTGAGCTGGAAGGGGCGCTGAACCGCGTCATTGC
</pre></div>
</div>
<p>The file contains the whole genome of the bacteria.</p>
<p>The first line contains the character <strong>&gt;</strong> and the name of the molecule / genome.</p>
<p>This format is called <a class="reference external" href="https://en.wikipedia.org/wiki/FASTA_format">**FASTA**</a> format and is universally used for storing one or multiple DNA/RNA/Protein sequences.</p>
<p>We can now download in the same ways the proteins:</p>
<p>&lt;img src=”images/righ_click_2.png” width=”800”/&gt;</p>
<p>and after unzipping the file we can have a look at it.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>more<span class="w"> </span>Escherichia_coli_bl21_gold_de3_plyss_ag_.ASM2366v1.pep.all.fa

&gt;ACT27082<span class="w"> </span>pep<span class="w"> </span>supercontig:ASM2366v1:CP001665:347:1750:1<span class="w"> </span>gene:ECBD_0001<span class="w"> </span>transcript:ACT27082<span class="w"> </span>gene_biotype:protein_coding<span class="w"> </span>transcript_biotype:protein_coding<span class="w"> </span>descrip
tion:chromosomal<span class="w"> </span>replication<span class="w"> </span>initiator<span class="w"> </span>protein<span class="w"> </span>DnaA
MSLSLWQQCLARLQDELPATEFSMWIRPLQAELSDNTLALYAPNRFVLDWVRDKYLNNIN
GLLTSFCGADAPQLRFEVGTKPVTQTPQAAVTSNVAAPAQVAQTQPQRAAPSTRSGWDNV
PAPAEPTYRSNVNVKHTFDNFVEGKSNQLARAAARQVADNPGGAYNPLFLYGGTGLGKTH
LLHAVGNGIMARKPNAKVVYMHSERFVQDMVKALQNNAIEEFKRYYRSVDALLIDDIQFF
ANKERSQEEFFHTFNALLEGNQQIILTSDRYPKEINGVEDRLKSRFGWGLTVAIEPPELE
TRVAILMKKADENDIRLPGEVAFFIAKRLRSNVRELEGALNRVIANANFTGRAITIDFVR
EALRDLLALQEKLVTIDNIQKTVAEYYKIKVADLLSKRRSRSVARPRQMAMALAKELTNH
SLPEIGDAFGGRDHTTVLHACRKIEQLREESHDIKEDFSNLIRTLSS
&gt;ACT27083<span class="w"> </span>pep<span class="w"> </span>supercontig:ASM2366v1:CP001665:1755:2855:1<span class="w"> </span>gene:ECBD_0002<span class="w"> </span>transcript:ACT27083<span class="w"> </span>gene_biotype:protein_coding<span class="w"> </span>transcript_biotype:protein_coding<span class="w"> </span>descri
ption:DNA<span class="w"> </span>polymerase<span class="w"> </span>III,<span class="w"> </span>beta<span class="w"> </span>subunit
MKFTVEREHLLKPLQQVSGPLGGRPTLPILGNLLLQVADGTLSLTGTDLEMEMVARVALV
QPHEPGATTVPARKFFDICRGLPEGAEIAVQLEGERMLVRSGRSRFSLSTLPAADFPNLD
DWQSEVEFTLPQATMKRLIEATQFSMAHQDVRYYLNGMLFETEGEELRTVATDGHRLAVC
SMPIGQSLPSHSVIVPRKGVIELMRMLDGGDNPLRVQIGSNNIRAHVGDFIFTSKLVDGR
FPDYRRVLPKNPDKHLEAGCDLLKQAFARAAILSNEKFRGVRLYVSENQLKITANNPEQE
EAEEILDVTYSGAEMEIGFNVSYVLDVLNALKCENVRMMLTDSVSSVQIEDAASQSAAYV
VMPMRL
&gt;ACT27084<span class="w"> </span>pep<span class="w"> </span>supercontig:ASM2366v1:CP001665:2855:3928:1<span class="w"> </span>gene:ECBD_0003<span class="w"> </span>transcript:ACT27084<span class="w"> </span>gene_biotype:protein_coding<span class="w"> </span>transcript_biotype:protein_coding<span class="w"> </span>descri
ption:DNA<span class="w"> </span>replication<span class="w"> </span>and<span class="w"> </span>repair<span class="w"> </span>protein<span class="w"> </span>RecF
MSLTRLLIRDFRNIETADLALSPGFNFLVGANGSGKTSVLEAIYTLGHGRAFRSLQIGRV
IRHEQEAFVLHGRLQGEERETAIGLTKDKQGDSKVRIDGTDGHKVAELAHLMPMQLITPE
GFTLLNGGPKYRRAFLDWGCFHNEPGFFTAWSNLKRLLKQRNAALRQVTRYEQLRPWDKE
--More--<span class="o">(</span><span class="m">0</span>%<span class="o">)</span>
</pre></div>
</div>
<p>We see that many protein sequences are embedded in the files and separated by their name, always preceded by the character <strong>“&gt;”</strong>.</p>
<p>To know how many sequences are in the files we can use the program <strong>grep</strong> with the option <strong>-c</strong> for counting the number of rows containg the character <strong>“&gt;”</strong>:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>grep<span class="w"> </span><span class="s2">&quot;&gt;&quot;</span><span class="w"> </span>-c<span class="w"> </span>Escherichia_coli_bl21_gold_de3_plyss_ag_.ASM2366v1.pep.all.fa
<span class="m">4228</span>
</pre></div>
</div>
<section id="id1">
<h4><strong>Recap</strong><a class="headerlink" href="#id1" title="Permalink to this heading"></a></h4>
<ul class="simple">
<li><p><strong>wget</strong> downloads file from a URL</p></li>
<li><p><strong>more</strong> prints a part of the content of a file in interactive way</p></li>
<li><p><strong>grep</strong> extract the rows containing a particular character / pattern.</p></li>
</ul>
<p>&lt;h3&gt;Next Session&lt;/h3&gt;</p>
<p>[Manipulate files, piping, parsing, reformatting](<a class="reference external" href="https://biocorecrg.github.io/advanced_linux_2019/parsing">https://biocorecrg.github.io/advanced_linux_2019/parsing</a>)</p>
</section>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="about.html" class="btn btn-neutral float-left" title="About the course" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="containers.html" class="btn btn-neutral float-right" title="Containers" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, CRG.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>