<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Introduction to Linux &mdash; ROPES summer school - CRG - May 2023  documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
    <link rel="canonical" href="https://biocorecrg.github.io/ropes-linux-mop2-2023/linux_basics.html" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="author" title="About these documents" href="about.html" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Containers" href="containers.html" />
    <link rel="prev" title="About the course" href="about.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            ROPES summer school - CRG - May 2023
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="about.html">About the course</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Introduction to Linux</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#what-is-linux">What is Linux?</a></li>
<li class="toctree-l2"><a class="reference internal" href="#what-is-a-command-line-interface-gui">What is a Command Line Interface (GUI)?</a></li>
<li class="toctree-l2"><a class="reference internal" href="#create-directory-and-practice-moving-around">Create directory and practice moving around</a></li>
<li class="toctree-l2"><a class="reference internal" href="#download-files-from-repositories">Download files from repositories</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#recap"><strong>Recap</strong></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#manipulate-files-piping-parsing-reformatting">Manipulate files, piping, parsing, reformatting</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id1"><strong>Recap</strong></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#sequence-file-formats-fasta-and-fastq">Sequence file formats: Fasta and fastq</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id2"><strong>Recap</strong></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#bed-format-and-regular-expressions">Bed format and regular expressions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#genome-browser">Genome browser</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id3">Recap</a></li>
<li class="toctree-l3"><a class="reference internal" href="#exercises">Exercises</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#space-in-volumes-and-permissions">Space in volumes and permissions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#volume-sizes-and-disk-space">Volume sizes and disk space</a></li>
<li class="toctree-l3"><a class="reference internal" href="#permissions">Permissions</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#use-octal-notation-for-file-permissions">Use octal notation for file permissions</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="containers.html">Containers</a></li>
<li class="toctree-l1"><a class="reference internal" href="nextflow.html">Nextflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="mop2_practical.html">Master of Pores 2</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">ROPES summer school - CRG - May 2023</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Introduction to Linux</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/linux_basics.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="introduction-to-linux">
<span id="linux-basics-page"></span><h1>Introduction to Linux<a class="headerlink" href="#introduction-to-linux" title="Permalink to this heading"></a></h1>
<ul class="simple">
<li><p><a class="reference internal" href="#what-is-linux">What is Linux?</a></p></li>
<li><p><a class="reference internal" href="#what-is-a-command-line-interface-gui">What is a Command Line Interface (GUI)?</a></p></li>
<li><p><a class="reference internal" href="#create-directory-and-practice-moving-around">Create directory and practice moving around</a></p></li>
<li><p><a class="reference internal" href="#download-files-from-repositories">Download files from repositories</a></p></li>
<li><p><a class="reference internal" href="#manipulate-files-piping-parsing-reformatting">Manipulate files, piping, parsing, reformatting</a></p></li>
<li><p><a class="reference internal" href="#sequence-file-formats-fasta-and-fastq">Sequence file formats: Fasta and fastq</a></p></li>
<li><p><a class="reference internal" href="#bed-format-and-regular-expressions">Bed format and regular expressions</a></p></li>
<li><p><a class="reference internal" href="#space-in-volumes-and-permissions">Space in volumes and permissions</a></p></li>
</ul>
<section id="what-is-linux">
<h2>What is Linux?<a class="headerlink" href="#what-is-linux" title="Permalink to this heading"></a></h2>
<a class="reference internal image-reference" href="_images/Linux-Logo.png"><img alt="_images/Linux-Logo.png" src="_images/Linux-Logo.png" style="width: 400px;" /></a>
<p>Linux is an open-source operating system that was developed in the 1990s by <a class="reference external" href="https://en.wikipedia.org/wiki/Linus_Torvalds">Linus Torvalds</a>. Unlike other popular operating systems like Windows and MacOS, Linux is free to download, use, and distribute.</p>
<p>One of the most distinctive features of Linux is that it is highly customizable. Users can choose from a wide variety of “distributions” (or “distros”) of Linux, each with its own unique set of pre-installed software and user interface. Some popular distros include <a class="reference external" href="https://ubuntu.com/">Ubuntu</a>, <a class="reference external" href="https://www.debian.org/">Debian</a>, and <a class="reference external" href="https://fedoraproject.org/">Fedora</a>.</p>
<p>Linux is known for being incredibly stable and secure. It is also highly efficient, which means it can run well even on older or less powerful hardware. This has made Linux a popular choice for servers and other industrial applications.</p>
<p>Because it is <strong>open-source</strong>, Linux has a large and active community of developers who are constantly working to improve the operating system. This means that there are always new updates and features being added to Linux, and users can be confident that their operating system is always up-to-date and secure.</p>
<p>Linux is a is widely used in <strong>bioinformatics</strong> due to its ability to handle large data sets and complex computational tasks. Bioinformaticians use Linux to analyze DNA sequences, study gene expression, and develop computational models of biological systems.</p>
<p>One of the most important features of Linux for bioinformatics is its support for powerful command-line tools. These tools allow bioinformaticians to perform complex data processing and analysis tasks with ease. In addition, Linux provides a rich array of programming languages and libraries that are essential for bioinformatics research, such as <strong>Python</strong>, <strong>Java</strong>, <strong>Perl</strong>, <strong>C</strong> and <strong>C++</strong>.</p>
<p>One of the key advantages of Linux for bioinformatics is its ability to run on a wide range of hardware, from powerful servers to small embedded devices. This makes it possible to run bioinformatics analyses on a variety of different systems, including <strong>high-performance computing clusters</strong>, <strong>cloud-based platforms</strong>, and <strong>personal computers</strong>.</p>
<p>Another advantage of Linux for bioinformatics is its strong focus on security and reliability. Because Linux is <strong>open source</strong>, the community is constantly working to improve the operating system and fix any security vulnerabilities that are discovered. This makes Linux a trusted platform for sensitive bioinformatics data.</p>
</section>
<section id="what-is-a-command-line-interface-gui">
<h2>What is a Command Line Interface (GUI)?<a class="headerlink" href="#what-is-a-command-line-interface-gui" title="Permalink to this heading"></a></h2>
<a class="reference internal image-reference" href="_images/cli.png"><img alt="_images/cli.png" src="_images/cli.png" style="width: 600px;" /></a>
<p>A command line interface (<strong>CLI</strong>) is a text-based interface used to interact with a computer’s operating system or software by entering commands through a command prompt.</p>
<p>The command prompt usually consists of a text area where the user can enter a command, and the output of the command is displayed in the same area.</p>
<p><strong>Commands</strong> can be entered using specific keywords or phrases, which are interpreted by the operating system or software.</p>
<p>For example, in the Windows command prompt, the user can type “<strong>dir</strong>” to list the files and directories in the current directory, and in the Unix/Linux command line, the user can type “<strong>ls</strong>” to achieve the same result.</p>
<p>In addition to simple commands, more complex operations can be performed by chaining commands together using special symbols.</p>
<p>Overall, command line interfaces offer a fast and powerful way to interact with a computer or software system, and are often preferred to graphical user interface (<strong>GUI</strong>) by experienced users or developers due to their flexibility and efficiency.</p>
</section>
<section id="create-directory-and-practice-moving-around">
<h2>Create directory and practice moving around<a class="headerlink" href="#create-directory-and-practice-moving-around" title="Permalink to this heading"></a></h2>
<p>To create file and folders in linux is quite simple. You can use a number of programs for creating an empty file (<strong>touch</strong>) or an empty directory (<strong>mkdir</strong>)</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>touch<span class="w"> </span>my_beautiful_file.txt

mkdir<span class="w"> </span>my_beautiful_folder
</pre></div>
</div>
<p>To display the list of files and folder we can use the command <strong>ls</strong></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>ls
my_beautiful_file.txt<span class="w">  </span>my_beautiful_folder
</pre></div>
</div>
<p>To change the name of a file (or a directory) you can use the command <strong>mv</strong> while for copying the file you can use <strong>cp</strong>. Adding the option <strong>-r</strong> (recursive) to <strong>cp</strong> allows to copy a whole folder and its content.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>mv<span class="w"> </span>my_beautiful_file.txt<span class="w"> </span>my_ugly_file.txt
mv<span class="w"> </span>my_beautiful_folder<span class="w"> </span>my_ugly_folder

cp<span class="w"> </span>my_ugly_file.txt<span class="w"> </span>my_beautiful_file.txt
cp<span class="w"> </span>my_ugly_folder<span class="w"> </span>-r<span class="w"> </span>my_beautiful_folder
</pre></div>
</div>
<p>If you omit the <strong>-r</strong> option the system will complain</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>cp<span class="w"> </span>my_ugly_folder<span class="w"> </span>my_other_folder
</pre></div>
</div>
<p>You can use <strong>mv</strong> also for moving a file (or a directory) inside a folder. Also <strong>cp</strong> will allow you to make a copy inside a folder.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>mv<span class="w"> </span>my_beautiful_file.txt<span class="w"> </span>my_beautiful_folder
cp<span class="w"> </span>my_ugly_file.txt<span class="w"> </span>my_ugly_folder

ls

my_beautiful_folder<span class="w">  </span>my_ugly_file.txt<span class="w">  </span>my_ugly_folder
</pre></div>
</div>
<p>For entering in a folder we can use the tool <strong>cd</strong></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">cd</span><span class="w"> </span>my_ugly_folder

ls

my_ugly_file.txt
</pre></div>
</div>
<p>For going out we can move one level out</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">cd</span><span class="w"> </span>../

ls

my_beautiful_folder<span class="w">  </span>my_ugly_file.txt<span class="w">  </span>my_ugly_folder
</pre></div>
</div>
<p>Sometimes we get lost and would like to know where we are.</p>
<a class="reference internal image-reference" href="_images/lost.png"><img alt="_images/lost.png" src="_images/lost.png" style="width: 600px;" /></a>
<p>We can use the command <strong>pwd</strong>.</p>
<p>We can write to a file using the character <strong>&gt;</strong>, that means output redirection.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">echo</span><span class="w"> </span><span class="s2">&quot;ATGTACTGACTGCATGCATGCCATGCA&quot;</span><span class="w"> </span>&gt;<span class="w"> </span>my_dna.txt
</pre></div>
</div>
<p>And display the content of the file using the program <strong>cat</strong></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>cat<span class="w"> </span>my_dna.txt

ATGTACTGACTGCATGCATGCCATGCA
</pre></div>
</div>
<p>To convert this sequence to a RNA one we can just replace the <strong>T</strong> base with <strong>U</strong> by using the program <strong>sed</strong>. The sintax of this program is the following <strong>s/&lt;TO BE REPLACED&gt;/&lt;TO REPLACE&gt;/</strong>.</p>
<p>You can add a <strong>g</strong> at the end if you want to replace every character found <strong>s/&lt;TO BE REPLACED&gt;/&lt;TO REPLACE&gt;/g</strong>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sed<span class="w"> </span>s/T/U/g<span class="w"> </span>my_dna.txt<span class="w"> </span>&gt;<span class="w"> </span>my_rna.txt

cat<span class="w"> </span>my_rna.txt

AUGUACUGACUGCAUGCAUGCCAUGCA
</pre></div>
</div>
<p>Every command has a manual, you can read it by using the program <strong>man</strong> with the name of the tool.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>man<span class="w"> </span>ls

LS<span class="o">(</span><span class="m">1</span><span class="o">)</span><span class="w">                                                                   </span>User<span class="w"> </span>Commands<span class="w">                                                                   </span>LS<span class="o">(</span><span class="m">1</span><span class="o">)</span>

NAME
<span class="w">      </span>ls<span class="w"> </span>-<span class="w"> </span>list<span class="w"> </span>directory<span class="w"> </span>contents

SYNOPSIS
<span class="w">       </span>ls<span class="w"> </span><span class="o">[</span>OPTION<span class="o">]</span>...<span class="w"> </span><span class="o">[</span>FILE<span class="o">]</span>...

DESCRIPTION
<span class="w">       </span>List<span class="w"> </span>information<span class="w"> </span>about<span class="w"> </span>the<span class="w"> </span>FILEs<span class="w"> </span><span class="o">(</span>the<span class="w"> </span>current<span class="w"> </span>directory<span class="w"> </span>by<span class="w"> </span>default<span class="o">)</span>.<span class="w">  </span>Sort<span class="w"> </span>entries<span class="w"> </span>alphabetically<span class="w"> </span><span class="k">if</span><span class="w"> </span>none<span class="w"> </span>of<span class="w"> </span>-cftuvSUX<span class="w"> </span>nor<span class="w"> </span>--sort<span class="w"> </span>is<span class="w"> </span>specified.

<span class="w">       </span>Mandatory<span class="w"> </span>arguments<span class="w"> </span>to<span class="w"> </span>long<span class="w"> </span>options<span class="w"> </span>are<span class="w"> </span>mandatory<span class="w"> </span><span class="k">for</span><span class="w"> </span>short<span class="w"> </span>options<span class="w"> </span>too.

<span class="w">       </span>-a,<span class="w"> </span>--all
<span class="w">              </span><span class="k">do</span><span class="w"> </span>not<span class="w"> </span>ignore<span class="w"> </span>entries<span class="w"> </span>starting<span class="w"> </span>with<span class="w"> </span>.

<span class="w">      </span>-A,<span class="w"> </span>--almost-all
<span class="w">              </span><span class="k">do</span><span class="w"> </span>not<span class="w"> </span>list<span class="w"> </span>implied<span class="w"> </span>.<span class="w"> </span>and<span class="w"> </span>..

<span class="w">      </span>--author
<span class="w">            </span>with<span class="w"> </span>-l,<span class="w"> </span>print<span class="w"> </span>the<span class="w"> </span>author<span class="w"> </span>of<span class="w"> </span>each<span class="w"> </span>file

<span class="w">     </span>-b,<span class="w"> </span>--escape
<span class="w">            </span>print<span class="w"> </span>C-style<span class="w"> </span>escapes<span class="w"> </span><span class="k">for</span><span class="w"> </span>nongraphic<span class="w"> </span>characters
Manual<span class="w"> </span>page<span class="w"> </span>ls<span class="o">(</span><span class="m">1</span><span class="o">)</span><span class="w"> </span>line<span class="w"> </span><span class="m">1</span><span class="w"> </span><span class="o">(</span>press<span class="w"> </span>h<span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="nb">help</span><span class="w"> </span>or<span class="w"> </span>q<span class="w"> </span>to<span class="w"> </span>quit<span class="o">)</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><strong>RECAP</strong>
* <strong>touch</strong> writes empty files <strong>mkdir</strong> empty directories
* <strong>mv</strong> move files (or directory) or change their name
* <strong>ls</strong> list files and directories
* <strong>cp</strong> copy files and direcotries
* <strong>cd</strong> change the directory
* <strong>echo</strong> print values to standard output
* <strong>cat</strong> print the content of a file to standard output
* <strong>sed</strong> replace a string with another
* <strong>man</strong> print the manual for a function</p>
</div>
</section>
<section id="download-files-from-repositories">
<h2>Download files from repositories<a class="headerlink" href="#download-files-from-repositories" title="Permalink to this heading"></a></h2>
<p>Several institutions host different kind of genomics data.</p>
<p>For example the genome browser <a class="reference external" href="https://www.ensembl.org/index.html">Ensembl</a> is also a public repository of genomes and annotation that can be freely downloaded and used for any kind of analysis</p>
<p>The resource <a class="reference external" href="https://bacteria.ensembl.org/index.html">Ensembl Bacteria</a> contains a large number of bacterial genomes and their annotation. As an example we can browse the page corresponding to <a class="reference external" href="https://bacteria.ensembl.org/Escherichia_coli_bl21_gold_de3_plyss_ag_/Info/Index/">Escherichia coli ‘BL21-Gold(DE3)pLysS AG’</a></p>
<a class="reference internal image-reference" href="_images/ensembl_escherichia.png"><img alt="_images/ensembl_escherichia.png" src="_images/ensembl_escherichia.png" style="width: 800px;" /></a>
<p>We can click on “Download genes, cDNAs, ncRNA, proteins <strong>FASTA</strong>”</p>
<a class="reference internal image-reference" href="_images/list_ensembl_escherichia.png"><img alt="_images/list_ensembl_escherichia.png" src="_images/list_ensembl_escherichia.png" style="width: 800px;" /></a>
<p>And then on <strong>DNA</strong></p>
<a class="reference internal image-reference" href="_images/file_list_escherichia.png"><img alt="_images/file_list_escherichia.png" src="_images/file_list_escherichia.png" style="width: 800px;" /></a>
<p>Then as an example we can use the copy the link address of the <strong>README</strong> file using the mouse right button.</p>
<a class="reference internal image-reference" href="_images/right_click.png"><img alt="_images/right_click.png" src="_images/right_click.png" style="width: 800px;" /></a>
<p>Then we can go back to our command line and use the program <strong>wget</strong> to download that file and using <strong>CTRL+C</strong> to paste the address:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>wget<span class="w"> </span>ftp://ftp.ensemblgenomes.org/pub/bacteria/release-42/fasta/bacteria_22_collection/escherichia_coli_bl21_gold_de3_plyss_ag_/dna/README

--2019-03-06<span class="w"> </span><span class="m">18</span>:59:13--<span class="w">  </span>ftp://ftp.ensemblgenomes.org/pub/bacteria/release-42/fasta/bacteria_22_collection/escherichia_coli_bl21_gold_de3_plyss_ag_/dna/README
<span class="w">           </span><span class="o">=</span>&gt;<span class="w"> </span>‘README’
Resolving<span class="w"> </span>ftp.ensemblgenomes.org<span class="w"> </span><span class="o">(</span>ftp.ensemblgenomes.org<span class="o">)</span>...<span class="w"> </span><span class="m">193</span>.62.197.94
Connecting<span class="w"> </span>to<span class="w"> </span>ftp.ensemblgenomes.org<span class="w"> </span><span class="o">(</span>ftp.ensemblgenomes.org<span class="o">)</span><span class="p">|</span><span class="m">193</span>.62.197.94<span class="p">|</span>:21...<span class="w"> </span>connected.
Logging<span class="w"> </span><span class="k">in</span><span class="w"> </span>as<span class="w"> </span>anonymous<span class="w"> </span>...<span class="w"> </span>Logged<span class="w"> </span><span class="k">in</span>!
<span class="o">==</span>&gt;<span class="w"> </span>SYST<span class="w"> </span>...<span class="w"> </span><span class="k">done</span>.<span class="w">    </span><span class="o">==</span>&gt;<span class="w"> </span>PWD<span class="w"> </span>...<span class="w"> </span><span class="k">done</span>.
<span class="o">==</span>&gt;<span class="w"> </span>TYPE<span class="w"> </span>I<span class="w"> </span>...<span class="w"> </span><span class="k">done</span>.<span class="w">  </span><span class="o">==</span>&gt;<span class="w"> </span>CWD<span class="w"> </span><span class="o">(</span><span class="m">1</span><span class="o">)</span><span class="w"> </span>/pub/bacteria/release-42/fasta/bacteria_22_collection/escherichia_coli_bl21_gold_de3_plyss_ag_/dna<span class="w"> </span>...<span class="w"> </span><span class="k">done</span>.
<span class="o">==</span>&gt;<span class="w"> </span>SIZE<span class="w"> </span>README<span class="w"> </span>...<span class="w"> </span><span class="nv">4923</span>
<span class="o">==</span>&gt;<span class="w"> </span>PASV<span class="w"> </span>...<span class="w"> </span><span class="k">done</span>.<span class="w">    </span><span class="o">==</span>&gt;<span class="w"> </span>RETR<span class="w"> </span>README<span class="w"> </span>...<span class="w"> </span><span class="k">done</span>.
Length:<span class="w"> </span><span class="m">4923</span><span class="w"> </span><span class="o">(</span><span class="m">4</span>.8K<span class="o">)</span><span class="w"> </span><span class="o">(</span>unauthoritative<span class="o">)</span>

<span class="m">100</span>%<span class="o">[======================================================================================================================</span>&gt;<span class="o">]</span><span class="w"> </span><span class="m">4</span>,923<span class="w">       </span>--.-K/s<span class="w">   </span><span class="k">in</span><span class="w"> </span>0s

<span class="m">2019</span>-03-06<span class="w"> </span><span class="m">18</span>:59:14<span class="w"> </span><span class="o">(</span><span class="m">295</span><span class="w"> </span>MB/s<span class="o">)</span><span class="w"> </span>-<span class="w"> </span>‘README’<span class="w"> </span>saved<span class="w"> </span><span class="o">[</span><span class="m">4923</span><span class="o">]</span>
</pre></div>
</div>
<p>we can then use the program <strong>more</strong> to display part of the content of the file:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>more<span class="w"> </span>README


<span class="c1">#### README ####</span>

IMPORTANT:<span class="w"> </span>Please<span class="w"> </span>note<span class="w"> </span>you<span class="w"> </span>can<span class="w"> </span>download<span class="w"> </span>correlation<span class="w"> </span>data<span class="w"> </span>tables,
supported<span class="w"> </span>by<span class="w"> </span>Ensembl,<span class="w"> </span>via<span class="w"> </span>the<span class="w"> </span>highly<span class="w"> </span>customisable<span class="w"> </span>BioMart<span class="w"> </span>and
EnsMart<span class="w"> </span>data<span class="w"> </span>mining<span class="w"> </span>tools.<span class="w"> </span>See<span class="w"> </span>http://www.ensembl.org/biomart/martview<span class="w"> </span>or
http://www.ebi.ac.uk/biomart/<span class="w"> </span><span class="k">for</span><span class="w"> </span>more<span class="w"> </span>information.

The<span class="w"> </span>genome<span class="w"> </span>assembly<span class="w"> </span>represented<span class="w"> </span>here<span class="w"> </span>corresponds<span class="w"> </span>to
GCA_000023665.1

<span class="c1">#######################</span>
Fasta<span class="w"> </span>DNA<span class="w"> </span>dumps
<span class="c1">#######################</span>

-----------
FILE<span class="w"> </span>NAMES
------------
The<span class="w"> </span>files<span class="w"> </span>are<span class="w"> </span>consistently<span class="w"> </span>named<span class="w"> </span>following<span class="w"> </span>this<span class="w"> </span>pattern:
<span class="w">   </span>&lt;species&gt;.&lt;assembly&gt;.&lt;sequence<span class="w"> </span>type&gt;.&lt;id<span class="w"> </span>type&gt;.&lt;id&gt;.fa.gz

&lt;species&gt;:<span class="w">   </span>The<span class="w"> </span>systematic<span class="w"> </span>name<span class="w"> </span>of<span class="w"> </span>the<span class="w"> </span>species.
&lt;assembly&gt;:<span class="w">  </span>The<span class="w"> </span>assembly<span class="w"> </span>build<span class="w"> </span>name.
&lt;sequence<span class="w"> </span>type&gt;:
<span class="w"> </span>*<span class="w"> </span><span class="s1">&#39;dna&#39;</span><span class="w"> </span>-<span class="w"> </span>unmasked<span class="w"> </span>genomic<span class="w"> </span>DNA<span class="w"> </span>sequences.
--More--<span class="o">(</span><span class="m">14</span>%<span class="o">)</span>
</pre></div>
</div>
<p>Pressing the bar allows us to scroll down the file, while for exiting you just click <strong>CTRL+C</strong>.
After reading the README we can download the file named <strong>toplevel</strong> that contains chromosomes, regions not assembled into chromosomes and N padded haplotype/patch regions:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>wget<span class="w"> </span>ftp://ftp.ensemblgenomes.org/pub/bacteria/release-42/fasta/bacteria_22_collection/escherichia_coli_bl21_gold_de3_plyss_ag_/dna/Escherichia_coli_bl21_gold_de3_plyss_ag_.ASM2366v1.dna.toplevel.fa.gz
</pre></div>
</div>
<p>We can use the options <strong>-lh</strong> of the program <strong>ls</strong> to list attributes of the files and show in human readable format the size fo the files</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>ls<span class="w"> </span>-lh

total<span class="w"> </span><span class="m">2</span>.0M
drwxr-xr-x<span class="w"> </span><span class="m">5</span><span class="w"> </span>lcozzuto<span class="w"> </span>Bioinformatics_Unit<span class="w">  </span><span class="m">209</span><span class="w"> </span>Mar<span class="w">  </span><span class="m">7</span><span class="w"> </span><span class="m">11</span>:48<span class="w"> </span>advanced_linux_2019
-rw-r--r--<span class="w"> </span><span class="m">1</span><span class="w"> </span>lcozzuto<span class="w"> </span>Bioinformatics_Unit<span class="w"> </span><span class="m">1</span>.4M<span class="w"> </span>Mar<span class="w">  </span><span class="m">7</span><span class="w"> </span><span class="m">13</span>:06<span class="w"> </span>Escherichia_coli_bl21_gold_de3_plyss_ag_.ASM2366v1.dna.toplevel.fa.gz
drwxr-xr-x<span class="w"> </span><span class="m">2</span><span class="w"> </span>lcozzuto<span class="w"> </span>Bioinformatics_Unit<span class="w">   </span><span class="m">39</span><span class="w"> </span>Mar<span class="w">  </span><span class="m">6</span><span class="w"> </span><span class="m">18</span>:17<span class="w"> </span>my_beautiful_folder
-rw-r--r--<span class="w"> </span><span class="m">1</span><span class="w"> </span>lcozzuto<span class="w"> </span>Bioinformatics_Unit<span class="w">    </span><span class="m">0</span><span class="w"> </span>Mar<span class="w">  </span><span class="m">6</span><span class="w"> </span><span class="m">18</span>:15<span class="w"> </span>my_ugly_file.txt
drwxr-xr-x<span class="w"> </span><span class="m">2</span><span class="w"> </span>lcozzuto<span class="w"> </span>Bioinformatics_Unit<span class="w">   </span><span class="m">34</span><span class="w"> </span>Mar<span class="w">  </span><span class="m">6</span><span class="w"> </span><span class="m">18</span>:17<span class="w"> </span>my_ugly_folder
-rw-r--r--<span class="w"> </span><span class="m">1</span><span class="w"> </span>lcozzuto<span class="w"> </span>Bioinformatics_Unit<span class="w"> </span><span class="m">4</span>.9K<span class="w"> </span>Mar<span class="w">  </span><span class="m">6</span><span class="w"> </span><span class="m">18</span>:59<span class="w"> </span>README
</pre></div>
</div>
<p>For unzipping the file we can use the program <strong>gunzip</strong>. The uncompressed file is now <strong>4.5M</strong>.</p>
<p>Let’s see the content of the file.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>more<span class="w"> </span>Escherichia_coli_bl21_gold_de3_plyss_ag_.ASM2366v1.dna.toplevel.fa

&gt;CP001665<span class="w"> </span>dna:supercontig<span class="w"> </span>supercontig:ASM2366v1:CP001665:1:4570938:1<span class="w"> </span>REF
CGTCCTGGATCTTTATTAGATCGATTAAGCCAATTTTTGTCTATGGTCATTAAATTTTCC
AATATGCGGCGTAAATCGTGCCCGCCTCGCGGCAGGATCGTTTACACTTAGCGAGTTCTG
GAAAGTCCTGTGGATAAATCGGGAAAATCTGTGAGAAACAGAAGATCTCTTGCGCAGTTT
AGGCTATGATCCGCGGTCCCGATCGTTTTGCAGGATCTTGATCGGGCATATAACCGCAGA
CAGCGGTTCGTGCGTCACCCTCAAGCAGGGTCTTTTCGACGTACGTCAACAATCATGAAT
GTTTCAGCCTTAGTCATTATCGACTTTTGTTCGAGTGGAGTCCGCCGTGTCACTTTCGCT
TTGGCAGCAGTGTCTTGCCCGATTGCAGGATGAGTTACCAGCCACAGAATTCAGTATGTG
GATACGCCCATTGCAGGCGGAACTGAGCGATAACACGCTGGCCCTGTACGCGCCAAACCG
TTTTGTCCTCGATTGGGTACGGGACAAGTACCTTAATAATATCAATGGACTGCTAACCAG
TTTCTGCGGAGCGGATGCCCCACAGCTGCGTTTTGAAGTCGGCACCAAACCGGTGACGCA
AACGCCACAAGCGGCAGTGACGAGCAACGTCGCGGCCCCTGCACAGGTGGCGCAAACGCA
GCCGCAACGTGCTGCGCCTTCTACGCGCTCAGGTTGGGATAACGTCCCGGCCCCGGCAGA
ACCGACCTATCGTTCTAACGTAAACGTCAAACACACGTTTGATAACTTCGTTGAAGGTAA
ATCTAACCAACTGGCGCGCGCGGCGGCTCGCCAGGTGGCGGATAACCCTGGCGGTGCCTA
TAACCCGTTGTTCCTTTATGGCGGCACGGGTCTGGGTAAAACTCACCTGCTGCATGCGGT
GGGTAACGGCATTATGGCGCGCAAGCCGAATGCCAAAGTGGTTTATATGCACTCCGAGCG
CTTTGTTCAGGACATGGTTAAAGCCCTGCAAAACAACGCGATCGAAGAGTTTAAACGCTA
CTACCGTTCCGTAGATGCACTGCTGATCGACGATATTCAGTTTTTTGCTAATAAAGAACG
ATCTCAGGAAGAGTTTTTCCACACCTTCAACGCCCTGCTGGAAGGTAATCAACAGATCAT
TCTCACCTCGGATCGCTATCCGAAAGAGATCAACGGCGTTGAGGATCGTTTGAAATCCCG
CTTCGGTTGGGGACTGACTGTGGCGATCGAACCGCCAGAGCTGGAAACCCGTGTGGCGAT
CCTGATGAAAAAGGCCGACGAAAACGACATTCGTTTGCCGGGCGAAGTGGCGTTCTTTAT
CGCCAAGCGTCTACGATCTAACGTACGTGAGCTGGAAGGGGCGCTGAACCGCGTCATTGC
</pre></div>
</div>
<p>The file contains the whole genome of the bacteria.</p>
<p>The first line contains the character <strong>&gt;</strong> and the name of the molecule / genome.</p>
<p>This format is called <a class="reference external" href="https://en.wikipedia.org/wiki/FASTA_format">FASTA</a> format and is universally used for storing one or multiple DNA/RNA/Protein sequences.</p>
<p>We can now download in the same ways the proteins:</p>
<a class="reference internal image-reference" href="_images/righ_click_2.png"><img alt="_images/righ_click_2.png" src="_images/righ_click_2.png" style="width: 800px;" /></a>
<p>and after unzipping the file we can have a look at it.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>more<span class="w"> </span>Escherichia_coli_bl21_gold_de3_plyss_ag_.ASM2366v1.pep.all.fa

&gt;ACT27082<span class="w"> </span>pep<span class="w"> </span>supercontig:ASM2366v1:CP001665:347:1750:1<span class="w"> </span>gene:ECBD_0001<span class="w"> </span>transcript:ACT27082<span class="w"> </span>gene_biotype:protein_coding<span class="w"> </span>transcript_biotype:protein_coding<span class="w"> </span>descrip
tion:chromosomal<span class="w"> </span>replication<span class="w"> </span>initiator<span class="w"> </span>protein<span class="w"> </span>DnaA
MSLSLWQQCLARLQDELPATEFSMWIRPLQAELSDNTLALYAPNRFVLDWVRDKYLNNIN
GLLTSFCGADAPQLRFEVGTKPVTQTPQAAVTSNVAAPAQVAQTQPQRAAPSTRSGWDNV
PAPAEPTYRSNVNVKHTFDNFVEGKSNQLARAAARQVADNPGGAYNPLFLYGGTGLGKTH
LLHAVGNGIMARKPNAKVVYMHSERFVQDMVKALQNNAIEEFKRYYRSVDALLIDDIQFF
ANKERSQEEFFHTFNALLEGNQQIILTSDRYPKEINGVEDRLKSRFGWGLTVAIEPPELE
TRVAILMKKADENDIRLPGEVAFFIAKRLRSNVRELEGALNRVIANANFTGRAITIDFVR
EALRDLLALQEKLVTIDNIQKTVAEYYKIKVADLLSKRRSRSVARPRQMAMALAKELTNH
SLPEIGDAFGGRDHTTVLHACRKIEQLREESHDIKEDFSNLIRTLSS
&gt;ACT27083<span class="w"> </span>pep<span class="w"> </span>supercontig:ASM2366v1:CP001665:1755:2855:1<span class="w"> </span>gene:ECBD_0002<span class="w"> </span>transcript:ACT27083<span class="w"> </span>gene_biotype:protein_coding<span class="w"> </span>transcript_biotype:protein_coding<span class="w"> </span>descri
ption:DNA<span class="w"> </span>polymerase<span class="w"> </span>III,<span class="w"> </span>beta<span class="w"> </span>subunit
MKFTVEREHLLKPLQQVSGPLGGRPTLPILGNLLLQVADGTLSLTGTDLEMEMVARVALV
QPHEPGATTVPARKFFDICRGLPEGAEIAVQLEGERMLVRSGRSRFSLSTLPAADFPNLD
DWQSEVEFTLPQATMKRLIEATQFSMAHQDVRYYLNGMLFETEGEELRTVATDGHRLAVC
SMPIGQSLPSHSVIVPRKGVIELMRMLDGGDNPLRVQIGSNNIRAHVGDFIFTSKLVDGR
FPDYRRVLPKNPDKHLEAGCDLLKQAFARAAILSNEKFRGVRLYVSENQLKITANNPEQE
EAEEILDVTYSGAEMEIGFNVSYVLDVLNALKCENVRMMLTDSVSSVQIEDAASQSAAYV
VMPMRL
&gt;ACT27084<span class="w"> </span>pep<span class="w"> </span>supercontig:ASM2366v1:CP001665:2855:3928:1<span class="w"> </span>gene:ECBD_0003<span class="w"> </span>transcript:ACT27084<span class="w"> </span>gene_biotype:protein_coding<span class="w"> </span>transcript_biotype:protein_coding<span class="w"> </span>descri
ption:DNA<span class="w"> </span>replication<span class="w"> </span>and<span class="w"> </span>repair<span class="w"> </span>protein<span class="w"> </span>RecF
MSLTRLLIRDFRNIETADLALSPGFNFLVGANGSGKTSVLEAIYTLGHGRAFRSLQIGRV
IRHEQEAFVLHGRLQGEERETAIGLTKDKQGDSKVRIDGTDGHKVAELAHLMPMQLITPE
GFTLLNGGPKYRRAFLDWGCFHNEPGFFTAWSNLKRLLKQRNAALRQVTRYEQLRPWDKE
--More--<span class="o">(</span><span class="m">0</span>%<span class="o">)</span>
</pre></div>
</div>
<p>We see that many protein sequences are embedded in the files and separated by their name, always preceded by the character <strong>“&gt;”</strong>.</p>
<p>To know how many sequences are in the files we can use the program <strong>grep</strong> with the option <strong>-c</strong> for counting the number of rows containg the character <strong>“&gt;”</strong>:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>grep<span class="w"> </span><span class="s2">&quot;&gt;&quot;</span><span class="w"> </span>-c<span class="w"> </span>Escherichia_coli_bl21_gold_de3_plyss_ag_.ASM2366v1.pep.all.fa
<span class="m">4228</span>
</pre></div>
</div>
<section id="recap">
<h3><strong>Recap</strong><a class="headerlink" href="#recap" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p><strong>wget</strong> downloads file from a URL</p></li>
<li><p><strong>more</strong> prints a part of the content of a file in interactive way</p></li>
<li><p><strong>grep</strong> extract the rows containing a particular character / pattern.</p></li>
</ul>
</section>
</section>
<section id="manipulate-files-piping-parsing-reformatting">
<h2>Manipulate files, piping, parsing, reformatting<a class="headerlink" href="#manipulate-files-piping-parsing-reformatting" title="Permalink to this heading"></a></h2>
<p>Parsing a file means extracting meaningful parts from a data source.
In few words if you have table and are interested only in a number of columns, extracting those columns can be an example of <strong>parsing</strong>.
In our case, for example, we can extract the name of our sequences by using again the program <strong>grep</strong> and redirecting the output to a new file.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>grep<span class="w"> </span><span class="s2">&quot;&gt;&quot;</span><span class="w"> </span>Escherichia_coli_bl21_gold_de3_plyss_ag_.ASM2366v1.pep.all.fa<span class="w"> </span>&gt;<span class="w"> </span>seq_names.txt

more<span class="w"> </span>seq_names.txt

&gt;ACT27082<span class="w"> </span>pep<span class="w"> </span>supercontig:ASM2366v1:CP001665:347:1750:1<span class="w"> </span>gene:ECBD_0001<span class="w"> </span>transcript:ACT27082<span class="w"> </span>gene_biotype:protein_coding<span class="w"> </span>transcript_biotype:protein_coding<span class="w"> </span>descrip
tion:chromosomal<span class="w"> </span>replication<span class="w"> </span>initiator<span class="w"> </span>protein<span class="w"> </span>DnaA
&gt;ACT27083<span class="w"> </span>pep<span class="w"> </span>supercontig:ASM2366v1:CP001665:1755:2855:1<span class="w"> </span>gene:ECBD_0002<span class="w"> </span>transcript:ACT27083<span class="w"> </span>gene_biotype:protein_coding<span class="w"> </span>transcript_biotype:protein_coding<span class="w"> </span>descri
ption:DNA<span class="w"> </span>polymerase<span class="w"> </span>III,<span class="w"> </span>beta<span class="w"> </span>subunit
&gt;ACT27084<span class="w"> </span>pep<span class="w"> </span>supercontig:ASM2366v1:CP001665:2855:3928:1<span class="w"> </span>gene:ECBD_0003<span class="w"> </span>transcript:ACT27084<span class="w"> </span>gene_biotype:protein_coding<span class="w"> </span>transcript_biotype:protein_coding<span class="w"> </span>descri
ption:DNA<span class="w"> </span>replication<span class="w"> </span>and<span class="w"> </span>repair<span class="w"> </span>protein<span class="w"> </span>RecF
&gt;ACT27085<span class="w"> </span>pep<span class="w"> </span>supercontig:ASM2366v1:CP001665:3957:6371:1<span class="w"> </span>gene:ECBD_0004<span class="w"> </span>transcript:ACT27085<span class="w"> </span>gene_biotype:protein_coding<span class="w"> </span>transcript_biotype:protein_coding<span class="w"> </span>descri
ption:DNA<span class="w"> </span>gyrase,<span class="w"> </span>B<span class="w"> </span>subunit
</pre></div>
</div>
<p>We can also <strong>pipe</strong> the results of a program (via Standard output) to a new program (via Standard input) by using the character
<code class="docutils literal notranslate"><span class="pre">`|`</span></code>, the program <strong>head</strong> allows to extract the first N rows (indicated by the parameter <strong>-n</strong>). Tail, instead allows to get the latest N rows.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>grep<span class="w"> </span><span class="s2">&quot;&gt;&quot;</span><span class="w"> </span>-c<span class="w"> </span>Escherichia_coli_bl21_gold_de3_plyss_ag_.ASM2366v1.pep.all.fa
<span class="m">4228</span>

grep<span class="w"> </span><span class="s2">&quot;&gt;&quot;</span><span class="w"> </span>Escherichia_coli_bl21_gold_de3_plyss_ag_.ASM2366v1.pep.all.fa<span class="w"> </span><span class="p">|</span><span class="w"> </span>head<span class="w"> </span>-n<span class="w"> </span><span class="m">3</span>
&gt;ACT27082<span class="w"> </span>pep<span class="w"> </span>supercontig:ASM2366v1:CP001665:347:1750:1<span class="w"> </span>gene:ECBD_0001<span class="w"> </span>transcript:ACT27082<span class="w"> </span>gene_biotype:protein_coding<span class="w"> </span>transcript_biotype:protein_coding<span class="w"> </span>description:chromosomal<span class="w"> </span>replication<span class="w"> </span>initiator<span class="w"> </span>protein<span class="w"> </span>DnaA
&gt;ACT27083<span class="w"> </span>pep<span class="w"> </span>supercontig:ASM2366v1:CP001665:1755:2855:1<span class="w"> </span>gene:ECBD_0002<span class="w"> </span>transcript:ACT27083<span class="w"> </span>gene_biotype:protein_coding<span class="w"> </span>transcript_biotype:protein_coding<span class="w"> </span>description:DNA<span class="w"> </span>polymerase<span class="w"> </span>III,<span class="w"> </span>beta<span class="w"> </span>subunit
&gt;ACT27084<span class="w"> </span>pep<span class="w"> </span>supercontig:ASM2366v1:CP001665:2855:3928:1<span class="w"> </span>gene:ECBD_0003<span class="w"> </span>transcript:ACT27084<span class="w"> </span>gene_biotype:protein_coding<span class="w"> </span>transcript_biotype:protein_coding<span class="w"> </span>description:DNA<span class="w"> </span>replication<span class="w"> </span>and<span class="w"> </span>repair<span class="w"> </span>protein<span class="w"> </span>RecF

grep<span class="w"> </span><span class="s2">&quot;&gt;&quot;</span><span class="w"> </span>Escherichia_coli_bl21_gold_de3_plyss_ag_.ASM2366v1.pep.all.fa<span class="w"> </span><span class="p">|</span><span class="w"> </span>tail<span class="w"> </span>-n<span class="w"> </span><span class="m">3</span>
&gt;ACT31307<span class="w"> </span>pep<span class="w"> </span>supercontig:ASM2366v1:CP001665:4569941:4570198:-1<span class="w"> </span>gene:ECBD_4328<span class="w"> </span>transcript:ACT31307<span class="w"> </span>gene_biotype:protein_coding<span class="w"> </span>transcript_biotype:protein_coding<span class="w"> </span>description:protein<span class="w"> </span>of<span class="w"> </span>unknown<span class="w"> </span><span class="k">function</span><span class="w"> </span>DUF37
&gt;ACT31308<span class="w"> </span>pep<span class="w"> </span>supercontig:ASM2366v1:CP001665:4570162:4570488:-1<span class="w"> </span>gene:ECBD_4329<span class="w"> </span>transcript:ACT31308<span class="w"> </span>gene_biotype:protein_coding<span class="w"> </span>transcript_biotype:protein_coding<span class="w"> </span>description:ribonuclease<span class="w"> </span>P<span class="w"> </span>protein<span class="w"> </span>component
&gt;ACT31309<span class="w"> </span>pep<span class="w"> </span>supercontig:ASM2366v1:CP001665:4570538:4570678:-1<span class="w"> </span>gene:ECBD_4330<span class="w"> </span>transcript:ACT31309<span class="w"> </span>gene_biotype:protein_coding<span class="w"> </span>transcript_biotype:protein_coding<span class="w"> </span>description:ribosomal<span class="w"> </span>protein<span class="w"> </span>L34
</pre></div>
</div>
<p>Going back to the genome file, we can use a combination of <strong>grep</strong> and <strong>wc</strong> to count the number of bases.
The option <strong>-v</strong> of <strong>grep</strong> will remove the row with the indicated character.
The option <strong>-m</strong> of <strong>wc</strong> tool allows to count only the characters, while <strong>-l</strong> gives you the number of lines.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>grep<span class="w"> </span>-v<span class="w"> </span><span class="s2">&quot;&gt;&quot;</span><span class="w"> </span>Escherichia_coli_bl21_gold_de3_plyss_ag_.ASM2366v1.dna.toplevel.fa<span class="p">|</span><span class="w"> </span>wc<span class="w"> </span>-m
<span class="m">4647121</span>

grep<span class="w"> </span>-v<span class="w"> </span><span class="s2">&quot;&gt;&quot;</span><span class="w"> </span>Escherichia_coli_bl21_gold_de3_plyss_ag_.ASM2366v1.dna.toplevel.fa<span class="p">|</span><span class="w"> </span>wc<span class="w"> </span>-l
<span class="m">76183</span>
</pre></div>
</div>
<p>Now let’s try to extract only the identifiers from the protein file. As we can see they are located just before a
<strong>space</strong>. So we can slice the first column using the space as delimiter using the program <strong>cut</strong> and the option <strong>-d ” “</strong>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>cut<span class="w"> </span>-f<span class="w"> </span><span class="m">1</span><span class="w"> </span>-d<span class="w"> </span><span class="s2">&quot; &quot;</span><span class="w"> </span>seq_names.txt<span class="w"> </span><span class="p">|</span>head<span class="w"> </span>-n<span class="w"> </span><span class="m">5</span>
&gt;ACT27082
&gt;ACT27083
&gt;ACT27084
&gt;ACT27085
&gt;ACT27086
</pre></div>
</div>
<p>We still have the character <strong>&gt;</strong> from the fasta file. For removing it we can use the program <strong>tr</strong> with the option <strong>-d</strong> (delete).</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>cut<span class="w"> </span>-f<span class="w"> </span><span class="m">1</span><span class="w"> </span>-d<span class="w"> </span><span class="s2">&quot; &quot;</span><span class="w"> </span>seq_names.txt<span class="w"> </span><span class="p">|</span><span class="w"> </span>tr<span class="w"> </span>-d<span class="w"> </span><span class="s2">&quot;&gt;&quot;</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>head<span class="w"> </span>-n<span class="w"> </span><span class="m">5</span>
ACT27082
ACT27083
ACT27084
ACT27085
ACT27086
</pre></div>
</div>
<p>Sometimes it can be useful to have a random list of identifiers (for instance to have a random background). We can achieve this with the program <strong>shuf</strong>. The program <strong>cat</strong> shows the full content of a file.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>cut<span class="w"> </span>-f<span class="w"> </span><span class="m">1</span><span class="w"> </span>-d<span class="w"> </span><span class="s2">&quot; &quot;</span><span class="w"> </span>seq_names.txt<span class="w"> </span><span class="p">|</span><span class="w"> </span>tr<span class="w"> </span>-d<span class="w"> </span><span class="s2">&quot;&gt;&quot;</span><span class="w"> </span><span class="p">|</span>shuf<span class="w"> </span><span class="p">|</span><span class="w"> </span>head<span class="w"> </span>-n<span class="w"> </span><span class="m">5</span><span class="w"> </span>&gt;<span class="w"> </span>random.list

cat<span class="w"> </span>random.list
ACT31118
ACT27123
ACT31080
ACT28234
ACT29418
</pre></div>
</div>
<p>PS: the list is random, so it is unlikely you will get the same result.</p>
<p>A list of identifiers can be quite useful to go back to the original name list to extract the whole information.
We can do this using again the program <strong>grep</strong> with the options <strong>-F</strong> (it means search a fixed string, do not interpret it… we will explain this later) and <strong>-f</strong> for using patterns specified in a file.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>grep<span class="w"> </span>-Ff<span class="w"> </span>random.list<span class="w"> </span>seq_names.txt

&gt;ACT27123<span class="w"> </span>pep<span class="w"> </span>supercontig:ASM2366v1:CP001665:44295:44414:-1<span class="w"> </span>gene:ECBD_0043<span class="w"> </span>transcript:ACT27123<span class="w"> </span>gene_biotype:protein_coding<span class="w"> </span>transcript_biotype:protein_coding<span class="w"> </span>description:hypothetical<span class="w"> </span>protein
&gt;ACT28234<span class="w"> </span>pep<span class="w"> </span>supercontig:ASM2366v1:CP001665:1230560:1230991:1<span class="w"> </span>gene:ECBD_1168<span class="w"> </span>transcript:ACT28234<span class="w"> </span>gene_biotype:protein_coding<span class="w"> </span>transcript_biotype:protein_coding<span class="w"> </span>description:Nucleoside-diphosphate<span class="w"> </span>kinase
&gt;ACT29418<span class="w"> </span>pep<span class="w"> </span>supercontig:ASM2366v1:CP001665:2508388:2509098:-1<span class="w"> </span>gene:ECBD_2392<span class="w"> </span>transcript:ACT29418<span class="w"> </span>gene_biotype:protein_coding<span class="w"> </span>transcript_biotype:protein_coding<span class="w"> </span>description:nitrate<span class="w"> </span>reductase<span class="w"> </span>molybdenum<span class="w"> </span>cofactor<span class="w"> </span>assembly<span class="w"> </span>chaperone
&gt;ACT31080<span class="w"> </span>pep<span class="w"> </span>supercontig:ASM2366v1:CP001665:4316460:4317305:1<span class="w"> </span>gene:ECBD_4097<span class="w"> </span>transcript:ACT31080<span class="w"> </span>gene_biotype:protein_coding<span class="w"> </span>transcript_biotype:protein_coding<span class="w"> </span>description:MIP<span class="w"> </span>family<span class="w"> </span>channel<span class="w"> </span>protein
&gt;ACT31118<span class="w"> </span>pep<span class="w"> </span>supercontig:ASM2366v1:CP001665:4355734:4355916:-1<span class="w"> </span>gene:ECBD_4135<span class="w"> </span>transcript:ACT31118<span class="w"> </span>gene_biotype:protein_coding<span class="w"> </span>transcript_biotype:protein_coding<span class="w"> </span>description:hypothetical<span class="w"> </span>protein
</pre></div>
</div>
<p>If we want to extract also the corresponding sequence the situation is more complex.</p>
<p>First of all we need to convert the fasta format in a tab separated format with two columns: and id and a sequence.
And then use <strong>grep</strong> again to extract our sequences of interest.
The conversion can be achieved using one of the most powerful linux tool, that is a programming language: <strong>awk</strong></p>
<ul class="simple">
<li><p>Awk’s basic syntax:</p></li>
</ul>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>awk<span class="w"> </span><span class="s1">&#39;OPTIONAL PATTERN {SOME INSTRUCTIONS}&#39;</span><span class="w"> </span>FILENAME
</pre></div>
</div>
<p>Awk reads the files line by line.</p>
<p>As a naive example we can just print the content of the file using <strong>awk</strong> (<strong>$0</strong> is the whole line):</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>awk<span class="w"> </span><span class="s1">&#39;{print $0}&#39;</span><span class="w">  </span>Escherichia_coli_bl21_gold_de3_plyss_ag_.ASM2366v1.pep.all.fa<span class="w"> </span><span class="p">|</span>head<span class="w"> </span>-n<span class="w"> </span><span class="m">3</span>

&gt;ACT27082<span class="w"> </span>pep<span class="w"> </span>supercontig:ASM2366v1:CP001665:347:1750:1<span class="w"> </span>gene:ECBD_0001<span class="w"> </span>transcript:ACT27082<span class="w"> </span>gene_biotype:protein_coding<span class="w"> </span>transcript_biotype:protein_coding<span class="w"> </span>description:chromosomal<span class="w"> </span>replication<span class="w"> </span>initiator<span class="w"> </span>protein<span class="w"> </span>DnaA
MSLSLWQQCLARLQDELPATEFSMWIRPLQAELSDNTLALYAPNRFVLDWVRDKYLNNIN
GLLTSFCGADAPQLRFEVGTKPVTQTPQAAVTSNVAAPAQVAQTQPQRAAPSTRSGWDNV
</pre></div>
</div>
<p>Or we can remove the carriage return by setting the built-in variable <strong>ORS</strong> to empty (Output Record Separator Variable)</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>head<span class="w"> </span>-n<span class="w"> </span><span class="m">10</span><span class="w"> </span>Escherichia_coli_bl21_gold_de3_plyss_ag_.ASM2366v1.pep.all.fa<span class="w"> </span><span class="p">|</span><span class="w"> </span>awk<span class="w"> </span><span class="s1">&#39;{ORS=&quot;&quot;; print $0}&#39;</span>

&gt;ACT27082<span class="w"> </span>pep<span class="w"> </span>supercontig:ASM2366v1:CP001665:347:1750:1<span class="w"> </span>gene:ECBD_0001<span class="w"> </span>transcript:ACT27082<span class="w"> </span>gene_biotype:protein_coding<span class="w"> </span>transcript_biotype:protein_coding<span class="w"> </span>description:chromosomal<span class="w"> </span>replication<span class="w"> </span>initiator<span class="w"> </span>protein<span class="w"> </span>DnaAMSLSLWQQCLARLQDELPATEFSMWIRPLQAELSDNTLALYAPNRFVLDWVRDKYLNNINGLLTSFCGADAPQLRFEVGTKPVTQTPQAAVTSNVAAPAQVAQTQPQRAAPSTRSGWDNVPAPAEPTYRSNVNVKHTFDNFVEGKSNQLARAAARQVADNPGGAYNPLFLYGGTGLGKTHLLHAVGNGIMARKPNAKVVYMHSERFVQDMVKALQNNAIEEFKRYYRSVDALLIDDIQFFANKERSQEEFFHTFNALLEGNQQIILTSDRYPKEINGVEDRLKSRFGWGLTVAIEPPELETRVAILMKKADENDIRLPGEVAFFIAKRLRSNVRELEGALNRVIANANFTGRAITIDFVREALRDLLALQEKLVTIDNIQKTVAEYYKIKVADLLSKRRSRSVARPRQMAMALAKELTNHSLPEIGDAFGGRDHTTVLHACRKIEQLREESHDIKEDFSNLIRTLSS&gt;ACT27083<span class="w"> </span>pep<span class="w"> </span>supercontig:ASM2366v1:CP001665:1755:2855:1<span class="w"> </span>gene:ECBD_0002<span class="w"> </span>transcript:ACT27083<span class="w"> </span>gene_biotype:protein_coding<span class="w"> </span>transcript_biotype:protein_coding<span class="w"> </span>description:DNA<span class="w"> </span>polymerase<span class="w"> </span>III,<span class="w"> </span>beta<span class="w"> </span>subunit
</pre></div>
</div>
<p>At this point using we need a <strong>if</strong> statement to reach the point. In few words this statement says: EXECUTE a piece of code IF a given condition is met OTHERWISE (<strong>else</strong>) do something else.</p>
<p>As an example we can use the <strong>if</strong> to select the header like a <strong>grep</strong> function using the matching expression tilde <strong>~</strong> with the character <strong>*&gt;*</strong></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>awk<span class="w"> </span><span class="s1">&#39;{if ($0~&quot;&gt;&quot;) {print $0}}&#39;</span><span class="w"> </span>Escherichia_coli_bl21_gold_de3_plyss_ag_.ASM2366v1.pep.all.fa<span class="w"> </span><span class="p">|</span>head<span class="w"> </span>-n<span class="w"> </span><span class="m">3</span>

&gt;ACT27082<span class="w"> </span>pep<span class="w"> </span>supercontig:ASM2366v1:CP001665:347:1750:1<span class="w"> </span>gene:ECBD_0001<span class="w"> </span>transcript:ACT27082<span class="w"> </span>gene_biotype:protein_coding<span class="w"> </span>transcript_biotype:protein_coding<span class="w"> </span>description:chromosomal<span class="w"> </span>replication<span class="w"> </span>initiator<span class="w"> </span>protein<span class="w"> </span>DnaA
&gt;ACT27083<span class="w"> </span>pep<span class="w"> </span>supercontig:ASM2366v1:CP001665:1755:2855:1<span class="w"> </span>gene:ECBD_0002<span class="w"> </span>transcript:ACT27083<span class="w"> </span>gene_biotype:protein_coding<span class="w"> </span>transcript_biotype:protein_coding<span class="w"> </span>description:DNA<span class="w"> </span>polymerase<span class="w"> </span>III,<span class="w"> </span>beta<span class="w"> </span>subunit
&gt;ACT27084<span class="w"> </span>pep<span class="w"> </span>supercontig:ASM2366v1:CP001665:2855:3928:1<span class="w"> </span>gene:ECBD_0003<span class="w"> </span>transcript:ACT27084<span class="w"> </span>gene_biotype:protein_coding<span class="w"> </span>transcript_biotype:protein_coding<span class="w"> </span>description:DNA<span class="w"> </span>replication<span class="w"> </span>and<span class="w"> </span>repair<span class="w"> </span>protein<span class="w"> </span>RecF
</pre></div>
</div>
<p>Note that this syntax can be simplified when looking for patterns:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>awk<span class="w"> </span><span class="s1">&#39;$0 ~ &quot;&gt;&quot; {print $0}&#39;</span><span class="w"> </span>Escherichia_coli_bl21_gold_de3_plyss_ag_.ASM2366v1.pep.all.fa<span class="w"> </span><span class="p">|</span><span class="w"> </span>head<span class="w"> </span>-n<span class="w"> </span><span class="m">3</span>

&gt;ACT27082<span class="w"> </span>pep<span class="w"> </span>supercontig:ASM2366v1:CP001665:347:1750:1<span class="w"> </span>gene:ECBD_0001<span class="w"> </span>transcript:ACT27082<span class="w"> </span>gene_biotype:protein_coding<span class="w"> </span>transcript_biotype:protein_coding<span class="w"> </span>description:chromosomal<span class="w"> </span>replication<span class="w"> </span>initiator<span class="w"> </span>protein<span class="w"> </span>DnaA
&gt;ACT27083<span class="w"> </span>pep<span class="w"> </span>supercontig:ASM2366v1:CP001665:1755:2855:1<span class="w"> </span>gene:ECBD_0002<span class="w"> </span>transcript:ACT27083<span class="w"> </span>gene_biotype:protein_coding<span class="w"> </span>transcript_biotype:protein_coding<span class="w"> </span>description:DNA<span class="w"> </span>polymerase<span class="w"> </span>III,<span class="w"> </span>beta<span class="w"> </span>subunit
&gt;ACT27084<span class="w"> </span>pep<span class="w"> </span>supercontig:ASM2366v1:CP001665:2855:3928:1<span class="w"> </span>gene:ECBD_0003<span class="w"> </span>transcript:ACT27084<span class="w"> </span>gene_biotype:protein_coding<span class="w"> </span>transcript_biotype:protein_coding<span class="w"> </span>description:DNA<span class="w"> </span>replication<span class="w"> </span>and<span class="w"> </span>repair<span class="w"> </span>protein<span class="w"> </span>RecF
</pre></div>
</div>
<p>So, combining the previous example, we can remove the carriage return and in case we found the <strong>&gt;</strong> character we print that row preceded by a carriage return and followed by a tab (<strong>t</strong>)</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>awk<span class="w"> </span><span class="s1">&#39;{ORS=&quot;&quot;; if ($0~&quot;&gt;&quot;) {print &quot;\n&quot;$0&quot;\t&quot;} else {print $0}}&#39;</span><span class="w"> </span>Escherichia_coli_bl21_gold_de3_plyss_ag_.ASM2366v1.pep.all.fa<span class="w"> </span><span class="p">|</span>head<span class="w"> </span>-n<span class="w"> </span><span class="m">3</span>

&gt;ACT27082<span class="w"> </span>pep<span class="w"> </span>supercontig:ASM2366v1:CP001665:347:1750:1<span class="w"> </span>gene:ECBD_0001<span class="w"> </span>transcript:ACT27082<span class="w"> </span>gene_biotype:protein_coding<span class="w"> </span>transcript_biotype:protein_coding<span class="w"> </span>description:chromosomal<span class="w"> </span>replication<span class="w"> </span>initiator<span class="w"> </span>protein<span class="w"> </span>DnaA<span class="w">     </span>MSLSLWQQCLARLQDELPATEFSMWIRPLQAELSDNTLALYAPNRFVLDWVRDKYLNNINGLLTSFCGADAPQLRFEVGTKPVTQTPQAAVTSNVAAPAQVAQTQPQRAAPSTRSGWDNVPAPAEPTYRSNVNVKHTFDNFVEGKSNQLARAAARQVADNPGGAYNPLFLYGGTGLGKTHLLHAVGNGIMARKPNAKVVYMHSERFVQDMVKALQNNAIEEFKRYYRSVDALLIDDIQFFANKERSQEEFFHTFNALLEGNQQIILTSDRYPKEINGVEDRLKSRFGWGLTVAIEPPELETRVAILMKKADENDIRLPGEVAFFIAKRLRSNVRELEGALNRVIANANFTGRAITIDFVREALRDLLALQEKLVTIDNIQKTVAEYYKIKVADLLSKRRSRSVARPRQMAMALAKELTNHSLPEIGDAFGGRDHTTVLHACRKIEQLREESHDIKEDFSNLIRTLSS
&gt;ACT27083<span class="w"> </span>pep<span class="w"> </span>supercontig:ASM2366v1:CP001665:1755:2855:1<span class="w"> </span>gene:ECBD_0002<span class="w"> </span>transcript:ACT27083<span class="w"> </span>gene_biotype:protein_coding<span class="w"> </span>transcript_biotype:protein_coding<span class="w"> </span>description:DNA<span class="w"> </span>polymerase<span class="w"> </span>III,<span class="w"> </span>beta<span class="w"> </span>subunit<span class="w">  </span>MKFTVEREHLLKPLQQVSGPLGGRPTLPILGNLLLQVADGTLSLTGTDLEMEMVARVALVQPHEPGATTVPARKFFDICRGLPEGAEIAVQLEGERMLVRSGRSRFSLSTLPAADFPNLDDWQSEVEFTLPQATMKRLIEATQFSMAHQDVRYYLNGMLFETEGEELRTVATDGHRLAVCSMPIGQSLPSHSVIVPRKGVIELMRMLDGGDNPLRVQIGSNNIRAHVGDFIFTSKLVDGRFPDYRRVLPKNPDKHLEAGCDLLKQAFARAAILSNEKFRGVRLYVSENQLKITANNPEQEEAEEILDVTYSGAEMEIGFNVSYVLDVLNALKCENVRMMLTDSVSSVQIEDAASQSAAYVVMPMRL

awk<span class="w"> </span><span class="s1">&#39;{ORS=&quot;&quot;; if ($0~&quot;&gt;&quot;) {print &quot;\n&quot;$0&quot;\t&quot;} else {print $0}}&#39;</span><span class="w"> </span>Escherichia_coli_bl21_gold_de3_plyss_ag_.ASM2366v1.pep.all.fa<span class="w"> </span>&gt;<span class="w"> </span>proteins.tab

wc<span class="w"> </span>-l<span class="w"> </span>proteins.tab
<span class="m">4228</span><span class="w"> </span>proteins.tab

wc<span class="w"> </span>-l<span class="w"> </span>seq_names.txt
<span class="m">4228</span><span class="w"> </span>seq_names.txt
</pre></div>
</div>
<p>So now we can use the grep command to extract our sequence of interest.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>grep<span class="w"> </span>-Ff<span class="w"> </span>random.list<span class="w"> </span>proteins.tab

&gt;ACT27123<span class="w"> </span>pep<span class="w"> </span>supercontig:ASM2366v1:CP001665:44295:44414:-1<span class="w"> </span>gene:ECBD_0043<span class="w"> </span>transcript:ACT27123<span class="w"> </span>gene_biotype:protein_coding<span class="w"> </span>transcript_biotype:protein_coding<span class="w"> </span>description:hypothetical<span class="w"> </span>protein<span class="w">   </span>MEYKVWHFLLTTQARFVQHDESDESKLHLCFIRYTFVKG
&gt;ACT28234<span class="w"> </span>pep<span class="w"> </span>supercontig:ASM2366v1:CP001665:1230560:1230991:1<span class="w"> </span>gene:ECBD_1168<span class="w"> </span>transcript:ACT28234<span class="w"> </span>gene_biotype:protein_coding<span class="w"> </span>transcript_biotype:protein_coding<span class="w"> </span>description:Nucleoside-diphosphate<span class="w"> </span>kinase<span class="w">       </span>MAIERTFSIIKPNAVAKNVIGNIFARFEAAGFKIVGTKMLHLTVEQARGFYAEHDGKPFFDGLVEFMTSGPIVVSVLEGENAVQRHRDLLGATNPANALAGTLRADYADSLTENGTHGSDSVESAAREIAYFFGEGEVCPRTR
&gt;ACT29418<span class="w"> </span>pep<span class="w"> </span>supercontig:ASM2366v1:CP001665:2508388:2509098:-1<span class="w"> </span>gene:ECBD_2392<span class="w"> </span>transcript:ACT29418<span class="w"> </span>gene_biotype:protein_coding<span class="w"> </span>transcript_biotype:protein_coding<span class="w"> </span>description:nitrate<span class="w"> </span>reductase<span class="w"> </span>molybdenum<span class="w"> </span>cofactor<span class="w"> </span>assembly<span class="w"> </span>chaperone<span class="w">   </span>MIELVIVSRLLEYPDAALWQHQQEMFEAIAASKNLSKEDAHALGIFLRDLTAMDPLDAQAQYSELFDRGRATSLLLFEHVHGESRDRGQAMVDLLAQYEQHGLQLNSRELPDHLPLYLEYLSQLPQSEAVEGLKDIAPILALLSARLQQRESRYAVMFDLLLKLANTAIDSDKVAEKIADEARDDTPQALDAVWEEEQVKFFADKGCGDSAITAHQRRFAGAVAPQYLNITTGGQH
&gt;ACT31080<span class="w"> </span>pep<span class="w"> </span>supercontig:ASM2366v1:CP001665:4316460:4317305:1<span class="w"> </span>gene:ECBD_4097<span class="w"> </span>transcript:ACT31080<span class="w"> </span>gene_biotype:protein_coding<span class="w"> </span>transcript_biotype:protein_coding<span class="w"> </span>description:MIP<span class="w"> </span>family<span class="w"> </span>channel<span class="w"> </span>protein<span class="w">  </span>MSQTSTLKGQCIAEFLGTGLLIFFGVGCVAALKVAGASFGQWEISVIWGLGVAMAIYLTAGVSGAHLNPAVTIALWLFACFDKRKVIPFIVSQVAGAFCAAALVYGLYYNLFFDFEQTHHIVRGSVESVDLAGTFSTYPNPHINFVQAFAVEMVITAILMGLILALTDDGNGVPRGPLAPLLIGLLIAVIGASMGPLTGFAMNPARDFGPKVFAWLAGWGNVAFTGGRDIPYFLVPLFGPIVGAIVGAFAYRKLIGRHLPCDICVVEEKETTTPSEQKASL
&gt;ACT31118<span class="w"> </span>pep<span class="w"> </span>supercontig:ASM2366v1:CP001665:4355734:4355916:-1<span class="w"> </span>gene:ECBD_4135<span class="w"> </span>transcript:ACT31118<span class="w"> </span>gene_biotype:protein_coding<span class="w"> </span>transcript_biotype:protein_coding<span class="w"> </span>description:hypothetical<span class="w"> </span>protein<span class="w">       </span>MGKNDVNQIADNVRVVHAGCGVNALSGLQSRINSMYCSLLVGLISAAHQAILRLSSVSCP
</pre></div>
</div>
<section id="id1">
<h3><strong>Recap</strong><a class="headerlink" href="#id1" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p><strong>cut</strong> extract the indicated column</p></li>
<li><p><strong>awk</strong> allows several kind of parsing operations</p></li>
<li><p><strong>head</strong> extract the indicated number of rows from the beginning of a file</p></li>
</ul>
</section>
</section>
<section id="sequence-file-formats-fasta-and-fastq">
<h2>Sequence file formats: Fasta and fastq<a class="headerlink" href="#sequence-file-formats-fasta-and-fastq" title="Permalink to this heading"></a></h2>
<p>We already showed the fasta format. There is a header characterized by the presence of <strong>&gt;</strong> and a number of rows containing the sequence.
The format is used for both nucleic acids and proteins.</p>
<a class="reference internal image-reference" href="_images/fasta_format.png"><img alt="_images/fasta_format.png" src="_images/fasta_format.png" style="width: 600px;" /></a>
<p>Another way to store sequencing data, and particularly the short reads coming from the sequencers is the <strong>Fastq</strong> format.
This format allows to store also the information about the quality of that particular base, i.e. the probability that that base reading was true or not.</p>
<a class="reference internal image-reference" href="_images/fastq_format.png"><img alt="_images/fastq_format.png" src="_images/fastq_format.png" style="width: 600px;" /></a>
<p>The format contains four rows for sequence with:
* a header containing <strong>&#64;</strong> as the first character
* the sequence content
* a <strong>spacer</strong>
* the quality encoded using ASCII characters.</p>
<a class="reference internal image-reference" href="_images/phred_quality.png"><img alt="_images/phred_quality.png" src="_images/phred_quality.png" style="width: 600px;" /></a>
<p>Currently most of the journals require the submissions of NGS data in a public repository upon publishing.</p>
<p>The major repositories are:</p>
<ul class="simple">
<li><p><a class="reference external" href="https://www.ncbi.nlm.nih.gov/sra">SRA</a> (Sequence Read Archive) from US</p></li>
<li><p><a class="reference external" href="https://www.ebi.ac.uk/ena">ENA</a> (European Nucleotide Archive)</p></li>
<li><p><a class="reference external" href="https://www.ddbj.nig.ac.jp/dra/index-e.html">DDBJ-DRA</a> from Japan.</p></li>
</ul>
<p>They are interconnected and mirror the data among them and are connected to other databases that contain also the results of a given analysis such as <a class="reference external" href="https://www.ncbi.nlm.nih.gov/geo/">GEO</a> and <a class="reference external" href="https://www.ebi.ac.uk/arrayexpress/">Array-express</a> that contain expression data.</p>
<p>Let’s download a test fastq files we stored in our repository with the access <strong>SRR6466185</strong> using again <strong>wget</strong>. Then we untar the file.</p>
<p>We should have 32,345 reads. Let’s try to count them using a combination of <strong>zcat</strong> that uncompress the file and writes the content to the terminal and <strong>wc</strong>.</p>
<p>So considering that each sequence is defined by 4 rows we should have <code class="docutils literal notranslate"><span class="pre">32,345</span> <span class="pre">*</span> <span class="pre">4</span> <span class="pre">=</span> <span class="pre">129,380</span></code>. We can also use <strong>awk</strong> for embedding also the division:</p>
<section id="id2">
<h3><strong>Recap</strong><a class="headerlink" href="#id2" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p><strong>tar</strong> allows to compress / uncompress more files in one</p></li>
<li><p><strong>zcat</strong> equivalent of <strong>cat</strong> for gzipped files</p></li>
</ul>
</section>
</section>
<section id="bed-format-and-regular-expressions">
<h2>Bed format and regular expressions<a class="headerlink" href="#bed-format-and-regular-expressions" title="Permalink to this heading"></a></h2>
<p>The BED format is a “tab” separated text file. It consists of one line per feature, each containing 3-12 columns of data.
It is used for indicating genomic locations as the one of exons, binding sites, regulatory elements, etc.</p>
<p>Typical 6-fields bed format</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>chrom</p></th>
<th class="head"><p>chromStart</p></th>
<th class="head"><p>chromEnd</p></th>
<th class="head"><p>name</p></th>
<th class="head"><p>score</p></th>
<th class="head"><p>strand</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>chr7</p></td>
<td><p>127471196</p></td>
<td><p>127472363</p></td>
<td><p>Pos1</p></td>
<td><p>0</p></td>
<td><ul class="simple">
<li></li>
</ul>
</td>
</tr>
<tr class="row-odd"><td><p>chr7</p></td>
<td><p>127472363</p></td>
<td><p>127473530</p></td>
<td><p>Pos2</p></td>
<td><p>0</p></td>
<td><ul class="simple">
<li></li>
</ul>
</td>
</tr>
<tr class="row-even"><td><p>chr7</p></td>
<td><p>127473530</p></td>
<td><p>127474697</p></td>
<td><p>Pos3</p></td>
<td><p>0</p></td>
<td><ul class="simple">
<li></li>
</ul>
</td>
</tr>
<tr class="row-odd"><td><p>chr7</p></td>
<td><p>127474697</p></td>
<td><p>127475864</p></td>
<td><p>Pos4</p></td>
<td><p>0</p></td>
<td><ul class="simple">
<li></li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>Additionally you may have up to 6 more fields:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>thickStart</p></th>
<th class="head"><p>thickEnd</p></th>
<th class="head"><p>itemRgb</p></th>
<th class="head"><p>blockCount</p></th>
<th class="head"><p>blockSizes</p></th>
<th class="head"><p>blockStarts</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>127471196</p></td>
<td><p>127472363</p></td>
<td><p>255,0,0</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
</tr>
<tr class="row-odd"><td><p>127472363</p></td>
<td><p>127473530</p></td>
<td><p>255,0,0</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
</tr>
<tr class="row-even"><td><p>127473530</p></td>
<td><p>127474697</p></td>
<td><p>255,0,0</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
</tr>
<tr class="row-odd"><td><p>127474697</p></td>
<td><p>127475864</p></td>
<td><p>255,0,0</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
</tr>
</tbody>
</table>
<p>This kind of file can be fed to a genome browser like <a class="reference external" href="https://genome-euro.ucsc.edu/cgi-bin/hgGateway?redirect=manual&amp;source=genome.ucsc.edu">UCSC genome browser</a> to highlight the genomic positions. Here an example about our coordinates:</p>
<section id="genome-browser">
<h3>Genome browser<a class="headerlink" href="#genome-browser" title="Permalink to this heading"></a></h3>
<a class="reference internal image-reference" href="_images/genome_browser.png"><img alt="_images/genome_browser.png" src="_images/genome_browser.png" style="width: 900px;" /></a>
<p>BED files can be uploaded to public databases like <strong>GEO</strong> and <strong>ArrayExpress</strong>. As an example let’s download the putative binding site positions obtained by a ChIP-seq experiment on Suz12 transcription factor <a class="reference external" href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE41589">GSE41589</a>.</p>
<a class="reference internal image-reference" href="_images/GEO_chipseq.png"><img alt="_images/GEO_chipseq.png" src="_images/GEO_chipseq.png" style="width: 600px;" /></a>
<p>Let’s use the right-click on FTP link to copy the link and <strong>wget</strong> to download the file.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>wget<span class="w"> </span>ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE41nnn/GSE41589/suppl/GSE41589_Suz12_BindingSites.txt.gz

--2019-03-08<span class="w"> </span><span class="m">15</span>:03:40--<span class="w">  </span>ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE41nnn/GSE41589/suppl/GSE41589_Suz12_BindingSites.txt.gz
<span class="w">           </span><span class="o">=</span>&gt;<span class="w"> </span>‘GSE41589_Suz12_BindingSites.txt.gz’
Resolving<span class="w"> </span>ftp.ncbi.nlm.nih.gov<span class="w"> </span><span class="o">(</span>ftp.ncbi.nlm.nih.gov<span class="o">)</span>...<span class="w"> </span><span class="m">130</span>.14.250.12,<span class="w"> </span><span class="m">2607</span>:f220:41e:250::11
Connecting<span class="w"> </span>to<span class="w"> </span>ftp.ncbi.nlm.nih.gov<span class="w"> </span><span class="o">(</span>ftp.ncbi.nlm.nih.gov<span class="o">)</span><span class="p">|</span><span class="m">130</span>.14.250.12<span class="p">|</span>:21...<span class="w"> </span>connected.
Logging<span class="w"> </span><span class="k">in</span><span class="w"> </span>as<span class="w"> </span>anonymous<span class="w"> </span>...<span class="w"> </span>Logged<span class="w"> </span><span class="k">in</span>!
<span class="o">==</span>&gt;<span class="w"> </span>SYST<span class="w"> </span>...<span class="w"> </span><span class="k">done</span>.<span class="w">    </span><span class="o">==</span>&gt;<span class="w"> </span>PWD<span class="w"> </span>...<span class="w"> </span><span class="k">done</span>.
<span class="o">==</span>&gt;<span class="w"> </span>TYPE<span class="w"> </span>I<span class="w"> </span>...<span class="w"> </span><span class="k">done</span>.<span class="w">  </span><span class="o">==</span>&gt;<span class="w"> </span>CWD<span class="w"> </span><span class="o">(</span><span class="m">1</span><span class="o">)</span><span class="w"> </span>/geo/series/GSE41nnn/GSE41589/suppl<span class="w"> </span>...<span class="w"> </span><span class="k">done</span>.
<span class="o">==</span>&gt;<span class="w"> </span>SIZE<span class="w"> </span>GSE41589_Suz12_BindingSites.txt.gz<span class="w"> </span>...<span class="w"> </span><span class="nv">72526</span>
<span class="o">==</span>&gt;<span class="w"> </span>PASV<span class="w"> </span>...<span class="w"> </span><span class="k">done</span>.<span class="w">    </span><span class="o">==</span>&gt;<span class="w"> </span>RETR<span class="w"> </span>GSE41589_Suz12_BindingSites.txt.gz<span class="w"> </span>...<span class="w"> </span><span class="k">done</span>.
Length:<span class="w"> </span><span class="m">72526</span><span class="w"> </span><span class="o">(</span>71K<span class="o">)</span><span class="w"> </span><span class="o">(</span>unauthoritative<span class="o">)</span>

<span class="m">100</span>%<span class="o">[======================================================================================================================</span>&gt;<span class="o">]</span><span class="w"> </span><span class="m">72</span>,526<span class="w">       </span>258KB/s<span class="w">   </span><span class="k">in</span><span class="w"> </span><span class="m">0</span>.3s

<span class="m">2019</span>-03-08<span class="w"> </span><span class="m">15</span>:03:42<span class="w"> </span><span class="o">(</span><span class="m">258</span><span class="w"> </span>KB/s<span class="o">)</span><span class="w"> </span>-<span class="w"> </span>‘GSE41589_Suz12_BindingSites.txt.gz’<span class="w"> </span>saved<span class="w"> </span><span class="o">[</span><span class="m">72526</span><span class="o">]</span>
</pre></div>
</div>
<p>Since we have one row per feature, just counting the number of rows will give us the number of binding sites:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>zcat<span class="w"> </span>GSE41589_Suz12_BindingSites.txt.gz<span class="w"> </span><span class="p">|</span><span class="w"> </span>wc<span class="w"> </span>-l

<span class="m">8053</span>
</pre></div>
</div>
<p>While extracting the first column with <strong>cut</strong> and piping the result to <strong>uniq</strong> will give us the list of chromosomes in which there is at lease one binding site</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>zcat<span class="w"> </span>GSE41589_Suz12_BindingSites.txt.gz<span class="w"> </span><span class="p">|</span><span class="w"> </span>cut<span class="w"> </span>-f<span class="w"> </span><span class="m">1</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>uniq
chr1
chr10
chr11
chr12
chr13
chr14
chr15
chr16
chr17
chr18
chr19
chr2
chr3
chr4
chr5
chr6
chr7
chr8
chr9
chrX
chrY
</pre></div>
</div>
<p>We should not use <strong>uniq</strong> alone if we are not sure the values are not sorted… As an example if we shuffle the chromosome order <strong>uniq</strong> won’t work as expected</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>zcat<span class="w"> </span>GSE41589_Suz12_BindingSites.txt.gz<span class="w"> </span><span class="p">|</span><span class="w"> </span>cut<span class="w"> </span>-f<span class="w"> </span><span class="m">1</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>uniq<span class="w"> </span><span class="p">|</span><span class="w"> </span>wc<span class="w"> </span>-l
<span class="m">21</span>

zcat<span class="w"> </span>GSE41589_Suz12_BindingSites.txt.gz<span class="w"> </span><span class="p">|</span><span class="w"> </span>cut<span class="w"> </span>-f<span class="w"> </span><span class="m">1</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>shuf<span class="w"> </span><span class="p">|</span><span class="w"> </span>uniq<span class="w"> </span><span class="p">|</span><span class="w"> </span>wc<span class="w"> </span>-l
<span class="m">7633</span>
</pre></div>
</div>
<p>We can fix this using another tool named <strong>sort</strong></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>zcat<span class="w"> </span>GSE41589_Suz12_BindingSites.txt.gz<span class="w"> </span><span class="p">|</span><span class="w"> </span>cut<span class="w"> </span>-f<span class="w"> </span><span class="m">1</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>shuf<span class="w"> </span><span class="p">|</span><span class="w"> </span>sort<span class="w"> </span><span class="p">|</span><span class="w"> </span>uniq<span class="w"> </span><span class="p">|</span><span class="w"> </span>wc<span class="w"> </span>-l
<span class="m">21</span>
</pre></div>
</div>
<p>The tool <strong>uniq</strong> has an interesting parameter called <strong>-c</strong> that gives us the number of times that a particular row was found. In this way we have the number of binding site per chromosome.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>zcat<span class="w"> </span>GSE41589_Suz12_BindingSites.txt.gz<span class="w"> </span><span class="p">|</span><span class="w"> </span>cut<span class="w"> </span>-f<span class="w"> </span><span class="m">1</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>shuf<span class="w"> </span><span class="p">|</span><span class="w"> </span>sort<span class="w"> </span><span class="p">|</span><span class="w"> </span>uniq<span class="w"> </span>-c
<span class="w">    </span><span class="m">525</span><span class="w"> </span>chr1
<span class="w">    </span><span class="m">362</span><span class="w"> </span>chr10
<span class="w">    </span><span class="m">603</span><span class="w"> </span>chr11
<span class="w">    </span><span class="m">367</span><span class="w"> </span>chr12
<span class="w">    </span><span class="m">392</span><span class="w"> </span>chr13
<span class="w">    </span><span class="m">314</span><span class="w"> </span>chr14
<span class="w">    </span><span class="m">351</span><span class="w"> </span>chr15
<span class="w">    </span><span class="m">193</span><span class="w"> </span>chr16
<span class="w">    </span><span class="m">303</span><span class="w"> </span>chr17
<span class="w">    </span><span class="m">295</span><span class="w"> </span>chr18
<span class="w">    </span><span class="m">232</span><span class="w"> </span>chr19
<span class="w">    </span><span class="m">672</span><span class="w"> </span>chr2
<span class="w">    </span><span class="m">397</span><span class="w"> </span>chr3
<span class="w">    </span><span class="m">547</span><span class="w"> </span>chr4
<span class="w">    </span><span class="m">511</span><span class="w"> </span>chr5
<span class="w">    </span><span class="m">445</span><span class="w"> </span>chr6
<span class="w">    </span><span class="m">480</span><span class="w"> </span>chr7
<span class="w">    </span><span class="m">381</span><span class="w"> </span>chr8
<span class="w">    </span><span class="m">435</span><span class="w"> </span>chr9
<span class="w">    </span><span class="m">244</span><span class="w"> </span>chrX
<span class="w">      </span><span class="m">5</span><span class="w"> </span>chrY
</pre></div>
</div>
<p>Now let’s try to extract the binding sites from the <strong>chromosome 1</strong>. We can use <strong>grep</strong> but we will extract also unwanted things…</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>zcat<span class="w"> </span>GSE41589_Suz12_BindingSites.txt.gz<span class="p">|</span><span class="w"> </span>grep<span class="w"> </span>chr1<span class="p">|</span><span class="w"> </span>cut<span class="w"> </span>-f<span class="w"> </span><span class="m">1</span><span class="p">|</span><span class="w"> </span>uniq
chr1
chr10
chr11
chr12
chr13
chr14
chr15
chr16
chr17
chr18
chr19
</pre></div>
</div>
<p>To avoid to extract also other chromosomes that simply start for chr1 we can use the option <strong>-w</strong></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>zcat<span class="w"> </span>GSE41589_Suz12_BindingSites.txt.gz<span class="p">|</span><span class="w"> </span>grep<span class="w"> </span>-w<span class="w"> </span>chr1<span class="p">|</span><span class="w"> </span>cut<span class="w"> </span>-f<span class="w"> </span><span class="m">1</span><span class="p">|</span><span class="w"> </span>uniq
chr1
</pre></div>
</div>
<p>If you want more complex search we can use the <strong>regular expression</strong>: a sequence of characters that define a search pattern.</p>
<p>Single characters bewteen square brackets can be searched at the same time</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>zcat<span class="w"> </span>GSE41589_Suz12_BindingSites.txt.gz<span class="p">|</span><span class="w"> </span>grep<span class="w"> </span>-w<span class="w"> </span><span class="s2">&quot;chr[129]&quot;</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>cut<span class="w"> </span>-f<span class="w"> </span><span class="m">1</span><span class="p">|</span>uniq
chr1
chr2
chr9
</pre></div>
</div>
<p>For searching chromosomes with two digits you should use more intervals</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>zcat<span class="w"> </span>GSE41589_Suz12_BindingSites.txt.gz<span class="p">|</span><span class="w"> </span>grep<span class="w"> </span>-w<span class="w"> </span><span class="s2">&quot;chr[1][123]&quot;</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>cut<span class="w"> </span>-f<span class="w"> </span><span class="m">1</span><span class="p">|</span>uniq
chr11
chr12
chr13
</pre></div>
</div>
<p>Using a hiphen <strong>-</strong> allows to extract a whole interval from the first to the last number</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>zcat<span class="w"> </span>GSE41589_Suz12_BindingSites.txt.gz<span class="p">|</span><span class="w"> </span>grep<span class="w"> </span>-w<span class="w"> </span><span class="s2">&quot;chr[1][0-9]&quot;</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>cut<span class="w"> </span>-f<span class="w"> </span><span class="m">1</span><span class="p">|</span>uniq
chr10
chr11
chr12
chr13
chr14
chr15
chr16
chr17
chr18
chr19
</pre></div>
</div>
<p>Another useful regular expression is <strong>^</strong> and <strong>$</strong> that indicates that the pattern has to be found at the beginning or the end of the string. Also <strong>awk</strong> is able to use regular expressions, this will make the search more accurate since you can decide which column to scan.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>awk<span class="w"> </span>-F<span class="s2">&quot;\t&quot;</span><span class="w"> </span><span class="s1">&#39;{if ($2~&quot;^MCELDI&quot;) print}&#39;</span><span class="w"> </span>proteins.tab

&gt;ACT27842<span class="w"> </span>pep<span class="w"> </span>supercontig:ASM2366v1:CP001665:800015:801223:-1<span class="w"> </span>gene:ECBD_0774<span class="w"> </span>transcript:ACT27842<span class="w"> </span>gene_biotype:protein_coding<span class="w"> </span>transcript_biotype:protein_coding<span class="w"> </span>description:transposase<span class="w"> </span>IS4<span class="w"> </span>family<span class="w"> </span>protein<span class="w">       </span>MCELDILHDSLYQFCPELHLKRLNSLTLACHALLDCKTLTLTELGRNLPTKARTKHNIKRIDRLLGNRHLHKERLAVYRWHASFICSGNTMPIVLVDWSDIREQKRLMVLRASVALHGRSVTLYEKAFPLSEQCSKKAHDQFLADLASILPSNTTPLIVSDAGFKVPWYKSVEKLGWYWLSRVRGKVQYADLGAENWKPISNLHDMSSSHSKTLGYKRLTKSNPISCQILLYKSRSKGRKNQRSTRTHCHHPSPKIYSASAKEPWILATNLPVEIRTPKQLVNIYSKRMQIEETFRDLKSPAYGLGLRHSRTSSSERFDIMLLIALMLQLTCWLAGVHAQKQGWDKHFQANTVRNRNVLSTVRLGMEVLRHSGYTITREDSLVAATLLTQNLFTHGYVLGKL
&gt;ACT27843<span class="w"> </span>pep<span class="w"> </span>supercontig:ASM2366v1:CP001665:801353:802561:-1<span class="w"> </span>gene:ECBD_0775<span class="w"> </span>transcript:ACT27843<span class="w"> </span>gene_biotype:protein_coding<span class="w"> </span>transcript_biotype:protein_coding<span class="w"> </span>description:transposase<span class="w"> </span>IS4<span class="w"> </span>family<span class="w"> </span>protein<span class="w">       </span>MCELDILHDSLYQFCPELHLKRLNSLTLACHALLDCKTLTLTELGRNLPTKARTKHNIKRIDRLLGNRHLHKERLAVYRWHASFICSGNTMPIVLVDWSDIREQKRLMVLRASVALHGRSVTLYEKAFPLSEQCSKKAHDQFLADLASILPSNTTPLIVSDAGFKVPWYKSVEKLGWYWLSRVRGKVQYADLGAENWKPISNLHDMSSSHSKTLGYKRLTKSNPISCQILLYKSRSKGRKNQRSTRTHCHHPSPKIYSASAKEPWVLATNLPVEIRTPKQLVNIYSKRMQIEETFRDLKSPAYGLGLRHSRTSSSERFDIMLLIALMLQLTCWLAGVHAQKQGWDKHFQANTVRNRNVLSTVRLGMEVLRHSGYTITREDLLVAATLLAQNLFTHGYALGKL
&gt;ACT28724<span class="w"> </span>pep<span class="w"> </span>supercontig:ASM2366v1:CP001665:1791810:1793018:-1<span class="w"> </span>gene:ECBD_1670<span class="w"> </span>transcript:ACT28724<span class="w"> </span>gene_biotype:protein_coding<span class="w"> </span>transcript_biotype:protein_coding<span class="w"> </span>description:transposase<span class="w"> </span>IS4<span class="w"> </span>family<span class="w"> </span>protein<span class="w">     </span>MCELDILHDSLYQFCPELHLKRLNSLTLACHALLDCKTLTLTELGRNLPTKARTKHNIKRIDRLLGNRHLHKERLAVYRWHASFICSGNTMPIVLVDWSDIREQKRLMVLRASVALHGRSVTLYEKAFPLSEQCSKKAHDQFLADLASILPSNTTPLIVSDAGFKVPWYKSVEKLGWYWLSRVRGKVQYADLGAENWKPISNLHDMSSSHSKTLGYKRLTKSNPISCQILLYKSRSKGRKNQRSTRTHCHHPSPKIYSASAKEPWVLATNLPVEIRTPKQLVNIYSKRMQIEETFRDLKSPAYGLGLRHSRTSSSERFDIMLLIALMLQLTCWLAGVHAQKQGWDKHFQANTVRNRNVLSTVRLGMEVLRHSGYTITREDLLVAATLLAQNLFTHGYALGKL
&gt;ACT28727<span class="w"> </span>pep<span class="w"> </span>supercontig:ASM2366v1:CP001665:1795281:1796489:-1<span class="w"> </span>gene:ECBD_1674<span class="w"> </span>transcript:ACT28727<span class="w"> </span>gene_biotype:protein_coding<span class="w"> </span>transcript_biotype:protein_coding<span class="w"> </span>description:transposase<span class="w"> </span>IS4<span class="w"> </span>family<span class="w"> </span>protein<span class="w">     </span>MCELDILHDSLYQFCPELHLKRLNSLTLACHALLDCKTLTLTELGRNLPTKARTKHNIKRIDRLLGNRHLHKERLAVYRWHASFICSGNTMPIVLVDWSDIREQKRLMVLRASVALHGRSVTLYEKAFPLSEQCSKKAHDQFLADLASILPSNTTPLIVSDAGFKVPWYKSVEKLGWYWLSRVRGKVQYADLGAENWKPISNLHDMSSSHSKTLGYKRLTKSNPISCQILLYKSRSKGRKNQRSTRTHCHHPSPKIYSASAKEPWILATNLPVEIRTPKQLVNIYSKRMQIEETFRDLKSPAYGLGLRHSRTSSSERFDIMLLIALMLQLTCWLAGVHAQKQGWDKHFQANTVRNRNVLSTVRLGMEVLRHSGYTITREDLLVAATLLAQNLFTHGYALGKL
&gt;ACT30468<span class="w"> </span>pep<span class="w"> </span>supercontig:ASM2366v1:CP001665:3610918:3612126:-1<span class="w"> </span>gene:ECBD_3469<span class="w"> </span>transcript:ACT30468<span class="w"> </span>gene_biotype:protein_coding<span class="w"> </span>transcript_biotype:protein_coding<span class="w"> </span>description:transposase<span class="w"> </span>IS4<span class="w"> </span>family<span class="w"> </span>protein<span class="w">     </span>MCELDILHDSLYQFCPELHLKRLNSLTLACHALLDCKTLTLTELGRNLPTKARTKHNIKRIDRLLGNRHLHKERLAVYRWHASFICSGNTMPIVLVDWSDIREQKRLMVLRASVALHGRSVTLYEKAFPLSEQCSKKAHDQFLADLASILPSNTTPLIVSDAGFKVPWYKSVEKLGWYWLSRVRGKVQYADLGAENWKPISNLHDMSSSHSKTLGYKRLTKSNPISCQILLYKSRSKGRKNQRSTRTHCHHPSPKIYSASAKEPWILATNLPVEIRTPKQLVNIYSKRMQIEETFRDLKSPAYGLGLRHSRTSSSERFDIMLLIALMLQLTCWLAGVHAQKQGWDKHFQANTVRNRNVLSTVRLGMEVLRHSGYTITREDSLVAATLLTQNLFTHGYVLGKL
&gt;ACT31152<span class="w"> </span>pep<span class="w"> </span>supercontig:ASM2366v1:CP001665:4391870:4393078:-1<span class="w"> </span>gene:ECBD_4170<span class="w"> </span>transcript:ACT31152<span class="w"> </span>gene_biotype:protein_coding<span class="w"> </span>transcript_biotype:protein_coding<span class="w"> </span>description:transposase<span class="w"> </span>IS4<span class="w"> </span>family<span class="w"> </span>protein<span class="w">     </span>MCELDILHDSLYQFCPELHLKRLNSLTLACHALLDCKTLTLTELGRNLPTKARTKHNIKRIDRLLGNRHLHKERLAVYRWHASFICSGNTMPIVLVDWSDIREQKRLMVLRASVALHGRSVTLYEKAFPLSEQCSKKAHDQFLADLASILPSNTTPLIVSDAGFKVPWYKSVEKLGWYWLSRVRGKVQYADLGAENWKPISNLHDMSSSHSKTLGYKRLTKSNPISCQILLYKSRSKGRKNQRSTRTHCHHPSPKIYSASAKEPWILATNLPVEIRTPKQLVNIYSKRMQIEETFRDLKSPAYGLGLRHSRTSSSERFDIMLLIALMLQLTCWLAGVHAQKQGWDKHFQANTVRNRNVLSTVRLGMEVLRHSGYTITREDLLVAATLLAQNLFTHGYALGKL



awk<span class="w"> </span>-F<span class="s2">&quot;\t&quot;</span><span class="w"> </span><span class="s1">&#39;{if ($2~&quot;YFAKRLK$&quot;) print}&#39;</span><span class="w"> </span>proteins.tab

&gt;ACT27943<span class="w"> </span>pep<span class="w"> </span>supercontig:ASM2366v1:CP001665:911724:912023:-1<span class="w"> </span>gene:ECBD_0875<span class="w"> </span>transcript:ACT27943<span class="w"> </span>gene_biotype:protein_coding<span class="w"> </span>transcript_biotype:protein_coding<span class="w"> </span>description:transposase<span class="w"> </span>IS3/IS911<span class="w"> </span>family<span class="w"> </span>protein<span class="w"> </span>MTKTVSTSKKPRKQHSPEFRSEALKLAERIGVTAAARELSLYESQLYNWRSKQQNQQTSSERELEMSTEIARLKRQLAERDEELAILQKAATYFAKRLK
&gt;ACT29129<span class="w"> </span>pep<span class="w"> </span>supercontig:ASM2366v1:CP001665:2195088:2195387:-1<span class="w"> </span>gene:ECBD_2089<span class="w"> </span>transcript:ACT29129<span class="w"> </span>gene_biotype:protein_coding<span class="w"> </span>transcript_biotype:protein_coding<span class="w"> </span>description:transposase<span class="w"> </span>IS3/IS911<span class="w"> </span>family<span class="w"> </span>protein<span class="w">       </span>MTKTVSTSKKPRKQHSPEFRSEALKLAERIGVTAAARELSLYESQLYNWRSKQQNQQTSSERELEMSTEIARLKRQLAERDEELAILQKAATYFAKRLK
&gt;ACT29266<span class="w"> </span>pep<span class="w"> </span>supercontig:ASM2366v1:CP001665:2355429:2355728:-1<span class="w"> </span>gene:ECBD_2236<span class="w"> </span>transcript:ACT29266<span class="w"> </span>gene_biotype:protein_coding<span class="w"> </span>transcript_biotype:protein_coding<span class="w"> </span>description:transposase<span class="w"> </span>IS3/IS911<span class="w"> </span>family<span class="w"> </span>protein<span class="w">       </span>MTKTVSTSKKPRKQHSPEFRSEALKLAERIGVTAAARELSLYESQLYNWRSKQQNQQTSSERELEMSTEIARLKRQLAERDEELAILQKAATYFAKRLK
&gt;ACT29589<span class="w"> </span>pep<span class="w"> </span>supercontig:ASM2366v1:CP001665:2682665:2682964:1<span class="w"> </span>gene:ECBD_2567<span class="w"> </span>transcript:ACT29589<span class="w"> </span>gene_biotype:protein_coding<span class="w"> </span>transcript_biotype:protein_coding<span class="w"> </span>description:transposase<span class="w"> </span>IS3/IS911<span class="w"> </span>family<span class="w"> </span>protein<span class="w">        </span>MTKTVSTSKKPRKQHSPEFRSEALKLAERIGVTAAARELSLYESQLYNWRSKQQNQQTSSERELEMSTEIARLKRQLAERDEELAILQKAATYFAKRLK
</pre></div>
</div>
<p>Other useful pattern for the regular expression is the dot <strong>.</strong> that means any charachter.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>grep<span class="w"> </span>-v<span class="w"> </span><span class="s2">&quot;&gt;&quot;</span><span class="w"> </span>Escherichia_coli_bl21_gold_de3_plyss_ag_.ASM2366v1.pep.all.fa<span class="w"> </span><span class="p">|</span><span class="w"> </span>grep<span class="w"> </span><span class="s2">&quot;AAAAA.A&quot;</span>

MKLSRRSFMKANAV**AAAAAAA**GLSVPGVARAVVGQQEAIKWDKAPCRFCGTGCGVLVGTQ
EQQRRMEAERLAQMQQLSHQDDDS**AAAAALAA**QTGERKVGRNDPCPCGSGKKYKQCHGRL
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>grep<span class="w"> </span>-v<span class="w"> </span><span class="s2">&quot;&gt;&quot;</span><span class="w"> </span>Escherichia_coli_bl21_gold_de3_plyss_ag_.ASM2366v1.pep.all.fa<span class="w"> </span><span class="p">|</span><span class="w"> </span>grep<span class="w"> </span><span class="s2">&quot;A[GA]AAA.AA&quot;</span>

GE**AGAAAPAA**KQEAAPAAAPAPAAGVKEVNVPDIGGDEVEVTEVMVKVGDKVAAEQSLIT
EQQRRMEAERLAQMQQLSHQDDDS**AAAAALAA**QTGERKVGRNDPCPCGSGKKYKQCHGRL
</pre></div>
</div>
<p>More extended regular expression can be searched by using grep with the parameter <strong>-E</strong>. For instance we can search for repetition:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>grep<span class="w"> </span>-v<span class="w"> </span><span class="s2">&quot;&gt;&quot;</span><span class="w"> </span>Escherichia_coli_bl21_gold_de3_plyss_ag_.ASM2366v1.pep.all.fa<span class="w"> </span><span class="p">|</span><span class="w"> </span>grep<span class="w"> </span>-E<span class="w"> </span><span class="s2">&quot;(AT){3}&quot;</span>

GVGIGIETVDGVPVKINNNSGATFVLSDGSNTLLFNTWVQAKSGRDVTLGNFT**ATATAT**F
</pre></div>
</div>
</section>
<section id="id3">
<h3>Recap<a class="headerlink" href="#id3" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p><strong>uniq</strong> it removes the duplicated elements in a list</p></li>
<li><p><strong>sort</strong> it sorts a given list</p></li>
<li><p><strong>shuf</strong> it shuffles a given list</p></li>
</ul>
</section>
<section id="exercises">
<h3>Exercises<a class="headerlink" href="#exercises" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>We see that some of the protein sequences are repeated in <strong>proteins.tab</strong> file. So different genes produce the same protein. How many unique proteins do we have in our proteins.tab file?</p></li>
<li><p>How many sequences in SRR6466185_1.fastq.gz contain the pattern “GGGATGACGGC”? And how many in SRR6466185_2.fastq.gz?</p></li>
<li><p>Can you calculate the sum of the size of the first 10 sequences in <strong>proteins.tab</strong>?</p></li>
<li><p>Can you tell how many different kind of description there are in <strong>seq_names.txt</strong>?</p></li>
<li><p>In Ensembl the chromosomes are named differently than in UCSC (1,2,3… vs chr1, chr2, chr3). Can you convert the binding sites stored within <strong>GSE41589_Suz12_BindingSites.txt.gz</strong> in a way that is compatible with Ensembl?</p></li>
</ul>
</section>
</section>
<section id="space-in-volumes-and-permissions">
<h2>Space in volumes and permissions<a class="headerlink" href="#space-in-volumes-and-permissions" title="Permalink to this heading"></a></h2>
<section id="volume-sizes-and-disk-space">
<h3>Volume sizes and disk space<a class="headerlink" href="#volume-sizes-and-disk-space" title="Permalink to this heading"></a></h3>
<p>When we deal with analyzing valuable data we should consider different problems:
* Consuming too much disk space
* Consuming too much memory
* Corrupting / deleting files when utilizing all disk space</p>
<p>Check the size of a file with :
* <strong>ls -lh</strong>:</p>
<blockquote>
<div><ul class="simple">
<li><p><strong>l</strong>: list.</p></li>
<li><p><strong>h</strong>: human readable.</p></li>
</ul>
</div></blockquote>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>ls<span class="w"> </span>-lh<span class="w"> </span>Escherichia_coli_bl21_gold_de3_plyss_ag_.ASM2366v1.pep.all.fa
</pre></div>
</div>
<dl class="simple">
<dt>Get a summary of the disk usage size of a directory:</dt><dd><ul class="simple">
<li><p><strong>du -sh</strong>:</p></li>
<li><p><strong>du</strong> stands for Disk Usage.</p></li>
<li><p><strong>s</strong>: summarize.</p></li>
<li><p><strong>h</strong>: human readable format.</p></li>
</ul>
</dd>
</dl>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
<p>du -sh my_beautiful_folder/</p>
<blockquote>
<div><p>3.5K    my_beautiful_folder/</p>
</div></blockquote>
<dl class="simple">
<dt>Display disk usage of all the files and directories:</dt><dd><ul class="simple">
<li><p><strong>du -ah</strong>:</p></li>
<li><p><strong>a</strong>: all.</p></li>
</ul>
</dd>
</dl>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>du<span class="w"> </span>-ah<span class="w"> </span>my_beautiful_folder/
</pre></div>
</div>
<ul class="simple">
<li><p>Reduce the space by:
+ Compressing files.
+ Using programs such as <strong>zcat</strong> or <strong>gunzip -c</strong> to extract information of zipped files on the fly instead of extracting the data.
+ Creating <strong>symbolic links</strong> instead of copying files and folders.</p></li>
</ul>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># compress files using gzip:</span>
gzip<span class="w"> </span>Escherichia_coli_bl21_gold_de3_plyss_ag_.ASM2366v1.pep.all.fa

<span class="c1"># show the content of a gzipped file without uncompressing it:</span>
zcat<span class="w"> </span>SRR6466185_1.fastq.gz<span class="w"> </span><span class="p">|</span><span class="w"> </span>head
gunzip<span class="w"> </span>-c<span class="w"> </span>SRR6466185_1.fastq.gz<span class="w"> </span><span class="p">|</span><span class="w"> </span>head

<span class="c1"># link files instead of copying them:</span>

mkdir<span class="w"> </span>test_links
<span class="nb">cd</span><span class="w"> </span>test_links

<span class="c1"># ln for linking, -s for symbolic link: first comes the file to link, then where to link it!</span>
ln<span class="w"> </span>-s<span class="w"> </span>../SRR6466185_1.fastq.gz<span class="w"> </span>.
<span class="c1"># give the copy of the file another name:</span>
ln<span class="w"> </span>-s<span class="w"> </span>../SRR6466185_2.fastq.gz<span class="w"> </span>./sample1_read2.fastq.gz

<span class="nb">cd</span><span class="w"> </span>..
</pre></div>
</div>
<p>Show the system disk space statistics (file system disk space usage):
* <strong>df -h</strong>.</p>
<blockquote>
<div><ul class="simple">
<li><p><strong>df</strong> stands for: <a href="#id4"><span class="problematic" id="id5">**</span></a>D**isk <a href="#id6"><span class="problematic" id="id7">**</span></a>F**ilesystem.</p></li>
<li><p><strong>h</strong>: human readable.</p></li>
</ul>
</div></blockquote>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>df<span class="w"> </span>-h

Filesystem<span class="w">                              </span>Size<span class="w">  </span>Used<span class="w"> </span>Avail<span class="w"> </span>Use%<span class="w"> </span>Mounted<span class="w"> </span>on
/dev/sda3<span class="w">                                </span>20G<span class="w">  </span><span class="m">8</span>.1G<span class="w">   </span>11G<span class="w">  </span><span class="m">44</span>%<span class="w"> </span>/
devtmpfs<span class="w">                                 </span>63G<span class="w">     </span><span class="m">0</span><span class="w">   </span>63G<span class="w">   </span><span class="m">0</span>%<span class="w"> </span>/dev
tmpfs<span class="w">                                    </span>63G<span class="w">     </span><span class="m">0</span><span class="w">   </span>63G<span class="w">   </span><span class="m">0</span>%<span class="w"> </span>/dev/shm
...
</pre></div>
</div>
</section>
<section id="permissions">
<h3>Permissions<a class="headerlink" href="#permissions" title="Permalink to this heading"></a></h3>
<p>Create a small file:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">echo</span><span class="w"> </span><span class="s2">&quot;my file&quot;</span><span class="w"> </span>&gt;<span class="w"> </span>test.txt
</pre></div>
</div>
<ul class="simple">
<li><p>Each file has particular permissions that restrict their access to the users.</p></li>
</ul>
<p><strong>ls -l</strong> shows those permissions:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>ls<span class="w"> </span>-l<span class="w"> </span>test.txt
-rw-r--r--<span class="w"> </span><span class="m">1</span><span class="w"> </span>lcozzuto<span class="w"> </span>Bioinformatics_Unit<span class="w"> </span><span class="m">5</span><span class="w"> </span>Mar<span class="w"> </span><span class="m">14</span><span class="w"> </span><span class="m">16</span>:29<span class="w"> </span>test.txt
</pre></div>
</div>
<p>Here is the owner is <strong>lcozzuto</strong> and the group it belongs to is <strong>Bioinformatics_Unit</strong>.</p>
<p>The first field here contains <strong>10 ticks</strong>:</p>
<ul class="simple">
<li><dl class="simple">
<dt>tick 1 (1 field):</dt><dd><ul>
<li><p><strong>d</strong>: directory</p></li>
<li><p><strong>-</strong>: regular file</p></li>
<li><p><strong>l</strong>: symbolic link</p></li>
</ul>
</dd>
</dl>
</li>
<li><p>ticks 2-4: permissions of the <strong>owner</strong> (3 fields)</p></li>
<li><p>ticks 5-7: permissions of the <strong>group</strong> (3 fields)</p></li>
<li><p>ticks 8-10: permissions of <strong>any other user</strong> (3 fields)</p></li>
</ul>
<ul class="simple">
<li><dl class="simple">
<dt>What kind of permissions are we talking about?</dt><dd><ul>
<li><p><strong>r</strong>: read</p></li>
<li><p><strong>w</strong>: write</p></li>
<li><p><strong>x</strong>: execute</p></li>
</ul>
</dd>
</dl>
</li>
<li><p>In the latter example:
+ <strong>lcozzuto</strong> can <strong>read</strong> and <strong>write</strong> the file, but <strong>NOT execute</strong> it.
+ Members of <strong>Bioinformatics_Unit</strong> can only <strong>read</strong> the file.
+ All other users can only <strong>read</strong> the file.</p></li>
<li><p><strong>chmod</strong> controls the changes of permissions:</p></li>
</ul>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>chmod<span class="w"> </span><span class="o">[</span>who<span class="o">][</span>+,-,<span class="o">=][</span>permissions<span class="o">]</span><span class="w"> </span>filename
</pre></div>
</div>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>User</p></th>
<th class="head"><p>letter</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>owner</p></td>
<td><p>u</p></td>
</tr>
<tr class="row-odd"><td><p>group</p></td>
<td><p>g</p></td>
</tr>
<tr class="row-even"><td><p>users
not in
the
group</p></td>
<td><p>o</p></td>
</tr>
<tr class="row-odd"><td><p>all
users</p></td>
<td><p>a</p></td>
</tr>
</tbody>
</table>
<ul class="simple">
<li><p>Add <strong>writing</strong> permissions to the <strong>group</strong>:</p></li>
</ul>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>chmod<span class="w"> </span>g+w<span class="w"> </span>test.txt

ls<span class="w"> </span>-l<span class="w"> </span>test.txt

-rw-rw-r--<span class="w"> </span><span class="m">1</span><span class="w"> </span>sbonnin<span class="w"> </span>Bioinformatics_Unit<span class="w"> </span><span class="m">5</span><span class="w"> </span>Mar<span class="w"> </span><span class="m">14</span><span class="w"> </span><span class="m">16</span>:29<span class="w"> </span>test.txt
</pre></div>
</div>
<ul class="simple">
<li><p>Add <strong>writing</strong> permissions to the <strong>all other users</strong>:</p></li>
</ul>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>chmod<span class="w"> </span>o+w<span class="w"> </span>test.txt

ls<span class="w"> </span>-l<span class="w"> </span>test.txt

-rw-rw-rw-<span class="w"> </span><span class="m">1</span><span class="w"> </span>sbonnin<span class="w"> </span>Bioinformatics_Unit<span class="w"> </span><span class="m">5</span><span class="w"> </span>Mar<span class="w"> </span><span class="m">14</span><span class="w"> </span><span class="m">16</span>:29<span class="w"> </span>test.txt
</pre></div>
</div>
<ul class="simple">
<li><p>Remove writing permissions to all <strong>but the owner</strong>:</p></li>
</ul>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>chmod<span class="w"> </span>og-w<span class="w"> </span>test.txt

ls<span class="w"> </span>-l<span class="w"> </span>test.txt

-rw-r--r--<span class="w"> </span><span class="m">1</span><span class="w"> </span>lcozzuto<span class="w"> </span>Bioinformatics_Unit<span class="w"> </span><span class="m">5</span><span class="w"> </span>Mar<span class="w"> </span><span class="m">14</span><span class="w"> </span><span class="m">16</span>:29<span class="w"> </span>test.txt
</pre></div>
</div>
<ul class="simple">
<li><p>Remove all permissions to all <strong>but the owner</strong>:</p></li>
</ul>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>chmod<span class="w"> </span>og-rw<span class="w"> </span>test.txt

-rw-------<span class="w"> </span><span class="m">1</span><span class="w"> </span>lcozzuto<span class="w"> </span>Bioinformatics_Unit<span class="w"> </span><span class="m">5</span><span class="w"> </span>Mar<span class="w"> </span><span class="m">14</span><span class="w"> </span><span class="m">16</span>:29<span class="w"> </span>test.txt
</pre></div>
</div>
<ul class="simple">
<li><p>Preserve the file from any modification <strong>even done by yourself</strong> using <strong>a</strong> (= all):</p></li>
</ul>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>chmod<span class="w"> </span>a-w<span class="w"> </span>test.txt

-r--------<span class="w"> </span><span class="m">1</span><span class="w"> </span>lcozzuto<span class="w"> </span>Bioinformatics_Unit<span class="w"> </span><span class="m">5</span><span class="w"> </span>Mar<span class="w"> </span><span class="m">14</span><span class="w"> </span><span class="m">16</span>:29<span class="w"> </span>test.txt

<span class="c1"># Try to overwrite the content of test.txt:</span>

<span class="nb">echo</span><span class="w"> </span><span class="s2">&quot;test&quot;</span><span class="w"> </span>&gt;<span class="w"> </span>test.txt
-bash:<span class="w"> </span>test.txt:<span class="w"> </span>Permission<span class="w"> </span>denied

<span class="c1"># Try to remove test.txt:</span>

rm<span class="w"> </span>test.txt
rm:<span class="w"> </span>remove<span class="w"> </span>write-protected<span class="w"> </span>regular<span class="w"> </span>file<span class="w"> </span>‘test.txt’?
</pre></div>
</div>
<ul class="simple">
<li><p>Control whether a file or folder is <strong>executable</strong>:</p></li>
</ul>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>chmod<span class="w"> </span>-x<span class="w"> </span>my_ugly_folder/

<span class="c1"># try to enter the folder:</span>
<span class="nb">cd</span><span class="w"> </span>my_ugly_folder/
-bash:<span class="w"> </span>cd:<span class="w"> </span>my_ugly_folder/:<span class="w"> </span>Permission<span class="w"> </span>denied
</pre></div>
</div>
<ul class="simple">
<li><p>A user that don’t have the <strong>executing</strong> rights can’t access the directory !</p></li>
<li><p>Apply permission changes <strong>recursively</strong>, i.e. to all files inside a directory:</p></li>
</ul>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>chmod<span class="w"> </span>-R<span class="w"> </span>+x<span class="w"> </span>my_ugly_folder/
</pre></div>
</div>
<section id="use-octal-notation-for-file-permissions">
<h4>Use octal notation for file permissions<a class="headerlink" href="#use-octal-notation-for-file-permissions" title="Permalink to this heading"></a></h4>
<ul class="simple">
<li><dl class="simple">
<dt>Each tick in the first field refers to:</dt><dd><ul>
<li><p>a type of permission: read, write, execute.</p></li>
<li><p>a user type: owner, group, all others.</p></li>
</ul>
</dd>
</dl>
</li>
</ul>
<p>Each tick can be replaced by <strong>0</strong> (does not have that permission) or <strong>1</strong> (has that permission): this creates a <strong>binary</strong> number at each <strong>user type</strong>, that can be converted into an <strong>octal</strong> number.&lt;br&gt;</p>
<p>Hence, each <strong>octal</strong> number represents a set of permissions:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Binary</p></th>
<th class="head"><p>Octal</p></th>
<th class="head"><p>Permission</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>000</p></td>
<td><p>0</p></td>
<td><ul class="simple">
<li><ul>
<li><ul>
<li></li>
</ul>
</li>
</ul>
</li>
</ul>
</td>
</tr>
<tr class="row-odd"><td><p>001</p></td>
<td><p>1</p></td>
<td><ul class="simple">
<li><ul>
<li><p>x</p></li>
</ul>
</li>
</ul>
</td>
</tr>
<tr class="row-even"><td><p>010</p></td>
<td><p>2</p></td>
<td><ul class="simple">
<li><p>w -</p></li>
</ul>
</td>
</tr>
<tr class="row-odd"><td><p>011</p></td>
<td><p>3</p></td>
<td><ul class="simple">
<li><p>w x</p></li>
</ul>
</td>
</tr>
<tr class="row-even"><td><p>100</p></td>
<td><p>4</p></td>
<td><p>r - -</p></td>
</tr>
<tr class="row-odd"><td><p>101</p></td>
<td><p>5</p></td>
<td><p>r - x</p></td>
</tr>
<tr class="row-even"><td><p>110</p></td>
<td><p>6</p></td>
<td><p>r w -</p></td>
</tr>
<tr class="row-odd"><td><p>111</p></td>
<td><p>7</p></td>
<td><p>r w x</p></td>
</tr>
</tbody>
</table>
<ul class="simple">
<li><p>Set the permissions of <strong>my_expression.txt</strong> (from Module 1) so that:
+ the owner can: read, write and execute.
+ the group can: read and write.
+ the other users don’t have any permission.</p></li>
</ul>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>chmod<span class="w"> </span><span class="m">760</span><span class="w"> </span>my_expression.txt
<span class="c1"># 7 for the owner</span>
<span class="c1"># 6 for the group</span>
<span class="c1"># 0 for other users</span>
</pre></div>
</div>
</section>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="about.html" class="btn btn-neutral float-left" title="About the course" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="containers.html" class="btn btn-neutral float-right" title="Containers" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, CRG.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>