<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Master of Pores 2 &mdash; ROPES summer school - CRG - May 2023  documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
    <link rel="canonical" href="https://biocorecrg.github.io/ropes-linux-mop2-2023/mop2_practical.html" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="author" title="About these documents" href="about.html" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Nextflow" href="nextflow.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            ROPES summer school - CRG - May 2023
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="about.html">About the course</a></li>
<li class="toctree-l1"><a class="reference internal" href="linux_basics.html">Introduction to Linux</a></li>
<li class="toctree-l1"><a class="reference internal" href="containers.html">Containers</a></li>
<li class="toctree-l1"><a class="reference internal" href="nextflow.html">Nextflow</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Master of Pores 2</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#basics-of-direct-rna-sequencing-drs">Basics of Direct RNA Sequencing (DRS)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#master-of-pores-2-mop2-preprocessing-and-polya-tail-estimation-1">Master Of Pores 2 (MOP2) - preprocessing and polyA tail estimation (1)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#basic-preprocessing-module-mop-preprocessing">Basic preprocessing (module: <em>mop_preprocessing</em>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#polya-tail-length-analysis-module-mop-tail">PolyA tail length analysis (module: <em>mop_tail</em>)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#hands-on-1-mop-preprocess-and-mop-tail">Hands-on 1: <em>mop_preprocess</em> and <em>mop_tail</em></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#mop2-installation-and-data-preprocessing">MOP2 installation and data preprocessing:</a></li>
<li class="toctree-l3"><a class="reference internal" href="#computational-resources">Computational resources</a></li>
<li class="toctree-l3"><a class="reference internal" href="#monitoring">Monitoring</a></li>
<li class="toctree-l3"><a class="reference internal" href="#results">Results</a></li>
<li class="toctree-l3"><a class="reference internal" href="#polya-tail-length-estimation">PolyA tail length estimation</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id1">Results</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#master-of-pores-2-mop2-rna-modification-detection-2">Master Of Pores 2 (MOP2) - RNA modification detection (2)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#detection-of-rna-modifications-module-mop-mod">Detection of RNA modifications (module: <em>mop_mod</em>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#consensus-detection-of-rna-modifications-module-mop-consensus">Consensus detection of RNA modifications (module: <em>mop_consensus</em>)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#hands-on-2-mop-mod-and-mop-consensus">Hands-on 2: <em>mop_mod</em> and <em>mop_consensus</em></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#detection-of-differentially-modified-sites">Detection of differentially modified sites</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id3">Results</a></li>
<li class="toctree-l3"><a class="reference internal" href="#detection-of-differentially-modified-sites-with-high-confidence">Detection of differentially modified sites with high confidence</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id4">Results</a></li>
</ul>
</li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">ROPES summer school - CRG - May 2023</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Master of Pores 2</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/mop2_practical.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="master-of-pores-2">
<span id="mop2-practical-page"></span><h1>Master of Pores 2<a class="headerlink" href="#master-of-pores-2" title="Permalink to this heading"></a></h1>
<section id="basics-of-direct-rna-sequencing-drs">
<h2>Basics of Direct RNA Sequencing (DRS)<a class="headerlink" href="#basics-of-direct-rna-sequencing-drs" title="Permalink to this heading"></a></h2>
<p>Direct RNA sequencing is a technology developed by Oxford Nanopore Technologies (ONT) that allows the sequencing of native RNA molecules without the need of previous amplification nor fragmentation. Therefore, it can provide information about the presence of RNA modifications, polyA tail length and composition at per-read level.</p>
<p>This method is based on the use of protein nanopores that are embedded into a membrane. As the molecule goes through the pore, it alters the ionic current that is applied to it. These changes are then stored in <strong>fast5 files</strong>.</p>
<a class="reference internal image-reference" href="_images/Pore.png"><img alt="_images/Pore.png" src="_images/Pore.png" style="width: 700px;" /></a>
</section>
<section id="master-of-pores-2-mop2-preprocessing-and-polya-tail-estimation-1">
<h2>Master Of Pores 2 (MOP2) - preprocessing and polyA tail estimation (1)<a class="headerlink" href="#master-of-pores-2-mop2-preprocessing-and-polya-tail-estimation-1" title="Permalink to this heading"></a></h2>
<p>The complexity of analysing current intensity data together with the lack of systematic and reproducible pipelines have hindered the access of this technology to the general users. To overcome this limitation, we developed Master Of Pores (<strong>MoP</strong>) - a nextflow based workflow that simplifies the analysis of DRS datasets and aims to make it accessible to non-bioinformatic experts. We further improved this suit making a new version (<strong>MoP2</strong>).</p>
<p>For more information you can read our pubblications:</p>
<ul class="simple">
<li><p><a class="reference external" href="https://www.frontiersin.org/articles/10.3389/fgene.2020.00211/full">“MasterOfPores: A Workflow for the Analysis of Oxford Nanopore Direct RNA Sequencing Datasets” Cozzuto L, Liu H, Pryszcz LP, Hermoso Pulido T,  Delgado-Tejedor A, Ponomarenko J and Novoa EM. Front. Genet., 17 March 2020.</a></p></li>
<li><p><a class="reference external" href="https://link.springer.com/protocol/10.1007/978-1-0716-2962-8_13">“Nanopore Direct RNA Sequencing Data Processing and Analysis Using MasterOfPores” Cozzuto L, Delgado-Tejedor A, Hermoso Pulido T, Novoa EM, Ponomarenko J. N. Methods Mol Biol. 2023;2624:185-205.</a></p></li>
</ul>
<p>MOP2 can perform all steps required to analyse DRS data - from converting raw current intensities into multiple types of processed data to RNA modified sites detection and polyA tail length predictions. This pipeline consists of four modules: <em>mop_preprocess</em>, <em>mop_tail</em>, <em>mop_mod</em> and <em>mop_consensus</em>.</p>
<a class="reference internal image-reference" href="_images/MOP_scheme.png"><img alt="_images/MOP_scheme.png" src="_images/MOP_scheme.png" style="width: 700px;" /></a>
<section id="basic-preprocessing-module-mop-preprocessing">
<h3>Basic preprocessing (module: <em>mop_preprocessing</em>)<a class="headerlink" href="#basic-preprocessing-module-mop-preprocessing" title="Permalink to this heading"></a></h3>
<p>The pre-processing module is able to perform basecalling, mapping (either to a genome or a transcriptome), feature counting (per-gene level) and quality control. Furthermore, if required by the user, it can run demultiplexing, filtering and discovery of novel transcripts. As the final step, the workflow generates a final report of the performance and results of each of the steps performed.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Before proceeding to any other downstream analysis such as RNA modification detection and polyA tail analysis, this module <strong>must</strong> be executed.</p>
</div>
<p><strong>Analysis overview:</strong></p>
<ul>
<li><p><strong>Step 1: Demultiplexing</strong></p>
<p>Demultiplexing is required when analysing a barcoded sample; otherwise, this step is not necessary. <strong>Deeplexicon</strong> is used when analysing dRNA. This algorithm converts the barcode’s signal into an image, which is then classified based on a machine-learning approach. For cDNA, DNA and metDNA you can specify within the parameters to use <strong>Guppy</strong>.</p>
<ul class="simple">
<li><p><strong>Input:</strong> Raw fast5 files</p></li>
<li><p><strong>Output:</strong> Demuxed raw fast5 files</p></li>
</ul>
</li>
<li><p><strong>Step 2: Basecalling</strong></p>
<p>It is the process by which the current intensity data is translated into a nucleotide sequence by a machine learning algorithm called basecaller. Currently, the most widely used is <strong>Guppy</strong>, which was developed by ONT and it is only available if you are part of the ONT community. The model that Guppy uses to analyse RNA data is not modification aware and therefore, it can only identify the four canonical bases (A, U, C and G).</p>
<ul class="simple">
<li><p><strong>Input:</strong> Raw fast5 files</p></li>
<li><p><strong>Output:</strong> Basecalled fast5 and fastq files</p></li>
</ul>
</li>
<li><p><strong>Step 3: Filtering</strong></p>
<p>Filter out reads based on either quality and/or length performed by <strong>Nanofilt</strong>. For RNA modification detection using DRS data, this step should be turned off as modified reads tend to have lower quality than unmodified ones and thus, filtering based on quality would bias the results.</p>
<ul class="simple">
<li><p><strong>Input:</strong> Fastq files</p></li>
<li><p><strong>Output:</strong> Filtered fastq files</p></li>
</ul>
</li>
<li><p><strong>Step 4: Alignment</strong></p>
<p>Mapping step performed by either <strong>minimap2</strong> or <strong>grapmap</strong>. Both can perform spliced or unspliced alignments. Briefly, we would use spliced alignments when using a genome as a reference and; unspliced for transcriptome. Furthermore, it has been reported that minimap2 fails to align highly modified reads and thus, it should not be used to analyse data from highly modified RNA species such as rRNAs.</p>
<ul class="simple">
<li><p><strong>Input:</strong> Fastq files and reference file (genome or transcriptome)</p></li>
<li><p><strong>Output:</strong> Bam (and bai) files</p></li>
</ul>
</li>
<li><p><strong>Step 5: Gene/transcript counts</strong></p>
<p>The software run by MOP2 to perform this step depends on the type of reference used in the mapping step. For transcriptome alignments, <strong>NanoCount</strong> is used and it reports per transcript abundances whereas for genome alignments, <strong>htseq-count</strong> is executed and it generates per-gene counts.</p>
<ul class="simple">
<li><p><strong>Input:</strong> Reference and alignment file (bam) for NanoCount // Reference, annotation (<em>.gtf) and alignment files (</em>.bam) for htseq-count</p></li>
<li><p><strong>Output:</strong> Transcript abundances’ estimations // Per-gene counts</p></li>
</ul>
</li>
<li><p><strong>Step 5: Transcript discovery</strong></p>
<p><strong>Bambu</strong> aims to identify novel transcripts from mapped reads. For more information about how to use this tool, please visit its <a class="reference external" href="https://github.com/GoekeLab/bambu#General-Usage">GitHub page</a>.</p>
<ul class="simple">
<li><p><strong>Input:</strong> Alignment (.bam), reference (.fa) and annotation (.gtf)</p></li>
<li><p><strong>Output:</strong> Transcript’s abundances and read id-transcript assignments</p></li>
</ul>
</li>
<li><p><strong>Step 6: Reporting and quality control</strong></p>
<p><strong>multiQC</strong> produces the final report, as a html page, which contains the quality control’s results generated by <strong>MinionQC</strong> together with the stats from the previous executed steps.</p>
<ul class="simple">
<li><p><strong>Input:</strong> all inputs and outputs</p></li>
<li><p><strong>Output:</strong> final report</p></li>
</ul>
</li>
</ul>
<p>We will show how to use and configurate this module in the next hands-on exercise.</p>
</section>
<section id="polya-tail-length-analysis-module-mop-tail">
<h3>PolyA tail length analysis (module: <em>mop_tail</em>)<a class="headerlink" href="#polya-tail-length-analysis-module-mop-tail" title="Permalink to this heading"></a></h3>
<p>This module estimates poly(A) tail length at read level provided by <strong>Tailfindr</strong> and/or <strong>Nanopolish</strong>. This workflow uses as input all the files generated by <em>mop_preprocess</em>.</p>
<p><strong>Analysis overview:</strong></p>
<ul class="simple">
<li><p><strong>Software 1: Tailfindr</strong></p>
<ul>
<li><p><strong>Input:</strong> Basecalled fast5 files and read id-gene assignments</p></li>
<li><p><strong>Output:</strong> PolyA tail length estimations</p></li>
</ul>
</li>
<li><p><strong>Software 2: Nanopolish</strong></p>
<ul>
<li><p><strong>Input:</strong> Fastq (.fq.gz), alignment (.bam) and reference (.fa) files</p></li>
<li><p><strong>Output:</strong> PolyA tail length estimations</p></li>
</ul>
</li>
</ul>
</section>
</section>
<section id="hands-on-1-mop-preprocess-and-mop-tail">
<h2>Hands-on 1: <em>mop_preprocess</em> and <em>mop_tail</em><a class="headerlink" href="#hands-on-1-mop-preprocess-and-mop-tail" title="Permalink to this heading"></a></h2>
<section id="mop2-installation-and-data-preprocessing">
<h3>MOP2 installation and data preprocessing:<a class="headerlink" href="#mop2-installation-and-data-preprocessing" title="Permalink to this heading"></a></h3>
<p>For installing the MoP2 pipeline and downloading guppy 3.4.2, please use the code below:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">git clone --depth 1 --recurse-submodules https://github.com/biocorecrg/MOP2.git</span>

<span class="go">cd MOP2; bash INSTALL.sh 3.4.5</span>
</pre></div>
</div>
<p>For this hands-on exercise, we will perform polyA tail length estimation and RNA modification detection on total RNA DRS samples from <em>Saccharomyces cerevisiae</em> (see list below):</p>
<ul class="simple">
<li><p>Sample 1: snR36 knock-out strain</p></li>
<li><p>Samples 2, 3 and 4: WT strains</p></li>
</ul>
<p>We need to downolad the test dataset that is bundled in this repository</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">mkdir mydata</span>
<span class="go">cd mydata</span>

<span class="go">wget https://biocorecrg.github.io/ropes-linux-mop2-2023/data/nanopore.tar.gz</span>

<span class="go">--2023-05-10 15:48:30--  https://biocorecrg.github.io/ropes-linux-mop2-2023/data/nanopore.tar.gz</span>
<span class="go">Resolving biocorecrg.github.io (biocorecrg.github.io)... 185.199.108.153, 185.199.111.153, 185.199.110.153, ...</span>
<span class="go">Connecting to biocorecrg.github.io (biocorecrg.github.io)|185.199.108.153|:443... connected.</span>
<span class="go">HTTP request sent, awaiting response... 200 OK</span>
<span class="go">Length: 31740190 (30M) [application/gzip]</span>
<span class="go">Saving to: ‘nanopore.tar.gz’</span>

<span class="go">100%[=======================================================================&gt;] 31,740,190   123MB/s   in 0.2s</span>

<span class="go">2023-05-10 15:48:35 (123 MB/s) - ‘nanopore.tar.gz’ saved [31740190/31740190]</span>

<span class="go">ls -alh nanopore.tar.gz</span>
<span class="go">-rw-r--r-- 1 lcozzuto Bioinformatics_Unit 31M May 10 15:45 nanopore.tar.gz</span>

<span class="go">tar -zvxf nanopore.tar.gz</span>
<span class="go">nanopore/</span>
<span class="go">nanopore/WT100_Cov50_Rep1/</span>
<span class="go">nanopore/WT100_Cov100_Rep1/</span>
<span class="go">nanopore/Saccharomyces_cerevisiae.rRNA.fa</span>
<span class="go">nanopore/snR36_KO/</span>
<span class="go">nanopore/WT50_Cov50_Rep1/</span>
<span class="go">nanopore/WT50_Cov50_Rep1/batch_unmodified0.fast5</span>
<span class="go">nanopore/WT50_Cov50_Rep1/batch_modified0.fast5</span>
<span class="go">nanopore/snR36_KO/batch0.fast5</span>
<span class="go">nanopore/WT100_Cov100_Rep1/batch_modified0.fast5</span>
<span class="go">nanopore/WT100_Cov50_Rep1/batch_modified0.fast5</span>
</pre></div>
</div>
<p>Before setting up <em>mop_preproceess</em> module, it is important that you think about which softwares and parameters should be used - otherwise you might run analysis that are not suitable to your sample (and you will lose time and resources). Please, answers the questions below:</p>
<ul class="simple">
<li><p><strong>Question 1:</strong> Which is the most abundant RNA specie in your samples? Is it highly or lowly modified? Which mapper would you use?</p></li>
<li><p><strong>Question 2:</strong> Which reference would you use (genome or transcriptome)?</p></li>
<li><p><strong>Question 3:</strong> Would you use spliced or unspliced alignment?</p></li>
<li><p><strong>Question 4:</strong> Which counter would you use?</p></li>
</ul>
<p>Now, we can start setting up the <em>mop_preproceess</em> module. Please follow the code below:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">#</span>Enter<span class="w"> </span>the<span class="w"> </span>mop_preprocess<span class="w"> </span>directory:
<span class="go">cd ../mop_preprocess</span>

<span class="gp">#</span>List<span class="w"> </span>all<span class="w"> </span>files<span class="w"> </span>and<span class="w"> </span>directories:
<span class="go">ls -l</span>

<span class="gp">#</span>Summary<span class="w"> </span>of<span class="w"> </span>files:
<span class="gp">#</span><span class="c1"># bin directory: it contains all the binaries used by this module. If you wanna change guppy version, you should go here.</span>
<span class="gp">#</span><span class="c1"># *_opt.tsv files: it is used to input additional parameters to the individual softwares executed by the workflow.</span>
<span class="gp">#</span><span class="c1"># params.config file: it is the file that the user must edit to introduce the inputs required by the workflow.</span>

<span class="gp">#</span>Create<span class="w"> </span>a<span class="w"> </span>new<span class="w"> </span>params<span class="w"> </span>file<span class="w"> </span>and<span class="w"> </span>link<span class="w"> </span>it<span class="w"> </span>to<span class="w"> </span>the<span class="w"> </span>original<span class="w"> </span>one:
<span class="go">cp params.config myparams.config</span>
<span class="go">rm params.config</span>
<span class="go">ln -s myparams.config params.config</span>

<span class="gp"># </span>edit<span class="w"> </span>the<span class="w"> </span>new<span class="w"> </span>file
<span class="go">nano params.config</span>

<span class="gp">#</span>Params.config<span class="w"> </span>content:
<span class="go">params {</span>
<span class="go">  conffile            = &quot;final_summary_01.txt&quot;</span>
<span class="go">  fast5               = &quot;${projectDir}/../mydata/nanopore/**/*.fast5&quot;</span>
<span class="go">  fastq               = &quot;&quot;</span>

<span class="go">  reference           = &quot;${projectDir}/../mydata/nanopore/Saccharomyces_cerevisiae.rRNA.fa&quot;</span>
<span class="go">  annotation          = &quot;&quot;</span>
<span class="go">  ref_type            = &quot;transcriptome&quot;</span>

<span class="go">  pars_tools          = &quot;drna_tool_unsplice_opt.tsv&quot;</span>
<span class="go">  output              = &quot;$baseDir/output&quot;</span>
<span class="go">  qualityqc           = 5</span>
<span class="go">  granularity         = 1</span>

<span class="go">  basecalling         = &quot;guppy&quot;</span>
<span class="go">  GPU                 = &quot;OFF&quot;</span>
<span class="go">  demultiplexing      = &quot;NO&quot;</span>
<span class="go">  demulti_fast5       = &quot;NO&quot;</span>

<span class="go">  filtering           = &quot;NO&quot;</span>

<span class="go">  mapping             = &quot;graphmap&quot;</span>
<span class="go">  counting            = &quot;nanocount&quot;</span>
<span class="go">  discovery           = &quot;NO&quot;</span>

<span class="go">  cram_conv           = &quot;NO&quot;</span>
<span class="go">  subsampling_cram    = 50</span>

<span class="go">  saveSpace           = &quot;NO&quot;</span>

<span class="go">  email               = &quot;username@domain&quot;</span>
<span class="go">}</span>

<span class="gp">#</span>Save<span class="w"> </span>file<span class="w"> </span>and<span class="w"> </span>exit:
<span class="go">CTRL+o</span>
<span class="go">CTRL+x</span>
</pre></div>
</div>
<p>As discussed earlier, these options are okay when analysing total RNA samples. However, depending on the type of sample, changes in the params.config file should be made. Click <a class="reference external" href="https://biocorecrg.github.io/MOP2/docs/mop_preprocess.html">here</a> to check all parameters accepted by <em>mop_preprocess</em>.</p>
<p>Now that the set up of the module is done, it is important to discuss two additional topics before running it:</p>
</section>
<section id="computational-resources">
<h3>Computational resources<a class="headerlink" href="#computational-resources" title="Permalink to this heading"></a></h3>
<p><strong>MoP2</strong> has different profiles with resources specified for several infrastructures. If you have a look at the folder <strong>conf</strong> you can have an idea of the possibility to fine tune the resources such as maximum execution time, the queue name, the maximum memory etc. Fo example let’s have a look at the <strong>local.config file</strong>, we can change it to use more processors and memory, since our workstations have 8 CPUs and 16 Gb of RAM Memory.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">cd ../conf</span>

<span class="go">vim local.config</span>

<span class="go">process {</span>
<span class="go">        executor = &#39;local&#39;</span>
<span class="go">        cpus = 3</span>
<span class="go">        memory = &#39;6GB&#39;</span>
<span class="go">    cache=&#39;lenient&#39;</span>
<span class="go">    container = &#39;biocorecrg/mopprepr:0.7&#39;</span>
<span class="go">    containerOptions = { workflow.containerEngine == &quot;docker&quot; ? &#39;-u $(id -u):$(i</span>
<span class="go">d -g)&#39;: null}</span>
<span class="go">    withLabel: big_cpus_ignore {</span>
<span class="go">        errorStrategy = &#39;ignore&#39;</span>

<span class="go">    }</span>
<span class="go">    withLabel: basecall_gpus {</span>
<span class="go">            maxForks = 1</span>
<span class="go">            containerOptions = { workflow.containerEngine == &quot;singularity&quot; ? &#39;--</span>
<span class="go">nv&#39;:</span>
<span class="go">                   ( workflow.containerEngine == &quot;docker&quot; ? &#39;-u $(id -u):$(id -g</span>
<span class="go">) --gpus all&#39;: null ) }</span>
<span class="go">    }</span>
<span class="go">}</span>
</pre></div>
</div>
</section>
<section id="monitoring">
<h3>Monitoring<a class="headerlink" href="#monitoring" title="Permalink to this heading"></a></h3>
<p><strong>Nextflow Tower</strong> is an open source monitoring and managing platform for Nextflow workflows. There are two versions:</p>
<ul class="simple">
<li><p>Open source for monitoring of single pipelines.</p></li>
<li><p>Commercial one for workflow management, monitoring and resource optimisation.</p></li>
</ul>
<p>We will show the open source one.</p>
<p>First, you need to access the <a class="reference external" href="https://tower.nf/">tower.nf</a> website and login.</p>
<a class="reference internal image-reference" href="_images/tower.png"><img alt="_images/tower.png" src="_images/tower.png" style="width: 800px;" /></a>
<p>We recommend you to use github or google for autentication. If you selected the email for receiving the instructions and the token to be used.</p>
<a class="reference internal image-reference" href="_images/tower0.png"><img alt="_images/tower0.png" src="_images/tower0.png" style="width: 800px;" /></a>
<p>Here you see there is not yet any information.</p>
<a class="reference internal image-reference" href="_images/tower2.png"><img alt="_images/tower2.png" src="_images/tower2.png" style="width: 800px;" /></a>
<p>You can generate your token at <a class="reference external" href="https://tower.nf/tokens">https://tower.nf/tokens</a> and copy paste it in your pipeline using this snippet in the configuration file:</p>
<div class="highlight-groovy notranslate"><div class="highlight"><pre><span></span><span class="n">tower</span><span class="w"> </span><span class="o">{</span>
<span class="w">  </span><span class="n">accessToken</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">&#39;&lt;YOUR TOKEN&gt;&#39;</span>
<span class="w">  </span><span class="n">enabled</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">true</span>
<span class="o">}</span>
</pre></div>
</div>
<p>or exporting those environmental variables:</p>
<div class="highlight-groovy notranslate"><div class="highlight"><pre><span></span><span class="n">export</span><span class="w"> </span><span class="n">TOWER_ACCESS_TOKEN</span><span class="o">=*******</span><span class="n">YOUR</span><span class="o">***</span><span class="n">TOKEN</span><span class="o">*****</span><span class="n">HERE</span><span class="o">*******</span>
</pre></div>
</div>
<p>Now we can launch the pipeline:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">#</span>Run<span class="w"> </span>the<span class="w"> </span>module<span class="w"> </span><span class="k">in</span><span class="w"> </span>the<span class="w"> </span>background,<span class="w"> </span>with<span class="w"> </span>docker<span class="w"> </span>and<span class="w"> </span><span class="k">in</span><span class="w"> </span>the<span class="w"> </span><span class="nb">local</span><span class="w"> </span>computer:

<span class="go">nextflow run mop_preprocess.nf -with-docker -bg -profile local -with-tower &gt; log_preprocess.txt</span>

<span class="go">more log_preprocess.txt</span>
<span class="go">N E X T F L O W  ~  version 20.10.0</span>
<span class="go">Launching `mop_preprocess.nf` [goofy_spence] - revision: ec40fe0af4</span>

<span class="go">╔╦╗╔═╗╔═╗  ╔═╗┬─┐┌─┐┌─┐┬─┐┌─┐┌─┐┌─┐┌─┐┌─┐</span>
<span class="go">║║║║ ║╠═╝  ╠═╝├┬┘├┤ ├─┘├┬┘│ ││  ├┤ └─┐└─┐</span>
<span class="go">╩ ╩╚═╝╩    ╩  ┴└─└─┘┴  ┴└─└─┘└─┘└─┘└─┘└─┘</span>

<span class="go">====================================================</span>
<span class="go">BIOCORE@CRG Master of Pores 2. Preprocessing - N F  ~  version 2.0</span>
<span class="go">====================================================</span>

<span class="go">conffile.                 : final_summary_01.txt</span>

<span class="go">fast5                     : /nfs/no_backup/enovoa/users/andelgado/MOP2/mop_preprocess/../mydata/nanopore/**/*.fast5</span>
<span class="go">fastq                     :</span>

<span class="go">reference                 : /nfs/no_backup/enovoa/users/andelgado/MOP2/mop_preprocess/../mydata/nanopore/Saccharomyces_cerevisiae.rRNA.fa</span>
<span class="go">annotation                :</span>

<span class="go">granularity.              : 1</span>

<span class="go">ref_type                  : transcriptome</span>
<span class="go">pars_tools                : drna_tool_unsplice_opt.tsv</span>

<span class="go">output                    : /nfs/no_backup/enovoa/users/andelgado/MOP2/mop_preprocess/output</span>

<span class="go">GPU                       : OFF</span>

<span class="go">basecalling               : guppy</span>
<span class="go">demultiplexing            : NO</span>
<span class="go">demulti_fast5             : NO</span>

<span class="go">filtering                 : NO</span>
<span class="go">mapping                   : graphmap</span>

<span class="go">counting                  : nanocount</span>
<span class="go">discovery                 : NO</span>

<span class="go">cram_conv                 : NO</span>
<span class="go">subsampling_cram          : 50</span>


<span class="go">saveSpace                 : NO</span>
<span class="go">email                     : username@domain</span>

<span class="go">Sending the email to username@domain</span>

<span class="go">----------------------CHECK TOOLS -----------------------------</span>
<span class="go">basecalling : guppy</span>
<span class="go">&gt; demultiplexing will be skipped</span>
<span class="go">mapping : graphmap</span>
<span class="go">&gt; filtering will be skipped</span>
<span class="go">counting : nanocount</span>
<span class="go">&gt; discovery will be skipped</span>
<span class="go">--------------------------------------------------------------</span>
<span class="go">Monitor the execution with Nextflow Tower using this url https://tower.nf/user/anna-delgado-tejedor/watch/5MePdfTHV9xGyK</span>
</pre></div>
</div>
<p>Now, you can go to the tower website again using the link in the log:</p>
<a class="reference internal image-reference" href="_images/tower_mop.png"><img alt="_images/tower_mop.png" src="_images/tower_mop.png" style="width: 800px;" /></a>
<p>When the pipeline is finished we can also receive a mail.</p>
<p>Here you can see an example of the resource report that is generated at the end:</p>
<a class="reference internal image-reference" href="_images/tower_mop2.png"><img alt="_images/tower_mop2.png" src="_images/tower_mop2.png" style="width: 800px;" /></a>
</section>
<section id="results">
<h3>Results<a class="headerlink" href="#results" title="Permalink to this heading"></a></h3>
<p>Once the module has finished, these directories should be in your output folder:</p>
<ul class="simple">
<li><p><strong>fast5_files</strong>: Contains the basecalled fast5 files.</p></li>
</ul>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p><strong>How do we know if fast5 files are bassecalled or not?</strong></p>
<p>Raw and basecall fast5 files have the same extension (.fast5) and in consequence, the only way of knowing if a fast5 file is basecalled or not is to check its contents. Please use the code below:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span>#Install h5ls as sudo user:
sudo apt-get install hdf5-tools

[sudo] password for training:
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following additional packages will be installed:
libaec0 libhdf5-103-1 libhdf5-hl-100 libsz2
[....]

#Investigate fast5 files&#39; structure:
h5ls /path/to/fast5 | head -n15
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/raw.png"><img alt="_images/raw.png" src="_images/raw.png" style="width: 700px;" /></a>
<a class="reference internal image-reference" href="_images/basecalled.png"><img alt="_images/basecalled.png" src="_images/basecalled.png" style="width: 700px;" /></a>
</div>
<ul class="simple">
<li><p><strong>fastq_files</strong>: Contains one or, in case of demultiplexing, more fastq files.</p></li>
<li><p><strong>QC_files</strong>: Contains each single QC produced by the pipeline.</p></li>
<li><p><strong>alignment</strong>: Contains the bam and bai file(s).</p></li>
<li><p><strong>counts</strong>: Contains read counts per gene / transcript.</p></li>
<li><p><strong>assigned</strong>: Contains assignment of each read to a given gene / transcript.</p></li>
<li><p><strong>report</strong>: Contains the final multiqc report.</p></li>
</ul>
<p>Now, we would look at the alignments in IGV (genome browser) together with the stats reported in the multiQC html to decide if we have enough quality data to proceed with the polyA tail length estimation and RNA modification detection analysis. Due to time limitations, here you should decide if we can proceed or not only based on the multiQC report.</p>
<ul class="simple">
<li><p><strong>Question 5:</strong> Do we have enough data in all samples to proceed to the downstream analysis?</p></li>
</ul>
<a class="reference internal image-reference" href="_images/report1.png"><img alt="_images/report1.png" src="_images/report1.png" style="width: 700px;" /></a>
<a class="reference internal image-reference" href="_images/report2.png"><img alt="_images/report2.png" src="_images/report2.png" style="width: 700px;" /></a>
</section>
<section id="polya-tail-length-estimation">
<h3>PolyA tail length estimation<a class="headerlink" href="#polya-tail-length-estimation" title="Permalink to this heading"></a></h3>
<p>After preprocessing the data, we can go directly to run the <em>mop_tail</em> module which will output polyA tail length estimation at per read level. Please run the code below:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">#</span>Go<span class="w"> </span>to<span class="w"> </span>the<span class="w"> </span>directory:
<span class="go">cd ../mop_tail/</span>

<span class="gp">#</span>Edit<span class="w"> </span>params.config<span class="w"> </span>file:
<span class="go">nano params.config</span>

<span class="gp">#</span>Params.config<span class="w"> </span>content:

<span class="go">params {</span>

<span class="go">  input_path         = &quot;$projectDir/../mop_preprocess/output/&quot;</span>
<span class="go">  reference          = &quot;$projectDir/../mydata/nanopore/Saccharomyces_cerevisiae.rRNA.fa&quot;</span>

<span class="go">  pars_tools         = &quot;$projectDir/tools_opt.tsv&quot;</span>

<span class="go">  output             = &quot;$projectDir/outputPoly&quot;</span>

<span class="go">  tailfindr          = &quot;YES&quot;</span>
<span class="go">  nanopolish         = &quot;YES&quot;</span>

<span class="go">  email              = &quot;username@domain&quot;</span>
<span class="go">}</span>

<span class="gp">#</span>Save<span class="w"> </span>file<span class="w"> </span>and<span class="w"> </span>exit:
<span class="go">CTRL+o</span>
<span class="go">CTRL+x</span>

<span class="gp">#</span>Run<span class="w"> </span>the<span class="w"> </span>module<span class="w"> </span><span class="k">in</span><span class="w"> </span>the<span class="w"> </span>background,<span class="w"> </span>with<span class="w"> </span>docker<span class="w"> </span>and<span class="w"> </span><span class="k">in</span><span class="w"> </span>the<span class="w"> </span><span class="nb">local</span><span class="w"> </span>computer:
<span class="go">nextflow run mop_tail.nf -with-docker -bg -profile local -with-tower &gt; log_tail.txt</span>
</pre></div>
</div>
</section>
<section id="id1">
<h3>Results<a class="headerlink" href="#id1" title="Permalink to this heading"></a></h3>
<p>Once the module has finished, these directories should be in your output folder:</p>
<ul class="simple">
<li><p><strong>nanopolish_flow</strong>: Contains nanopolish’s results.</p></li>
<li><p><strong>tailfindr_flow</strong>: Contains tailfindr’s results.</p></li>
<li><p><strong>polya_common</strong>: Contains the text files that include the combined polyA tail length predictions at per read-level.</p></li>
</ul>
<p>Check the generated files and answer these questions below:</p>
<ul class="simple">
<li><p><strong>Question 6:</strong> Should we have done this analysis? Why?</p></li>
</ul>
</section>
</section>
<section id="master-of-pores-2-mop2-rna-modification-detection-2">
<h2>Master Of Pores 2 (MOP2) - RNA modification detection (2)<a class="headerlink" href="#master-of-pores-2-mop2-rna-modification-detection-2" title="Permalink to this heading"></a></h2>
<p>Currently, due to the lack of RNA modification-aware basecallers, the most common methods to identify RNA modifications from DRS datasets rely on finding differences between two differentially modified samples (ie: comparing a wild-type sample against a knock-out/IVT one). All publicly available algortihms can be clustered into two groups based on the feature that they use to generate their predictions:</p>
<a class="reference internal image-reference" href="_images/Detection_methods.png"><img alt="_images/Detection_methods.png" src="_images/Detection_methods.png" style="width: 700px;" /></a>
<section id="detection-of-rna-modifications-module-mop-mod">
<h3>Detection of RNA modifications (module: <em>mop_mod</em>)<a class="headerlink" href="#detection-of-rna-modifications-module-mop-mod" title="Permalink to this heading"></a></h3>
<p>Using the files obtained from the preprocessing analysis as input, the module <em>mop_mod</em> from MOP2 runs four publicly available algorithms (<strong>EpiNano</strong>, <strong>Nanopolish</strong>, <strong>Tombo</strong> and <strong>Nanocompore</strong>) to detect RNA modifications from DRS data. These softwares use several features from the data to generate their predictions (see below):</p>
<ul class="simple">
<li><p><strong>EpiNano</strong>: difference in basecalling errors (mismatch, insertion and deletion frequency). The sum of all these errors can also be used.</p></li>
<li><p><strong>Nanopolish</strong>: difference in median current intensity per position - although it was not designed to identify RNA modifications.</p></li>
<li><p><strong>Tombo</strong>: difference in current intensity values’ distribution (KS test).</p></li>
<li><p><strong>Nanocompore</strong>: difference in clusters based on current intensity and dwell time values.</p></li>
</ul>
<p>When using these algorithms, it is important to be aware of their limitations (and when applicable, take them into account to interpret our results!):</p>
<ul class="simple">
<li><p>Requirement of a KO/KD/IVT sample</p></li>
<li><p>Signal across the 5mer differs between softwares</p></li>
<li><p>Performance depends on coverage, modification type and stoichiometry</p></li>
<li><p>Presence of false positives in their predictions</p></li>
</ul>
<a class="reference internal image-reference" href="_images/Limitations_CurrentMethods.png"><img alt="_images/Limitations_CurrentMethods.png" src="_images/Limitations_CurrentMethods.png" style="width: 700px;" /></a>
<p><strong>Analysis overview:</strong></p>
<ul class="simple">
<li><p><strong>Software 1: EpiNano</strong></p>
<ul>
<li><p><strong>Input:</strong> Alignment (.bam) and reference (.fa) files</p></li>
<li><p><strong>Output:</strong> Basecalling errors (mismatch, insertion, deletion frequencies), coverage and base quality per-position.</p></li>
</ul>
</li>
<li><p><strong>Software 2: Nanopolish</strong></p>
<ul>
<li><p><strong>Input:</strong> Fastq (.fq.gz), alignment (.bam), reference (.fa) and summary (_final_summary.stats) files</p></li>
<li><p><strong>Output:</strong> Median current intensity values per position as well as how many <em>resquiggled</em> reads do cover that position.</p></li>
</ul>
</li>
</ul>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p><strong>What is resquiggling?</strong></p>
<p><em>Resquiggling</em> is the process by which basecalled reads are mapped to a genome/transcriptome reference and then, the raw signal is assigned to a sequence context in the reference. <strong>Nanopolish</strong>, <strong>Tombo</strong> and <strong>Nanocompore</strong> require all input reads to be resquiggled. Currently, the two most commonly used resquiggling algorithms are the ones from <strong>Nanopolish</strong> and <strong>Tombo</strong>.</p>
<p><strong>IMPORTANT:</strong> Some reads might fail to be resquiggled and thus, won’t be included in any downstream processes!</p>
</div>
<ul class="simple">
<li><p><strong>Software 3: Tombo</strong></p>
<ul>
<li><p><strong>Input:</strong> Fastq (.fq.gz), alignment (.bam), reference (.fa) and summary (_final_summary.stats) files</p></li>
<li><p><strong>Output:</strong> -Log10(p-value) obtained from the KS-test when comparing current intensity values’ distribution as well as how many <em>resquiggled</em> reads do cover that position (per position).</p></li>
</ul>
</li>
<li><p><strong>Software 4: Nanocompore</strong></p>
<ul>
<li><p><strong>Input:</strong> Fastq (.fq.gz), alignment (.bam), reference (.fa) and summary (_final_summary.stats) files</p></li>
<li><p><strong>Output:</strong> P-values obtained when comparing clusters of reads from the two samples (per-position).</p></li>
</ul>
</li>
</ul>
<p>We will show how to use and configurate this module in the next hands-on exercise.</p>
</section>
<section id="consensus-detection-of-rna-modifications-module-mop-consensus">
<h3>Consensus detection of RNA modifications (module: <em>mop_consensus</em>)<a class="headerlink" href="#consensus-detection-of-rna-modifications-module-mop-consensus" title="Permalink to this heading"></a></h3>
<p>As we have previously seen, predictions generated by the four algorithms run by <em>mop_mod</em> show poor overlap due to different factors (see previous section). Based on this evidence, we reasoned that if the results from these algorithms would be combined in a consensual manner, the true differentially modified sites would be retained while the number of false positives would decrease. To this end, we developed <strong>NanoConsensus</strong>, which is run by the module <em>mop_consensus</em> inside MOP2.</p>
<p><strong>Analysis overview:</strong></p>
<a class="reference internal image-reference" href="_images/NanoConsensus.png"><img alt="_images/NanoConsensus.png" src="_images/NanoConsensus.png" style="width: 700px;" /></a>
<ul class="simple">
<li><p><strong>Input:</strong> Predictions from EpiNano, Nanopolish, Tombo and NanoCompore (at least two of them).</p></li>
<li><p><strong>Output:</strong> Consensus prediction of differentially modified sites.</p></li>
</ul>
<p>For more information about <strong>NanoConsensus</strong> and the benchamrking results obtained when comparing its performance to other publicly available tools - please click <a class="reference external" href="https://www.biorxiv.org/content/10.1101/2023.03.21.533606v1">here</a>.</p>
</section>
</section>
<section id="hands-on-2-mop-mod-and-mop-consensus">
<h2>Hands-on 2: <em>mop_mod</em> and <em>mop_consensus</em><a class="headerlink" href="#hands-on-2-mop-mod-and-mop-consensus" title="Permalink to this heading"></a></h2>
<section id="detection-of-differentially-modified-sites">
<h3>Detection of differentially modified sites<a class="headerlink" href="#detection-of-differentially-modified-sites" title="Permalink to this heading"></a></h3>
<p>After preprocessing the data, we can run the <em>mop_mod</em> module which runs four algorithms to identify differentially modified sites. Please run the code below:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">#</span>Go<span class="w"> </span>to<span class="w"> </span>the<span class="w"> </span>directory:
<span class="go">cd ./../mop_mod/</span>

<span class="gp">#</span>Summary<span class="w"> </span>of<span class="w"> </span>files:
<span class="gp">#</span><span class="c1"># comparison.tsv file: it is used to input the pairwise comparisons that the workflow should analyse.</span>
<span class="gp">#</span><span class="c1"># params.config file: it is the file that the user must edit to introduce the inputs required by the workflow.</span>

<span class="gp">#</span>Edit<span class="w"> </span>params.config<span class="w"> </span>file:
<span class="go">nano params.config</span>

<span class="gp">#</span>Params.config<span class="w"> </span>content:
<span class="go">params {</span>
<span class="go">  input_path         = &quot;$projectDir/../mop_preprocess/output/&quot;</span>
<span class="go">  comparison         = &quot;$projectDir/comparison.tsv&quot;</span>

<span class="go">  reference          = &quot;$projectDir/../mydata/nanopore/Saccharomyces_cerevisiae.rRNA.fa&quot;</span>

<span class="go">  output             = &quot;$projectDir/output_mod&quot;</span>

<span class="go">  pars_tools         = &quot;$projectDir/tools_opt.tsv&quot;</span>

<span class="go">  // flows</span>
<span class="go">  epinano       = &quot;YES&quot;</span>
<span class="go">  nanocompore   = &quot;YES&quot;</span>
<span class="go">  tombo_lsc     = &quot;YES&quot;</span>
<span class="go">  tombo_msc     = &quot;YES&quot;</span>

<span class="go">  // epinano plots</span>
<span class="go">  epinano_plots = &quot;YES&quot;</span>

<span class="go">  email              = &quot;username@domain&quot;</span>
<span class="go">}</span>

<span class="gp">#</span>Save<span class="w"> </span>file<span class="w"> </span>and<span class="w"> </span>exit:
<span class="go">CTRL+o</span>
<span class="go">CTRL+x</span>

<span class="gp">#</span>Edit<span class="w"> </span>comparison.tsv<span class="w"> </span>file:
<span class="go">nano comparison.tsv</span>

<span class="gp">#</span>Comparison.tsv<span class="w"> </span>content<span class="w"> </span>-<span class="w"> </span>IT<span class="w"> </span>MUST<span class="w"> </span>BE<span class="w"> </span>TAB<span class="w"> </span>SEPARATED!
<span class="go">snR36_KO        WT100_Cov100_Rep1</span>
<span class="go">snR36_KO        WT100_Cov50_Rep1</span>
<span class="go">snR36_KO        WT50_Cov50_Rep1</span>

<span class="gp">#</span>Save<span class="w"> </span>file<span class="w"> </span>and<span class="w"> </span>exit:
<span class="go">CTRL+o</span>
<span class="go">CTRL+x</span>

<span class="gp">#</span>Run<span class="w"> </span>the<span class="w"> </span>module<span class="w"> </span><span class="k">in</span><span class="w"> </span>the<span class="w"> </span>background,<span class="w"> </span>with<span class="w"> </span>docker<span class="w"> </span>and<span class="w"> </span><span class="k">in</span><span class="w"> </span>the<span class="w"> </span><span class="nb">local</span><span class="w"> </span>computer:
<span class="go">nextflow run mop_mod.nf -with-docker -bg -profile local -with-tower &gt; log_mod.txt</span>
</pre></div>
</div>
</section>
<section id="id3">
<h3>Results<a class="headerlink" href="#id3" title="Permalink to this heading"></a></h3>
<p>Once the module has finished, these directories should be in your output folder:</p>
<ul class="simple">
<li><p><strong>epinano_flow</strong>: Contains EpiNano’s results (one .csv.gz file per sample) and the plots comparing each of the basecalilng features (mismatch, insertion and deletion frequency) between two samples per transcript.</p></li>
<li><p><strong>nanopolish-compore_flow</strong>: Contains Nanopolish’s (one .csv.gz file per sample) and Nanocompore’s (one directory per comparison) results.</p></li>
<li><p><strong>tombo_flow</strong>: Contains Tombo’s results (one .tsv.gz file per comparison)</p></li>
</ul>
<p>Check the generated files and answer these questions below:</p>
<ul class="simple">
<li><p><strong>Question 7:</strong> Were all expected files generated? If not, which one(s) are missing? Could you hypothesize why?</p></li>
</ul>
<p>To fix this issue, please run the code below:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">#</span>Edit<span class="w"> </span>the<span class="w"> </span>tools_opt.tsv<span class="w"> </span>file:
<span class="go">nano tools_opt.tsv</span>

<span class="gp">#</span>Contents<span class="w"> </span>of<span class="w"> </span>the<span class="w"> </span>tools_opt.tsv<span class="w"> </span>file:
<span class="gp">#</span>flows<span class="w">  </span>tool<span class="w">    </span>extrapars
<span class="go">epinano epinano &quot;&quot;</span>
<span class="go">nanocompore     nanopolish      &quot;&quot;</span>
<span class="go">nanocompore     nanocompore     &quot;--sequence_context 2 --downsample_high_coverage 10000&quot;</span>
<span class="go">tombo_resquiggling      tombo   &quot;&quot;</span>
<span class="go">tombo_msc       tombo   &quot;&quot;</span>
<span class="go">tombo_lsc       tombo   &quot;--minimum-test-reads 30&quot;</span>

<span class="gp">#</span>Save<span class="w"> </span>file<span class="w"> </span>and<span class="w"> </span>exit:
<span class="go">CTRL+o</span>
<span class="go">CTRL+x</span>

<span class="gp">#</span>Re-run<span class="w"> </span>the<span class="w"> </span>module<span class="w"> </span><span class="k">in</span><span class="w"> </span>the<span class="w"> </span>background,<span class="w"> </span>with<span class="w"> </span>docker<span class="w"> </span>and<span class="w"> </span><span class="k">in</span><span class="w"> </span>the<span class="w"> </span><span class="nb">local</span><span class="w"> </span>computer:
<span class="go">nextflow run mop_mod.nf -with-docker -bg -profile local -with-tower -resume &gt; log_mod_resumed.txt</span>
</pre></div>
</div>
<ul class="simple">
<li><p><strong>Question 8:</strong> Check the output from tombo and nanopolish - could you explain why the coverage reported is different?</p></li>
</ul>
</section>
<section id="detection-of-differentially-modified-sites-with-high-confidence">
<h3>Detection of differentially modified sites with high confidence<a class="headerlink" href="#detection-of-differentially-modified-sites-with-high-confidence" title="Permalink to this heading"></a></h3>
<p>Once we have obtained the predictions from the four algorithms run by <em>mop_mod</em>, we can run <strong>NanoConsensus</strong> using the <em>mop_consensus</em> module. Please, run the code below:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">#</span>Go<span class="w"> </span>to<span class="w"> </span>the<span class="w"> </span>directory:
<span class="go">cd ./../mop_consensus/</span>

<span class="gp">#</span>Summary<span class="w"> </span>of<span class="w"> </span>files:
<span class="gp">#</span><span class="c1"># comparison.tsv file: it is used to input the pairwise comparisons that the workflow should analyse (it must be the same as the one used when running mop_mod)</span>
<span class="gp">#</span><span class="c1"># params.config file: it is the file that the user must edit to introduce the inputs required by the workflow.</span>

<span class="gp">#</span>Edit<span class="w"> </span>params.config<span class="w"> </span>file:
<span class="go">nano params.config</span>

<span class="gp">#</span>Params.config<span class="w"> </span>content:
<span class="go">params {</span>

<span class="go">  input_path         = &quot;$projectDir/../mop_mod/output_mod&quot;</span>
<span class="go">  reference          = &quot;$projectDir/../mydata/nanopore/Saccharomyces_cerevisiae.rRNA.fa&quot;</span>

<span class="go">  comparison         = &quot;$projectDir/comparison.tsv&quot;</span>
<span class="go">  padsize            = 50</span>
<span class="go">  output             = &quot;$projectDir/output_consensus&quot;</span>

<span class="go">  email              = &quot;username@domain&quot;</span>
<span class="go">}</span>

<span class="gp">#</span>Save<span class="w"> </span>file<span class="w"> </span>and<span class="w"> </span>exit:
<span class="go">CTRL+o</span>
<span class="go">CTRL+x</span>

<span class="gp">#</span>For<span class="w"> </span>the<span class="w"> </span>comparison.tsv<span class="w"> </span>file,<span class="w"> </span>you<span class="w"> </span>have<span class="w"> </span>two<span class="w"> </span>options:
<span class="gp">#</span><span class="c1">#OPTION 1: Copy the comparison file from the mop_mod directory:</span>
<span class="go">cp ./../mop_mod/comparison.tsv .</span>

<span class="gp">#</span><span class="c1">#OPTION 2:Edit comparison.tsv file:</span>
<span class="go">nano comparison.tsv</span>

<span class="gp">#</span>Comparison.tsv<span class="w"> </span>content<span class="w"> </span>-<span class="w"> </span>IT<span class="w"> </span>MUST<span class="w"> </span>BE<span class="w"> </span>TAB<span class="w"> </span>SEPARATED!
<span class="go">snR36_KO        WT100_Cov100_Rep1</span>
<span class="go">snR36_KO        WT100_Cov50_Rep1</span>
<span class="go">snR36_KO        WT50_Cov50_Rep1</span>

<span class="gp">#</span>Save<span class="w"> </span>file<span class="w"> </span>and<span class="w"> </span>exit:
<span class="go">CTRL+o</span>
<span class="go">CTRL+x</span>

<span class="gp">#</span>Run<span class="w"> </span>the<span class="w"> </span>module<span class="w"> </span><span class="k">in</span><span class="w"> </span>the<span class="w"> </span>background,<span class="w"> </span>with<span class="w"> </span>docker<span class="w"> </span>and<span class="w"> </span><span class="k">in</span><span class="w"> </span>the<span class="w"> </span><span class="nb">local</span><span class="w"> </span>computer:
<span class="go">nextflow run mop_consensus.nf -with-docker -bg -profile local -with-tower &gt; log_consensus.txt</span>
</pre></div>
</div>
</section>
<section id="id4">
<h3>Results<a class="headerlink" href="#id4" title="Permalink to this heading"></a></h3>
<p>Once the module has finished, a directory per comparison and transcript should be generated. In this case, the three directories below should be generated:</p>
<ul class="simple">
<li><p>snR36_KO—WT100_Cov100_Rep1-18s</p></li>
<li><p>snR36_KO—WT100_Cov50_Rep1-18s</p></li>
<li><p>snR36_KO—WT50_Cov50_Rep1-18s</p></li>
<li><p><strong>Question 9:</strong> Inspect the log file - are there any errors reported? If there are, are they expected or not? Why?</p></li>
</ul>
<p>Now, let’s take a look at the NanoConsensus tracks that we have obtained from the <em>mop_consensus</em> module:</p>
<a class="reference internal image-reference" href="_images/snR36_KO−−−WT100_Cov100_Rep1.png"><img alt="_images/snR36_KO%E2%88%92%E2%88%92%E2%88%92WT100_Cov100_Rep1.png" src="_images/snR36_KO%E2%88%92%E2%88%92%E2%88%92WT100_Cov100_Rep1.png" style="width: 700px;" /></a>
<a class="reference internal image-reference" href="_images/snR36_KO−−−WT100_Cov50_Rep1.png"><img alt="_images/snR36_KO%E2%88%92%E2%88%92%E2%88%92WT100_Cov50_Rep1.png" src="_images/snR36_KO%E2%88%92%E2%88%92%E2%88%92WT100_Cov50_Rep1.png" style="width: 700px;" /></a>
<a class="reference internal image-reference" href="_images/snR36_KO−−−WT50_Cov50_Rep1.png"><img alt="_images/snR36_KO%E2%88%92%E2%88%92%E2%88%92WT50_Cov50_Rep1.png" src="_images/snR36_KO%E2%88%92%E2%88%92%E2%88%92WT50_Cov50_Rep1.png" style="width: 700px;" /></a>
<ul class="simple">
<li><p><strong>Question 10:</strong> Do you believe there might be false negatives/positive in the results? How would you deal with them?</p></li>
<li><p><strong>Question 11:</strong> If any, which position is differentially modified?</p></li>
</ul>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="nextflow.html" class="btn btn-neutral float-left" title="Nextflow" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, CRG.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>