<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Direct RNA Sequencing data analysis with Master of Pores 2 &mdash; ROPES summer school - CRG - May 2023  documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
    <link rel="canonical" href="https://biocorecrg.github.io/ropes-linux-mop2-2023/mop2_practical.html" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="author" title="About these documents" href="about.html" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Nextflow" href="nextflow.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            ROPES summer school - CRG - May 2023
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="about.html">About the course</a></li>
<li class="toctree-l1"><a class="reference internal" href="linux_basics.html">Introduction to Linux</a></li>
<li class="toctree-l1"><a class="reference internal" href="containers.html">Containers</a></li>
<li class="toctree-l1"><a class="reference internal" href="nextflow.html">Nextflow</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Direct RNA Sequencing data analysis with Master of Pores 2</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#basics-of-direct-rna-sequencing-drs">Basics of Direct RNA Sequencing (DRS)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#drs-data-analysis-with-master-of-pores-2-mop2">DRS data analysis with Master Of Pores 2 (MOP2)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#basic-preprocessing-module-mop-preprocessing">Basic preprocessing (module: <em>mop_preprocessing</em>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#polya-tail-length-analysis-module-mop-tail">PolyA tail length analysis (module: <em>mop_tail</em>)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#hands-on-1-mop-preprocess-and-mop-tail">Hands-on 1: <em>mop_preprocess</em> and <em>mop_tail</em></a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">ROPES summer school - CRG - May 2023</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Direct RNA Sequencing data analysis with Master of Pores 2</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/mop2_practical.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="direct-rna-sequencing-data-analysis-with-master-of-pores-2">
<span id="mop2-practical-page"></span><h1>Direct RNA Sequencing data analysis with Master of Pores 2<a class="headerlink" href="#direct-rna-sequencing-data-analysis-with-master-of-pores-2" title="Permalink to this heading"></a></h1>
<section id="basics-of-direct-rna-sequencing-drs">
<h2>Basics of Direct RNA Sequencing (DRS)<a class="headerlink" href="#basics-of-direct-rna-sequencing-drs" title="Permalink to this heading"></a></h2>
<p>Direct RNA sequencing is a technology developed by Oxford Nanopore Technologies (ONT) that allows the sequencing of native RNA molecules without the need of previous amplification nor fragmentation. Therefore, it provides information about the presence of RNA modifications, polyA tail length and composition at per-read level.</p>
<p>This method is based on the use of protein nanopores that are embedded into a membrane. As the molecule goes through the pore, it alters the ionic current that is applied to it. These changes are then stored in <strong>fast5 files</strong>.</p>
<p>TO ADD PORE HERE</p>
</section>
<section id="drs-data-analysis-with-master-of-pores-2-mop2">
<h2>DRS data analysis with Master Of Pores 2 (MOP2)<a class="headerlink" href="#drs-data-analysis-with-master-of-pores-2-mop2" title="Permalink to this heading"></a></h2>
<p>The complexity of analysing current intensity data together with the lack of systematic and reproducible pipelines have hindered the access of this technology to the general users. To overcome this limitation, we developed Master Of Pores 2 (MOP2) - a nextflow based workflow that simplifies the analysis of DRS datasets and aims to make it accessible to non-bioinformatic experts.</p>
<p>MOP2 can perform all steps required to analyse DRS data - from converting raw current intensities into multiple types of processed data to RNA modified sites detection and polyA tail length predictions. This pipeline consists of four modules: <em>mop_preprocess</em>, <em>mop_tail</em>, <em>mop_mod</em> and <em>mop_consensus</em>.</p>
<a class="reference internal image-reference" href="_images/MOP_scheme.png"><img alt="_images/MOP_scheme.png" src="_images/MOP_scheme.png" style="width: 700px;" /></a>
<section id="basic-preprocessing-module-mop-preprocessing">
<h3>Basic preprocessing (module: <em>mop_preprocessing</em>)<a class="headerlink" href="#basic-preprocessing-module-mop-preprocessing" title="Permalink to this heading"></a></h3>
<p>The pre-processing module is able to perform base-calling, mapping (either to a genome or a transcriptome), feature counting (per-gene level) and quality control. Furthermore, if required by the user, it can run demultiplexing, filtering and discovery of novel transcripts. As the final step, the workflow generates a final report of the performance and results of each of the steps performed.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Before proceeding to any other downstream analysis such as RNA modification detection and polyA tail analysis, this module <strong>must</strong> be executed.</p>
</div>
<p><strong>Analysis overview:</strong></p>
<ul>
<li><p><strong>Step 1a: Basecalling</strong></p>
<p>It is the process by which the current intensity data is translated into a nucleotide sequence by a machine learning algorithm called basecaller. Currently, the most widely used is <strong>Guppy</strong>, which was developed by ONT and it is only available if you are part of the ONT community. The model that Guppy uses to analyse RNA data is not modification aware and therefore, it can only identify the four canonical bases (A, U, C and G).</p>
<ul class="simple">
<li><p><strong>Input:</strong> Raw fast5 files</p></li>
<li><p><strong>Output:</strong> Basecalled fast5 and fastq files</p></li>
</ul>
</li>
<li><p><strong>Step 1b: Demultiplexing</strong></p>
<p>Demultiplexing is required when analysing a barcoded sample; otherwise, this step should be skipped. Here, <strong>Deeplexicon</strong> is used. This algorithm converts the barcode’s signal into an image, which is then classified based on a machine-learning approach.</p>
<ul class="simple">
<li><p><strong>Input:</strong> Raw fast5 files</p></li>
<li><p><strong>Output:</strong> Demuxed raw fast5 files</p></li>
</ul>
</li>
<li><p><strong>Step 2: Filtering</strong></p>
<p>Filter out reads based on either quality and/or length performed by <strong>Nanofilt</strong>. For RNA modification detection using DRS data, this step should be turned off as modified reads tend to have lower quality than unmodified ones and thus, filtering based on quality would bias the results.</p>
<ul class="simple">
<li><p><strong>Input:</strong> Fastq files</p></li>
<li><p><strong>Output:</strong> Filtered fastq files</p></li>
</ul>
</li>
<li><p><strong>Step 3: Alignment</strong></p>
<p>Mapping step performed by either <strong>minimap2</strong> or <strong>grapmap</strong>. Both can perform spliced or unspliced alignments. Briefly, we would use spliced alignments when using a genome as a reference and; unspliced for transcriptome. Furthermore, it has been reported that minimap2 fails to align highly modified reads and thus, it should not be used to analyse data from highly modified RNA species such as rRNAs.</p>
<ul class="simple">
<li><p><strong>Input:</strong> Fastq files and reference file (genome or transcriptome)</p></li>
<li><p><strong>Output:</strong> Bam (and bai) files</p></li>
</ul>
</li>
<li><p><strong>Step 4: Feature counts</strong></p>
<p>The software run by MOP2 to perform this step depends on the type of reference used in the mapping step. For transcriptome alignments, <strong>NanoCount</strong> is used and it reports per transcript abundances whereas for genome alignments, <strong>htseq-count</strong> is executed and it generates per-gene counts.</p>
<ul class="simple">
<li><p><strong>Input:</strong> Reference and alignment file (bam) for NanoCount // Reference, annotation (<em>.gtf) and alignment files (</em>.bam) for htseq-count</p></li>
<li><p><strong>Output:</strong> Transcript abundances’ estimations // Per-gene counts</p></li>
</ul>
</li>
<li><p><strong>Step 5: Transcript discovery</strong></p>
<p><strong>Bambu</strong> aims to identify novel transcripts from mapped reads. For more information about how to use this tool, please visit its <a class="reference external" href="https://github.com/GoekeLab/bambu#General-Usage">GitHub page</a>.</p>
<ul class="simple">
<li><p><strong>Input:</strong> Alignment (.bam), reference (.fa) and annotation (.gtf)</p></li>
<li><p><strong>Output:</strong> Transcript’s abundances and read id-transcript assignments</p></li>
</ul>
</li>
<li><p><strong>Step 6: Reporting and quality control</strong></p>
<p><strong>multiQC</strong> produces the final report, as a html page, which contains the quality control’s results generated by <strong>MinionQC</strong> together with the stats from the previous executed steps.</p>
<ul class="simple">
<li><p><strong>Input:</strong> all inputs and outputs</p></li>
<li><p><strong>Output:</strong> final report</p></li>
</ul>
</li>
</ul>
<p>We will show how to use and configurate this module in the next hands-on exercise.</p>
</section>
<section id="polya-tail-length-analysis-module-mop-tail">
<h3>PolyA tail length analysis (module: <em>mop_tail</em>)<a class="headerlink" href="#polya-tail-length-analysis-module-mop-tail" title="Permalink to this heading"></a></h3>
<p>This module estimates poly(A) tail length at read level provided by <strong>Tailfindr</strong> and/or <strong>Nanopolish</strong>. This workflow uses as input all the files generated by <em>mop_preprocess</em>.</p>
<p><strong>Analysis overview:</strong></p>
<ul class="simple">
<li><p><strong>Software 1: Tailfindr</strong></p>
<ul>
<li><p><strong>Input:</strong> Basecalled fast5 files and read id-gene assignments</p></li>
<li><p><strong>Output:</strong> PolyA tail length estimations</p></li>
</ul>
</li>
<li><p><strong>Software 2: Nanopolish</strong></p>
<ul>
<li><p><strong>Input:</strong> Fastq (.fq.gz), alignment (.bam) and reference (.fa) files</p></li>
<li><p><strong>Output:</strong> PolyA tail length estimations</p></li>
</ul>
</li>
</ul>
</section>
</section>
<section id="hands-on-1-mop-preprocess-and-mop-tail">
<h2>Hands-on 1: <em>mop_preprocess</em> and <em>mop_tail</em><a class="headerlink" href="#hands-on-1-mop-preprocess-and-mop-tail" title="Permalink to this heading"></a></h2>
<p>For installing the MoP2 pipeline and downloading guppy 3.4.2, please use the code below:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">git clone --depth 1 --recurse-submodules https://github.com/biocorecrg/MOP2.git</span>

<span class="go">cd MoP2; bash INSTALL.sh</span>
</pre></div>
</div>
<p>For this hands-on exercise, we will analyse several total RNA DRS samples from <em>Saccharomyces cerevisiae</em>:</p>
<ul class="simple">
<li><p>Sample 1: snR36 knock-out strain</p></li>
<li><p>Samples 2, 3 and 4: wild-type strains</p></li>
</ul>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="nextflow.html" class="btn btn-neutral float-left" title="Nextflow" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, CRG.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>